{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "make a new dir `cell_type_classification` and run the below code, using the R image in the attached .sif"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "singularity run --cleanenv -H $PWD -B /lustre1,/staging,${VSC_SCRATCH}/tmp:/tmp /staging/leuven/stg_00002/lcb/cflerin/containers/cflerin-seurat-4.0.3-plus.sif"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "use this for `pbmc.rna`\n",
    "```\n",
    "pbmc.rna = readRDS('/staging/leuven/stg_00002/lcb/cflerin/analysis/pbmc_atac/analysis2/scRNA/cell_type_classification/seurat/reference/pbmc_ssc_mat__integrated.rds')\n",
    "```"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# not necessary to run, just use the provided annotation: "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "# RNA-ATAC integration\n",
    "################################################################################\n",
    "# https://satijalab.org/seurat/articles/atacseq_integration_vignette.html\n",
    "\n",
    "library(Seurat)\n",
    "library(SeuratDisk)\n",
    "library(Signac)\n",
    "library(EnsDb.Hsapiens.v86)\n",
    "library(ggplot2)\n",
    "\n",
    "################################################################################\n",
    "################################################################################\n",
    "# reference data:\n",
    "################################################################################\n",
    "\n",
    "f_ref = '/lustre1/project/stg_00002/lcb/cflerin/analysis/pbmc_atac/analysis2/scRNA/cell_type_classification/seurat/reference/pbmc_ssc_mat.rds'\n",
    "f_meta = '/lustre1/project/stg_00002/lcb/cflerin/analysis/pbmc_atac/analysis2/scRNA/cell_type_classification/seurat/reference/pbmc_ssc_metadata.rds'\n",
    "\n",
    "pbmc.data <- readRDS(f_ref)\n",
    "pbmc.metadata <- readRDS(f_meta)\n",
    "\n",
    "pbmc <- CreateSeuratObject(counts = pbmc.data, meta.data = pbmc.metadata)\n",
    "pbmc <- subset(pbmc, subset = nFeature_RNA > 200)\n",
    "\n",
    "pbmc.list <- SplitObject(pbmc, split.by = \"Method\")\n",
    "#pbmc.list$bench = qdata\n",
    "\n",
    "for (i in names(pbmc.list)) {\n",
    "    pbmc.list[[i]] = NormalizeData(pbmc.list[[i]])\n",
    "    pbmc.list[[i]] = FindVariableFeatures(pbmc.list[[i]], selection.method = \"vst\", nfeatures = 2000, verbose = FALSE)\n",
    "    pbmc.list[[i]] = ScaleData(pbmc.list[[i]])\n",
    "    pbmc.list[[i]] = RunPCA(pbmc.list[[i]])\n",
    "    pbmc.list[[i]] = RunUMAP(pbmc.list[[i]], dims = 1:30)\n",
    "}\n",
    "\n",
    "##################################################\n",
    "### integrate reference:\n",
    "\n",
    "reference.list <- pbmc.list[\n",
    "    c(\"10x Chromium (v2) A\", \"10x Chromium (v2) B\", \"10x Chromium (v3)\", \"10x Chromium (v2)\")\n",
    "]\n",
    "#reference.list$query = qdata\n",
    "#pbmc.ref = pbmc.list$`10x Chromium (v3)`\n",
    "pbmc.anchors <- FindIntegrationAnchors(object.list = reference.list, dims = 1:30)\n",
    "\n",
    "pbmc.integrated <- IntegrateData(anchorset = pbmc.anchors, dims = 1:30)\n",
    "DefaultAssay(pbmc.integrated) <- \"integrated\"\n",
    "\n",
    "pbmc.integrated <- ScaleData(pbmc.integrated, verbose = FALSE)\n",
    "pbmc.integrated <- RunPCA(pbmc.integrated, npcs = 30, verbose = FALSE)\n",
    "pbmc.integrated <- RunUMAP(pbmc.integrated, reduction = \"pca\", dims = 1:30, verbose = FALSE)\n",
    "\n",
    "##################################################\n",
    "\n",
    "pbmc.rna = pbmc.list$`10x Chromium (v3)`\n",
    "\n",
    "saveRDS(pbmc.rna, file='pbmc_RNA_reference_10xv3.rds')\n",
    "\n",
    "# pbmc.rna <- NormalizeData(pbmc.rna)\n",
    "# pbmc.rna <- FindVariableFeatures(pbmc.rna)\n",
    "# pbmc.rna <- ScaleData(pbmc.rna)\n",
    "# pbmc.rna <- RunPCA(pbmc.rna)\n",
    "# pbmc.rna <- RunUMAP(pbmc.rna, dims = 1:30)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "and megakaryocytes need to be removed even from this one. so best to just use chris' RDS."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# run this"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# RNA-ATAC integration\n",
    "################################################################################\n",
    "# https://satijalab.org/seurat/articles/atacseq_integration_vignette.html\n",
    "\n",
    "library(Seurat)\n",
    "library(SeuratDisk)\n",
    "library(Signac)\n",
    "library(EnsDb.Hsapiens.v86)\n",
    "library(ggplot2)\n",
    "\n",
    "################################################################################\n",
    "################################################################################\n",
    "# reference data:\n",
    "################################################################################\n",
    "\n",
    "f_ref = '/lustre1/project/stg_00002/lcb/cflerin/analysis/pbmc_atac/analysis2/scRNA/cell_type_classification/seurat/reference/pbmc_ssc_mat.rds'\n",
    "f_meta = '/lustre1/project/stg_00002/lcb/cflerin/analysis/pbmc_atac/analysis2/scRNA/cell_type_classification/seurat/reference/pbmc_ssc_metadata.rds'\n",
    "\n",
    "\n",
    "pbmc.data <- readRDS(f_ref)\n",
    "pbmc.metadata <- readRDS(f_meta)\n",
    "\n",
    "pbmc <- CreateSeuratObject(counts = pbmc.data, meta.data = pbmc.metadata)\n",
    "pbmc <- subset(pbmc, subset = nFeature_RNA > 200)\n",
    "\n",
    "pbmc.list <- SplitObject(pbmc, split.by = \"Method\")\n",
    "#pbmc.list$bench = qdata\n",
    "\n",
    "for (i in names(pbmc.list)) {\n",
    "    pbmc.list[[i]] = NormalizeData(pbmc.list[[i]])\n",
    "    pbmc.list[[i]] = FindVariableFeatures(pbmc.list[[i]], selection.method = \"vst\", nfeatures = 2000, verbose = FALSE)\n",
    "    pbmc.list[[i]] = ScaleData(pbmc.list[[i]])\n",
    "    pbmc.list[[i]] = RunPCA(pbmc.list[[i]])\n",
    "    pbmc.list[[i]] = RunUMAP(pbmc.list[[i]], dims = 1:30)\n",
    "}\n",
    "\n",
    "##################################################\n",
    "### integrate reference:\n",
    "\n",
    "reference.list <- pbmc.list[\n",
    "    c(\"10x Chromium (v2) A\", \"10x Chromium (v2) B\", \"10x Chromium (v3)\", \"10x Chromium (v2)\")\n",
    "]\n",
    "#reference.list$query = qdata\n",
    "#pbmc.ref = pbmc.list$`10x Chromium (v3)`\n",
    "pbmc.anchors <- FindIntegrationAnchors(object.list = reference.list, dims = 1:30)\n",
    "\n",
    "pbmc.integrated <- IntegrateData(anchorset = pbmc.anchors, dims = 1:30)\n",
    "DefaultAssay(pbmc.integrated) <- \"integrated\"\n",
    "\n",
    "pbmc.integrated <- ScaleData(pbmc.integrated, verbose = FALSE)\n",
    "pbmc.integrated <- RunPCA(pbmc.integrated, npcs = 30, verbose = FALSE)\n",
    "pbmc.integrated <- RunUMAP(pbmc.integrated, reduction = \"pca\", dims = 1:30, verbose = FALSE)\n",
    "\n",
    "##################################################\n",
    "\n",
    "# pbmc.rna = pbmc.list$`10x Chromium (v3)`\n",
    "\n",
    "pbmc.rna = readRDS('pbmc_ssc_mat__integrated.rds')\n",
    "\n",
    "# pbmc.rna <- NormalizeData(pbmc.rna)\n",
    "# pbmc.rna <- FindVariableFeatures(pbmc.rna)\n",
    "# pbmc.rna <- ScaleData(pbmc.rna)\n",
    "# pbmc.rna <- RunPCA(pbmc.rna)\n",
    "# pbmc.rna <- RunUMAP(pbmc.rna, dims = 1:30)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "then run this per sample, change:\n",
    "* f_loom\n",
    "* f_frag\n",
    "* sample_id"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "################################################################################\n",
    "# ATAC\n",
    "################################################################################\n",
    "\n",
    "### get data from loom:\n",
    "\n",
    "f_loom = 'downstream_analysis/region_acc_loom/Hydrop_1__libDS_region_accessibility.loom'\n",
    "f_frag = 'fragments_mm/VIB_Hydrop_1.sinto.mm.fragments.sorted.tsv.gz'\n",
    "\n",
    "atacloomcon <- Connect(filename = f_loom, mode = \"r\")\n",
    "atacloomcon\n",
    "atac_tmp <- as.Seurat(atacloomcon, assay='ATAC')\n",
    "atacloomcon$close_all()\n",
    "\n",
    "# correctly parse regions (default delims are '-','-')\n",
    "regions = StringToGRanges(\n",
    "    rownames(GetAssayData(atac_tmp, slot = \"counts\", assay='ATAC')),\n",
    "    sep=c(':','-')\n",
    "    )\n",
    "# create chromatin assay\n",
    "chromatinassay = CreateChromatinAssay(\n",
    "    counts=GetAssayData(atac_tmp, slot = \"counts\", assay='ATAC'),\n",
    "    # genome='hg38',\n",
    "    fragments = f_frag,\n",
    "    ranges=regions\n",
    "    )\n",
    "    #annotation=annotation)\n",
    "\n",
    "atac <- CreateSeuratObject(counts = chromatinassay, assay='ATAC')\n",
    "\n",
    "library(stringr)\n",
    "\n",
    "\n",
    "# ATAC analysis add gene annotation information\n",
    "annotations <- GetGRangesFromEnsDb(ensdb = EnsDb.Hsapiens.v86)\n",
    "\n",
    "# use chr prefix:\n",
    "ucsc.levels <- str_replace(string=paste(\"chr\",seqlevels(annotations),sep=\"\"), pattern=\"chrMT\", replacement=\"chrM\")\n",
    "seqlevels(annotations) <- ucsc.levels\n",
    "\n",
    "genome(annotations) <- \"hg38\"\n",
    "\n",
    "saveRDS(annotations, file=\"granges_annotation.rds\")\n",
    "\n",
    "Annotation(atac) <- annotations\n",
    "\n",
    "# We exclude the first dimension as this is typically correlated with sequencing depth\n",
    "atac <- RunTFIDF(atac)\n",
    "atac <- FindTopFeatures(atac, min.cutoff = \"q0\")\n",
    "atac <- RunSVD(atac)\n",
    "atac <- RunUMAP(atac, reduction = \"lsi\", dims = 2:30, reduction.name = \"umap.atac\", reduction.key = \"atacUMAP_\")\n",
    "\n",
    "##################################################\n",
    "# Identify RNA-ATAC anchors\n",
    "\n",
    "# quantify gene activity\n",
    "gene.activities <- GeneActivity(atac, features = VariableFeatures(pbmc.rna))\n",
    "# gene.activities <- gene.activities[rownames(gene.activities)!=\"\",]\n",
    "\n",
    "# add gene activities as a new assay\n",
    "atac[[\"ACTIVITY\"]] <- CreateAssayObject(counts = gene.activities)\n",
    "\n",
    "# normalize gene activities\n",
    "DefaultAssay(atac) <- \"ACTIVITY\"\n",
    "atac <- NormalizeData(atac)\n",
    "atac <- ScaleData(atac, features = rownames(atac))\n",
    "\n",
    "# Identify anchors\n",
    "transfer.anchors <- FindTransferAnchors(reference = pbmc.rna,\n",
    "    query = atac,\n",
    "    features = VariableFeatures(object = pbmc.rna),\n",
    "    reference.assay = \"RNA\",\n",
    "    query.assay = \"ACTIVITY\",\n",
    "    reduction = \"cca\")\n",
    "\n",
    "\n",
    "# predict celltype\n",
    "celltype.predictions <- TransferData(\n",
    "    anchorset = transfer.anchors,\n",
    "    refdata = pbmc.rna$CellType,\n",
    "    weight.reduction = atac[[\"lsi\"]],\n",
    "    dims = 2:30)\n",
    "\n",
    "pbmc.atac <- AddMetaData(pbmc.atac, metadata = celltype.predictions)\n",
    "\n",
    "\n",
    "\n",
    "md = celltype.predictions\n",
    "sample_id = 'VIB_Hydrop_1'\n",
    "pred_thr = 0.7\n",
    "\n",
    "tmp = data.frame(\n",
    "      composite_sample_id = paste0(rownames(md),'-',sample_id),\n",
    "      barcode = rownames(md),\n",
    "      sample_id = sample_id,\n",
    "      cell_type = md$predicted.id,\n",
    "      cell_type_pred_score = md$prediction.score.max\n",
    "      )\n",
    "tmp$cell_type_hiconf_70 = tmp$cell_type\n",
    "tmp$cell_type_hiconf_70[tmp$cell_type_pred_score<pred_thr] = 'Unknown'\n",
    "\n",
    "table(tmp$cell_type)\n",
    "table(tmp$cell_type_hiconf_70)\n",
    "\n",
    "write.table(tmp, file=paste0('pycistopic_consensus_peaks/cell_type_classification/predictions/',sample_id,'__cell_type_seurat.IMPUTEDACCESSIBILITY.txt'), sep='\\t', row.names=FALSE, quote=FALSE)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "when running\n",
    "```\n",
    "> chromatinassay = CreateChromatinAssay(\n",
    "+     counts=GetAssayData(atac_tmp, slot = \"counts\", assay='ATAC'),\n",
    "+     genome='hg38',\n",
    "+     fragments = f_frag,\n",
    "+     ranges=regions\n",
    "+     )\n",
    "```\n",
    "i get the error\n",
    "```\n",
    "Computing hash\n",
    "Checking for 2208 cell barcodes\n",
    "Error in CreateFragmentObject(path = fragments, cells = cells, validate.fragments = validate.fragments,  :\n",
    "  Not all cells requested could be found in the fragment file.\n",
    "In addition: Warning message:\n",
    "In CreateChromatinAssay(counts = GetAssayData(atac_tmp, slot = \"counts\",  :\n",
    "  Overlapping ranges supplied. Ranges should be non-overlapping.\n",
    "```\n",
    "and `chromatinassay` is empty. \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "chr1\t10151\t10183\tGCGTTGACTTCGTCGACGTA-12\t1\n",
      "chr1\t16228\t16272\tATTGTAGAGGTAGGAGTCAA-12\t6\n",
      "chr1\t17490\t17527\tTTGTACTGCAAGGTGCCAGA-12\t3\n",
      "chr1\t30857\t31030\tCATGAACGTTGTGACGTGGA-12\t1\n",
      "chr1\t68280\t68311\tCATTGGTGCAGGCGAGATTC-12\t1\n",
      "chr1\t74871\t74900\tATCATCCGTTAATAGGCAGG-11\t11\n",
      "chr1\t79643\t79855\tGGAGCGGATTCAAGGTCGAT-11\t2\n",
      "chr1\t88228\t88302\tAAGACCAAGCGCTGATGGTA-12\t4\n",
      "chr1\t89893\t89944\tTCCGATTCAGCACGTTAAGA-12\t3\n",
      "chr1\t91177\t91237\tCTCCTCACTACTTGAAGAAG-12\t4\n",
      "\n",
      "gzip: stdout: Broken pipe\n"
     ]
    }
   ],
   "source": [
    "zcat fragments_postbap/VIB_Hydrop_1.sinto.mm.fragments.tsv.gz | head"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "this error doesn't make sense, because the barcodes are present in the fragments file."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "the output of the above code can be read in the cell_type_redictions.ipynb notebook"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "it turns out the loom file writing had a bug in it that did not add the `-1?` do the barcode. i fixed this in the loom writing."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "now that this is fixed, i got an error\n",
    "\n",
    "```\n",
    "> chromatinassay = CreateChromatinAssay(\n",
    "+     counts=GetAssayData(atac_tmp, slot = \"counts\", assay='ATAC'),\n",
    "+     genome='hg38',\n",
    "+     fragments = f_frag,\n",
    "+     ranges=regions\n",
    "+     )\n",
    "Computing hash\n",
    "Checking for 2208 cell barcodes\n",
    "Error in .order_seqlevels(chrom_sizes[, \"chrom\"]) :\n",
    "  !anyNA(m32) is not TRUE\n",
    "In addition: Warning message:\n",
    "In CreateChromatinAssay(counts = GetAssayData(atac_tmp, slot = \"counts\",  :\n",
    "  Overlapping ranges supplied. Ranges should be non-overlapping.\n",
    "```\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "it turns out that this is a known bug (https://github.com/timoast/signac/issues/687) and that commenting out the line `genome='hg38'`"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# below code isn't necessary"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "hist(celltype.predictions$prediction.score.max)\n",
    "\n",
    "################################################################################\n",
    "################################################################################\n",
    "\n",
    "object=atac\n",
    "features=VariableFeatures(pbmc.rna)\n",
    "biotypes = \"protein_coding\"\n",
    "max.width = 500000\n",
    "extend.upstream = 2000\n",
    "extend.downstream = 0\n",
    "\n",
    "assay <- 'ATAC'#SetIfNull(x = assay, y = DefaultAssay(object = object))\n",
    "annotation <- Annotation(object = object[[assay]])\n",
    "transcripts <- Signac:::CollapseToLongestTranscript(ranges = annotation)\n",
    "\n",
    "if (!is.null(x = biotypes)) {\n",
    "    transcripts <- transcripts[transcripts$gene_biotype %in% biotypes]\n",
    "}\n",
    "\n",
    "if (!is.null(x = features)) {\n",
    "    transcripts <- transcripts[transcripts$gene_name %in% features]\n",
    "}\n",
    "if (!is.null(x = max.width)) {\n",
    "    transcript.keep <- which(x = width(x = transcripts) < max.width)\n",
    "    transcripts <- transcripts[transcript.keep]\n",
    "}\n",
    "\n",
    "transcripts <- Extend(\n",
    "    x = transcripts,\n",
    "    upstream = extend.upstream,\n",
    "    downstream = extend.downstream\n",
    ")\n",
    "\n",
    "frags <- Fragments(object = object[[assay]])\n",
    "cells <- colnames(x = object[[assay]])\n",
    "counts <- FeatureMatrix(\n",
    "    fragments = frags,\n",
    "    features = transcripts,\n",
    "    cells = cells\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "singularity run -B /lustre1,/staging /staging/leuven/stg_00002/lcb/cflerin/containers/cflerin-seurat-4.0.3-plus.sif R\n",
    "\n",
    "singularity run --cleanenv -H $PWD -B /lustre1,/staging,${VSC_SCRATCH}/tmp:/tmp /staging/leuven/stg_00002/lcb/cflerin/containers/cflerin-seurat-4.0.3-plus.sif R\n",
    "\n",
    "################################################################################\n",
    "\n",
    "library(Seurat)\n",
    "library(SeuratDisk)\n",
    "library(Signac)\n",
    "library(ggplot2)\n",
    "#library(EnsDb.Hsapiens.v86)\n",
    "#library(stringr)\n",
    "\n",
    "\n",
    "\n",
    "#pbmc.rna = readRDS('/staging/leuven/stg_00002/lcb/cflerin/analysis/pbmc_atac/analysis2/data_freeze_Jun2021/cell_type_classification/pbmc_RNA_reference_10xv3.rds')\n",
    "annotations = readRDS(\"/staging/leuven/stg_00002/lcb/cflerin/analysis/pbmc_atac/analysis2/data_freeze_Jun2021/cell_type_classification/granges_annotation.rds\")\n",
    "pred_thr = 0.7\n",
    "\n",
    "pbmc.rna = readRDS('/staging/leuven/stg_00002/lcb/cflerin/analysis/pbmc_atac/analysis2/scRNA/cell_type_classification/seurat/reference/pbmc_ssc_mat__integrated.rds')\n",
    "\n",
    "\n",
    "sample_ids = c(\n",
    "    'Broad_1', 'Broad_2',\n",
    "    'Broad_mito_1',\n",
    "    'Broad_mito_2', 'CNAG_1', 'CNAG_2', 's3atac', 'Sanger_1', 'Sanger_2', 'Stanford_1', 'Stanford_2', 'VIB_1', 'VIB_2')\n",
    "\n",
    "\n",
    "# f_loomdir = '/staging/leuven/stg_00002/lcb/cflerin/analysis/pbmc_atac/analysis2/data_freeze_Jun2021_library_downsampled/atac_qc_multiplet_merged/jupyter/cell_region_loom/'\n",
    "f_fragdir = '/staging/leuven/stg_00002/lcb/cflerin/analysis/pbmc_atac/analysis2/data_freeze_Jun2021/atac_preprocess/multiplet_tagged/fragments/'\n",
    "\n",
    "\n",
    "f_loomdir = '/staging/leuven/stg_00002/lcb/cflerin/analysis/pbmc_atac/analysis2/data_freeze_Jun2021_cell_downsampled/atac_qc/jupyter/cell_region_loom__screen/'\n",
    "\n",
    "f_outdir = 'predictions__screen/'\n",
    "\n",
    "################################################################################\n",
    "################################################################################\n",
    "for(sample_id in sample_ids) {\n",
    "    #break}\n",
    "\n",
    "print(sample_id)\n",
    "f_loom = paste0(f_loomdir, sample_id, 'cell_region-all.loom')\n",
    "f_frag = paste0(f_fragdir, sample_id, '.sinto.mm.fragments.tsv.gz')\n",
    "\n",
    "### get data from loom:\n",
    "atacloomcon <- Connect(filename = f_loom, mode = \"r\")\n",
    "atacloomcon\n",
    "atac_tmp <- as.Seurat(atacloomcon, assay='ATAC')\n",
    "atacloomcon$close_all()\n",
    "\n",
    "# create chromatin assay\n",
    "chromatinassay = CreateChromatinAssay(\n",
    "    counts=GetAssayData(atac_tmp, slot = \"counts\", assay='ATAC'),\n",
    "    genome='hg38',\n",
    "    fragments = f_frag,\n",
    "    sep=c(':','-'),\n",
    "    validate.fragments=FALSE\n",
    "    )\n",
    "\n",
    "atac <- CreateSeuratObject(counts = chromatinassay, assay='ATAC')\n",
    "Annotation(atac) <- annotations\n",
    "\n",
    "# We exclude the first dimension as this is typically correlated with sequencing depth\n",
    "atac <- RunTFIDF(atac)\n",
    "atac <- FindTopFeatures(atac, min.cutoff = \"q0\")\n",
    "atac <- RunSVD(atac)\n",
    "atac <- RunUMAP(atac, reduction = \"lsi\", dims = 2:30, reduction.name = \"umap.atac\", reduction.key = \"atacUMAP_\")\n",
    "\n",
    "\n",
    "##################################################\n",
    "# Identify RNA-ATAC anchors\n",
    "\n",
    "# quantify gene activity\n",
    "gene.activities <- GeneActivity(atac, features = VariableFeatures(pbmc.rna))\n",
    "# gene.activities <- gene.activities[rownames(gene.activities)!=\"\",]\n",
    "\n",
    "# add gene activities as a new assay\n",
    "atac[[\"ACTIVITY\"]] <- CreateAssayObject(counts = gene.activities)\n",
    "\n",
    "# normalize gene activities\n",
    "DefaultAssay(atac) <- \"ACTIVITY\"\n",
    "atac <- NormalizeData(atac)\n",
    "atac <- ScaleData(atac, features = rownames(atac))\n",
    "\n",
    "# Identify anchors\n",
    "transfer.anchors <- FindTransferAnchors(reference = pbmc.rna,\n",
    "    query = atac,\n",
    "    features = VariableFeatures(object = pbmc.rna),\n",
    "    reference.assay = \"RNA\",\n",
    "    query.assay = \"ACTIVITY\",\n",
    "    reduction = \"cca\")\n",
    "\n",
    "\n",
    "# predict celltype\n",
    "celltype.predictions <- TransferData(\n",
    "    anchorset = transfer.anchors,\n",
    "    refdata = pbmc.rna$CellType,\n",
    "    weight.reduction = atac[[\"lsi\"]],\n",
    "    dims = 2:30)\n",
    "\n",
    "#pbmc.atac <- AddMetaData(pbmc.atac, metadata = celltype.predictions)\n",
    "\n",
    "md = celltype.predictions\n",
    "\n",
    "tmp = data.frame(\n",
    "      composite_sample_id = paste0(rownames(md),'-',sample_id),\n",
    "      barcode = rownames(md),\n",
    "      sample_id = sample_id,\n",
    "      cell_type = md$predicted.id,\n",
    "      cell_type_pred_score = md$prediction.score.max\n",
    "      )\n",
    "tmp$cell_type_hiconf_70 = tmp$cell_type\n",
    "tmp$cell_type_hiconf_70[tmp$cell_type_pred_score<pred_thr] = 'Unknown'\n",
    "\n",
    "\n",
    "write.table(tmp,\n",
    "            file=paste0(f_outdir,sample_id,'__cell_type_seurat.txt'),\n",
    "            sep='\\t', row.names=FALSE, quote=FALSE\n",
    "            )\n",
    "\n",
    "}\n",
    "\n",
    "#table(tmp$cell_type)\n",
    "#table(tmp$cell_type_hiconf_70)\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Bash",
   "language": "bash",
   "name": "bash"
  },
  "language_info": {
   "codemirror_mode": "shell",
   "file_extension": ".sh",
   "mimetype": "text/x-sh",
   "name": "bash"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
