{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "8bdd2493-0fd7-45ca-b73b-9c6d0c2bd941",
   "metadata": {},
   "source": [
    "# pycisTopic analysis\n",
    "\n",
    "Cell downsampled dataset, using consensus peak regions."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "53c07230-73b7-41e2-af18-0bb31aa7be57",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'0.1.dev295+gbd4bf4d.d20210830'"
      ]
     },
     "execution_count": 1,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import pycisTopic\n",
    "pycisTopic.__version__"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "1f5fdcc5-0d8f-48e5-8c29-302acbb9ad66",
   "metadata": {},
   "outputs": [],
   "source": [
    "import warnings\n",
    "warnings.filterwarnings('ignore')\n",
    "warnings.simplefilter('ignore')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "f7fc6d2d-f633-40d2-8c64-ae88679c7b4a",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pickle\n",
    "import pandas as pd"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "55296406-e6b7-4e59-ab84-163be3093c26",
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "wdir = '/staging/leuven/stg_00002/lcb/cflerin/analysis/pbmc_atac/analysis2/data_freeze_Jun2021_cell_downsampled/atac_qc/jupyter/'\n",
    "os.chdir( wdir )"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9e50af45-d2a4-4f12-8161-08c41a5bacf3",
   "metadata": {},
   "source": [
    "---"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "339e3f62-398d-4cec-bcc4-16dbbb5046f3",
   "metadata": {},
   "outputs": [],
   "source": [
    "fragments_dict = {\n",
    "'Broad_1':      '/staging/leuven/stg_00002/lcb/cflerin/analysis/pbmc_atac/analysis2/data_freeze_Jun2021_cell_downsampled/fragments/Broad_1.sinto.fragments.tsv.gz',\n",
    "'Broad_2':      '/staging/leuven/stg_00002/lcb/cflerin/analysis/pbmc_atac/analysis2/data_freeze_Jun2021_cell_downsampled/fragments/Broad_2.sinto.fragments.tsv.gz',\n",
    "'Broad_mito_1': '/staging/leuven/stg_00002/lcb/cflerin/analysis/pbmc_atac/analysis2/data_freeze_Jun2021_cell_downsampled/fragments/Broad_mito_1.sinto.fragments.tsv.gz',\n",
    "'Broad_mito_2': '/staging/leuven/stg_00002/lcb/cflerin/analysis/pbmc_atac/analysis2/data_freeze_Jun2021_cell_downsampled/fragments/Broad_mito_2.sinto.fragments.tsv.gz',\n",
    "'CNAG_1':       '/staging/leuven/stg_00002/lcb/cflerin/analysis/pbmc_atac/analysis2/data_freeze_Jun2021_cell_downsampled/fragments/CNAG_1.sinto.fragments.tsv.gz',\n",
    "'CNAG_2':       '/staging/leuven/stg_00002/lcb/cflerin/analysis/pbmc_atac/analysis2/data_freeze_Jun2021_cell_downsampled/fragments/CNAG_2.sinto.fragments.tsv.gz',\n",
    "'s3atac':       '/staging/leuven/stg_00002/lcb/cflerin/analysis/pbmc_atac/analysis2/data_freeze_Jun2021_cell_downsampled/fragments/s3atac.sinto.fragments.tsv.gz',\n",
    "'Sanger_1':     '/staging/leuven/stg_00002/lcb/cflerin/analysis/pbmc_atac/analysis2/data_freeze_Jun2021_cell_downsampled/fragments/Sanger_1.sinto.fragments.tsv.gz',\n",
    "'Sanger_2':     '/staging/leuven/stg_00002/lcb/cflerin/analysis/pbmc_atac/analysis2/data_freeze_Jun2021_cell_downsampled/fragments/Sanger_2.sinto.fragments.tsv.gz',\n",
    "'Stanford_1':   '/staging/leuven/stg_00002/lcb/cflerin/analysis/pbmc_atac/analysis2/data_freeze_Jun2021_cell_downsampled/fragments/Stanford_1.sinto.fragments.tsv.gz',\n",
    "'Stanford_2':   '/staging/leuven/stg_00002/lcb/cflerin/analysis/pbmc_atac/analysis2/data_freeze_Jun2021_cell_downsampled/fragments/Stanford_2.sinto.fragments.tsv.gz',\n",
    "'VIB_1':        '/staging/leuven/stg_00002/lcb/cflerin/analysis/pbmc_atac/analysis2/data_freeze_Jun2021_cell_downsampled/fragments/VIB_1.sinto.fragments.tsv.gz',\n",
    "'VIB_2':        '/staging/leuven/stg_00002/lcb/cflerin/analysis/pbmc_atac/analysis2/data_freeze_Jun2021_cell_downsampled/fragments/VIB_2.sinto.fragments.tsv.gz',\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "e963ecfc-ed5f-41b9-8a44-8f8c082d70a8",
   "metadata": {},
   "outputs": [],
   "source": [
    "consensus_peaks_dict = {\n",
    "'Broad_1':      '/staging/leuven/stg_00002/lcb/cflerin/analysis/pbmc_atac/analysis2/data_freeze_Jun2021_cell_downsampled/atac_qc/jupyter/consensus_peak_calling/Broad_1/Broad_1__consensus_regions.bed',\n",
    "'Broad_2':      '/staging/leuven/stg_00002/lcb/cflerin/analysis/pbmc_atac/analysis2/data_freeze_Jun2021_cell_downsampled/atac_qc/jupyter/consensus_peak_calling/Broad_2/Broad_2__consensus_regions.bed',\n",
    "'Broad_mito_1': '/staging/leuven/stg_00002/lcb/cflerin/analysis/pbmc_atac/analysis2/data_freeze_Jun2021_cell_downsampled/atac_qc/jupyter/consensus_peak_calling/Broad_mito_1/Broad_mito_1__consensus_regions.bed',\n",
    "'Broad_mito_2': '/staging/leuven/stg_00002/lcb/cflerin/analysis/pbmc_atac/analysis2/data_freeze_Jun2021_cell_downsampled/atac_qc/jupyter/consensus_peak_calling/Broad_mito_2/Broad_mito_2__consensus_regions.bed',\n",
    "'CNAG_1':       '/staging/leuven/stg_00002/lcb/cflerin/analysis/pbmc_atac/analysis2/data_freeze_Jun2021_cell_downsampled/atac_qc/jupyter/consensus_peak_calling/CNAG_1/CNAG_1__consensus_regions.bed',\n",
    "'CNAG_2':       '/staging/leuven/stg_00002/lcb/cflerin/analysis/pbmc_atac/analysis2/data_freeze_Jun2021_cell_downsampled/atac_qc/jupyter/consensus_peak_calling/CNAG_2/CNAG_2__consensus_regions.bed',\n",
    "'merged':       '/staging/leuven/stg_00002/lcb/cflerin/analysis/pbmc_atac/analysis2/data_freeze_Jun2021_cell_downsampled/atac_qc/jupyter/consensus_peak_calling/merged/merged__consensus_regions.bed',\n",
    "'s3atac':       '/staging/leuven/stg_00002/lcb/cflerin/analysis/pbmc_atac/analysis2/data_freeze_Jun2021_cell_downsampled/atac_qc/jupyter/consensus_peak_calling/s3atac/s3atac__consensus_regions.bed',\n",
    "'Sanger_1':     '/staging/leuven/stg_00002/lcb/cflerin/analysis/pbmc_atac/analysis2/data_freeze_Jun2021_cell_downsampled/atac_qc/jupyter/consensus_peak_calling/Sanger_1/Sanger_1__consensus_regions.bed',\n",
    "'Sanger_2':     '/staging/leuven/stg_00002/lcb/cflerin/analysis/pbmc_atac/analysis2/data_freeze_Jun2021_cell_downsampled/atac_qc/jupyter/consensus_peak_calling/Sanger_2/Sanger_2__consensus_regions.bed',\n",
    "'Stanford_1':   '/staging/leuven/stg_00002/lcb/cflerin/analysis/pbmc_atac/analysis2/data_freeze_Jun2021_cell_downsampled/atac_qc/jupyter/consensus_peak_calling/Stanford_1/Stanford_1__consensus_regions.bed',\n",
    "'Stanford_2':   '/staging/leuven/stg_00002/lcb/cflerin/analysis/pbmc_atac/analysis2/data_freeze_Jun2021_cell_downsampled/atac_qc/jupyter/consensus_peak_calling/Stanford_2/Stanford_2__consensus_regions.bed',\n",
    "'VIB_1':        '/staging/leuven/stg_00002/lcb/cflerin/analysis/pbmc_atac/analysis2/data_freeze_Jun2021_cell_downsampled/atac_qc/jupyter/consensus_peak_calling/VIB_1/VIB_1__consensus_regions.bed',\n",
    "'VIB_2':        '/staging/leuven/stg_00002/lcb/cflerin/analysis/pbmc_atac/analysis2/data_freeze_Jun2021_cell_downsampled/atac_qc/jupyter/consensus_peak_calling/VIB_2/VIB_2__consensus_regions.bed',    \n",
    "}"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "fac6eb51-433d-424a-a55c-d8ab3c8bf18b",
   "metadata": {},
   "source": [
    "## Library QC metrics"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "25dbdb6e-7416-45ab-a0a9-c613b9cb4b31",
   "metadata": {},
   "outputs": [],
   "source": [
    "from pycisTopic.qc import compute_qc_stats"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "1f0aede4-8013-4fda-928d-299e0933fc2c",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Start</th>\n",
       "      <th>Strand</th>\n",
       "      <th>Gene</th>\n",
       "      <th>Transcript_type</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>8986</th>\n",
       "      <td>chrY</td>\n",
       "      <td>9337464</td>\n",
       "      <td>1</td>\n",
       "      <td>TSPY4</td>\n",
       "      <td>protein_coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8987</th>\n",
       "      <td>chrY</td>\n",
       "      <td>9337510</td>\n",
       "      <td>1</td>\n",
       "      <td>TSPY4</td>\n",
       "      <td>protein_coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9024</th>\n",
       "      <td>chrY</td>\n",
       "      <td>22490397</td>\n",
       "      <td>1</td>\n",
       "      <td>PRY</td>\n",
       "      <td>protein_coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9100</th>\n",
       "      <td>chrY</td>\n",
       "      <td>14056227</td>\n",
       "      <td>1</td>\n",
       "      <td>VCY1B</td>\n",
       "      <td>protein_coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9136</th>\n",
       "      <td>chrY</td>\n",
       "      <td>12662368</td>\n",
       "      <td>1</td>\n",
       "      <td>USP9Y</td>\n",
       "      <td>protein_coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>236024</th>\n",
       "      <td>chr1</td>\n",
       "      <td>36479519</td>\n",
       "      <td>-1</td>\n",
       "      <td>CSF3R</td>\n",
       "      <td>protein_coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>236025</th>\n",
       "      <td>chr1</td>\n",
       "      <td>36471474</td>\n",
       "      <td>-1</td>\n",
       "      <td>CSF3R</td>\n",
       "      <td>protein_coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>236033</th>\n",
       "      <td>chr1</td>\n",
       "      <td>36482051</td>\n",
       "      <td>-1</td>\n",
       "      <td>CSF3R</td>\n",
       "      <td>protein_coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>236034</th>\n",
       "      <td>chr1</td>\n",
       "      <td>36323645</td>\n",
       "      <td>-1</td>\n",
       "      <td>EVA1B</td>\n",
       "      <td>protein_coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>236035</th>\n",
       "      <td>chr1</td>\n",
       "      <td>36324154</td>\n",
       "      <td>-1</td>\n",
       "      <td>EVA1B</td>\n",
       "      <td>protein_coding</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>77249 rows × 5 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       Chromosome     Start  Strand   Gene Transcript_type\n",
       "8986         chrY   9337464       1  TSPY4  protein_coding\n",
       "8987         chrY   9337510       1  TSPY4  protein_coding\n",
       "9024         chrY  22490397       1    PRY  protein_coding\n",
       "9100         chrY  14056227       1  VCY1B  protein_coding\n",
       "9136         chrY  12662368       1  USP9Y  protein_coding\n",
       "...           ...       ...     ...    ...             ...\n",
       "236024       chr1  36479519      -1  CSF3R  protein_coding\n",
       "236025       chr1  36471474      -1  CSF3R  protein_coding\n",
       "236033       chr1  36482051      -1  CSF3R  protein_coding\n",
       "236034       chr1  36323645      -1  EVA1B  protein_coding\n",
       "236035       chr1  36324154      -1  EVA1B  protein_coding\n",
       "\n",
       "[77249 rows x 5 columns]"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "f_biomart = '/staging/leuven/stg_00002/lcb/cflerin/analysis/pbmc_atac/analysis2/data_freeze_Jun2021/atac_qc_multiplet_merged/jupyter/biomart_annot.pickle'\n",
    "with open(f_biomart, 'rb') as f:\n",
    "    annot = pickle.load(f)\n",
    "annot"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "cd0753e9-c87b-4541-a21c-1cf9369601b6",
   "metadata": {
    "collapsed": true,
    "jupyter": {
     "outputs_hidden": true
    },
    "tags": []
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "2021-09-14 15:36:54,014\tINFO services.py:1263 -- View the Ray dashboard at \u001b[1m\u001b[32mhttp://127.0.0.1:8265\u001b[39m\u001b[22m\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\u001b[2m\u001b[36m(pid=4301)\u001b[0m 2021-09-14 15:36:56,933 cisTopic     INFO     Reading Broad_1\n",
      "\u001b[2m\u001b[36m(pid=4298)\u001b[0m 2021-09-14 15:36:56,842 cisTopic     INFO     Reading Broad_2\n",
      "\u001b[2m\u001b[36m(pid=4300)\u001b[0m 2021-09-14 15:36:56,842 cisTopic     INFO     Reading Broad_mito_1\n",
      "\u001b[2m\u001b[36m(pid=4299)\u001b[0m 2021-09-14 15:36:56,932 cisTopic     INFO     Reading CNAG_1\n",
      "\u001b[2m\u001b[36m(pid=4297)\u001b[0m 2021-09-14 15:36:56,834 cisTopic     INFO     Reading Broad_mito_2\n",
      "\u001b[2m\u001b[36m(pid=4299)\u001b[0m 2021-09-14 15:37:09,620 cisTopic     INFO     Computing barcode rank plot for CNAG_1\n",
      "\u001b[2m\u001b[36m(pid=4299)\u001b[0m 2021-09-14 15:37:09,620 cisTopic     INFO     Counting fragments\n",
      "\u001b[2m\u001b[36m(pid=4299)\u001b[0m 2021-09-14 15:37:10,188 cisTopic     INFO     Marking barcodes with more than 100\n",
      "\u001b[2m\u001b[36m(pid=4299)\u001b[0m 2021-09-14 15:37:10,211 numexpr.utils INFO     Note: NumExpr detected 36 cores but \"NUMEXPR_MAX_THREADS\" not set, so enforcing safe limit of 8.\n",
      "\u001b[2m\u001b[36m(pid=4299)\u001b[0m 2021-09-14 15:37:10,212 cisTopic     INFO     Returning plot data\n",
      "\u001b[2m\u001b[36m(pid=4299)\u001b[0m 2021-09-14 15:37:10,212 cisTopic     INFO     Returning valid barcodes\n",
      "\u001b[2m\u001b[36m(pid=4299)\u001b[0m 2021-09-14 15:37:11,047 cisTopic     INFO     Computing duplicate rate plot for CNAG_1\n",
      "\u001b[2m\u001b[36m(pid=4299)\u001b[0m 2021-09-14 15:37:11,910 cisTopic     INFO     Return plot data\n",
      "\u001b[2m\u001b[36m(pid=4299)\u001b[0m 2021-09-14 15:37:11,993 cisTopic     INFO     Computing insert size distribution for CNAG_1\n",
      "\u001b[2m\u001b[36m(pid=4299)\u001b[0m 2021-09-14 15:37:11,993 cisTopic     INFO     Counting fragments\n",
      "\u001b[2m\u001b[36m(pid=4299)\u001b[0m 2021-09-14 15:37:12,357 cisTopic     INFO     Returning plot data\n",
      "\u001b[2m\u001b[36m(pid=4297)\u001b[0m 2021-09-14 15:37:12,366 cisTopic     INFO     Computing barcode rank plot for Broad_mito_2\n",
      "\u001b[2m\u001b[36m(pid=4297)\u001b[0m 2021-09-14 15:37:12,366 cisTopic     INFO     Counting fragments\n",
      "\u001b[2m\u001b[36m(pid=4300)\u001b[0m 2021-09-14 15:37:13,056 cisTopic     INFO     Computing barcode rank plot for Broad_mito_1\n",
      "\u001b[2m\u001b[36m(pid=4300)\u001b[0m 2021-09-14 15:37:13,056 cisTopic     INFO     Counting fragments\n",
      "\u001b[2m\u001b[36m(pid=4297)\u001b[0m 2021-09-14 15:37:13,040 cisTopic     INFO     Marking barcodes with more than 100\n",
      "\u001b[2m\u001b[36m(pid=4297)\u001b[0m 2021-09-14 15:37:13,045 numexpr.utils INFO     Note: NumExpr detected 36 cores but \"NUMEXPR_MAX_THREADS\" not set, so enforcing safe limit of 8.\n",
      "\u001b[2m\u001b[36m(pid=4297)\u001b[0m 2021-09-14 15:37:13,045 cisTopic     INFO     Returning plot data\n",
      "\u001b[2m\u001b[36m(pid=4297)\u001b[0m 2021-09-14 15:37:13,046 cisTopic     INFO     Returning valid barcodes\n",
      "\u001b[2m\u001b[36m(pid=4300)\u001b[0m 2021-09-14 15:37:13,769 cisTopic     INFO     Marking barcodes with more than 100\n",
      "\u001b[2m\u001b[36m(pid=4300)\u001b[0m 2021-09-14 15:37:13,773 numexpr.utils INFO     Note: NumExpr detected 36 cores but \"NUMEXPR_MAX_THREADS\" not set, so enforcing safe limit of 8.\n",
      "\u001b[2m\u001b[36m(pid=4300)\u001b[0m 2021-09-14 15:37:13,774 cisTopic     INFO     Returning plot data\n",
      "\u001b[2m\u001b[36m(pid=4300)\u001b[0m 2021-09-14 15:37:13,774 cisTopic     INFO     Returning valid barcodes\n",
      "\u001b[2m\u001b[36m(pid=4297)\u001b[0m 2021-09-14 15:37:14,001 cisTopic     INFO     Computing duplicate rate plot for Broad_mito_2\n",
      "\u001b[2m\u001b[36m(pid=4300)\u001b[0m 2021-09-14 15:37:14,795 cisTopic     INFO     Computing duplicate rate plot for Broad_mito_1\n",
      "\u001b[2m\u001b[36m(pid=4297)\u001b[0m 2021-09-14 15:37:14,847 cisTopic     INFO     Return plot data\n",
      "\u001b[2m\u001b[36m(pid=4297)\u001b[0m 2021-09-14 15:37:14,921 cisTopic     INFO     Computing insert size distribution for Broad_mito_2\n",
      "\u001b[2m\u001b[36m(pid=4297)\u001b[0m 2021-09-14 15:37:14,921 cisTopic     INFO     Counting fragments\n",
      "\u001b[2m\u001b[36m(pid=4297)\u001b[0m 2021-09-14 15:37:15,344 cisTopic     INFO     Returning plot data\n",
      "\u001b[2m\u001b[36m(pid=4300)\u001b[0m 2021-09-14 15:37:15,700 cisTopic     INFO     Return plot data\n",
      "\u001b[2m\u001b[36m(pid=4300)\u001b[0m 2021-09-14 15:37:15,781 cisTopic     INFO     Computing insert size distribution for Broad_mito_1\n",
      "\u001b[2m\u001b[36m(pid=4300)\u001b[0m 2021-09-14 15:37:15,781 cisTopic     INFO     Counting fragments\n",
      "\u001b[2m\u001b[36m(pid=4300)\u001b[0m 2021-09-14 15:37:16,228 cisTopic     INFO     Returning plot data\n",
      "\u001b[2m\u001b[36m(pid=4299)\u001b[0m 2021-09-14 15:37:16,790 cisTopic     INFO     Computing TSS profile for CNAG_1\n",
      "\u001b[2m\u001b[36m(pid=4299)\u001b[0m 2021-09-14 15:37:17,532 cisTopic     INFO     Formatting annnotation\n",
      "\u001b[2m\u001b[36m(pid=4299)\u001b[0m 2021-09-14 15:37:17,618 cisTopic     INFO     Creating coverage matrix\n",
      "\u001b[2m\u001b[36m(pid=4297)\u001b[0m 2021-09-14 15:37:20,625 cisTopic     INFO     Computing TSS profile for Broad_mito_2\n",
      "\u001b[2m\u001b[36m(pid=4297)\u001b[0m 2021-09-14 15:37:21,478 cisTopic     INFO     Formatting annnotation\n",
      "\u001b[2m\u001b[36m(pid=4297)\u001b[0m 2021-09-14 15:37:21,562 cisTopic     INFO     Creating coverage matrix\n",
      "\u001b[2m\u001b[36m(pid=4300)\u001b[0m 2021-09-14 15:37:21,774 cisTopic     INFO     Computing TSS profile for Broad_mito_1\n",
      "\u001b[2m\u001b[36m(pid=4300)\u001b[0m 2021-09-14 15:37:22,646 cisTopic     INFO     Formatting annnotation\n",
      "\u001b[2m\u001b[36m(pid=4300)\u001b[0m 2021-09-14 15:37:22,722 cisTopic     INFO     Creating coverage matrix\n",
      "\u001b[2m\u001b[36m(pid=4298)\u001b[0m 2021-09-14 15:37:24,197 cisTopic     INFO     Computing barcode rank plot for Broad_2\n",
      "\u001b[2m\u001b[36m(pid=4298)\u001b[0m 2021-09-14 15:37:24,197 cisTopic     INFO     Counting fragments\n",
      "\u001b[2m\u001b[36m(pid=4298)\u001b[0m 2021-09-14 15:37:25,048 cisTopic     INFO     Marking barcodes with more than 100\n",
      "\u001b[2m\u001b[36m(pid=4298)\u001b[0m 2021-09-14 15:37:25,058 numexpr.utils INFO     Note: NumExpr detected 36 cores but \"NUMEXPR_MAX_THREADS\" not set, so enforcing safe limit of 8.\n",
      "\u001b[2m\u001b[36m(pid=4298)\u001b[0m 2021-09-14 15:37:25,060 cisTopic     INFO     Returning plot data\n",
      "\u001b[2m\u001b[36m(pid=4298)\u001b[0m 2021-09-14 15:37:25,060 cisTopic     INFO     Returning valid barcodes\n",
      "\u001b[2m\u001b[36m(pid=4298)\u001b[0m 2021-09-14 15:37:26,354 cisTopic     INFO     Computing duplicate rate plot for Broad_2\n",
      "\u001b[2m\u001b[36m(pid=4301)\u001b[0m 2021-09-14 15:37:26,476 cisTopic     INFO     Computing barcode rank plot for Broad_1\n",
      "\u001b[2m\u001b[36m(pid=4301)\u001b[0m 2021-09-14 15:37:26,476 cisTopic     INFO     Counting fragments\n",
      "\u001b[2m\u001b[36m(pid=4301)\u001b[0m 2021-09-14 15:37:27,353 cisTopic     INFO     Marking barcodes with more than 100\n",
      "\u001b[2m\u001b[36m(pid=4301)\u001b[0m 2021-09-14 15:37:27,357 numexpr.utils INFO     Note: NumExpr detected 36 cores but \"NUMEXPR_MAX_THREADS\" not set, so enforcing safe limit of 8.\n",
      "\u001b[2m\u001b[36m(pid=4301)\u001b[0m 2021-09-14 15:37:27,358 cisTopic     INFO     Returning plot data\n",
      "\u001b[2m\u001b[36m(pid=4301)\u001b[0m 2021-09-14 15:37:27,359 cisTopic     INFO     Returning valid barcodes\n",
      "\u001b[2m\u001b[36m(pid=4298)\u001b[0m 2021-09-14 15:37:27,664 cisTopic     INFO     Return plot data\n",
      "\u001b[2m\u001b[36m(pid=4298)\u001b[0m 2021-09-14 15:37:27,780 cisTopic     INFO     Computing insert size distribution for Broad_2\n",
      "\u001b[2m\u001b[36m(pid=4298)\u001b[0m 2021-09-14 15:37:27,780 cisTopic     INFO     Counting fragments\n",
      "\u001b[2m\u001b[36m(pid=4298)\u001b[0m 2021-09-14 15:37:28,401 cisTopic     INFO     Returning plot data\n",
      "\u001b[2m\u001b[36m(pid=4301)\u001b[0m 2021-09-14 15:37:28,605 cisTopic     INFO     Computing duplicate rate plot for Broad_1\n",
      "\u001b[2m\u001b[36m(pid=4301)\u001b[0m 2021-09-14 15:37:29,922 cisTopic     INFO     Return plot data\n",
      "\u001b[2m\u001b[36m(pid=4301)\u001b[0m 2021-09-14 15:37:30,031 cisTopic     INFO     Computing insert size distribution for Broad_1\n",
      "\u001b[2m\u001b[36m(pid=4301)\u001b[0m 2021-09-14 15:37:30,031 cisTopic     INFO     Counting fragments\n",
      "\u001b[2m\u001b[36m(pid=4301)\u001b[0m 2021-09-14 15:37:30,607 cisTopic     INFO     Returning plot data\n",
      "\u001b[2m\u001b[36m(pid=4298)\u001b[0m 2021-09-14 15:37:35,334 cisTopic     INFO     Computing TSS profile for Broad_2\n",
      "\u001b[2m\u001b[36m(pid=4298)\u001b[0m 2021-09-14 15:37:36,482 cisTopic     INFO     Formatting annnotation\n",
      "\u001b[2m\u001b[36m(pid=4298)\u001b[0m 2021-09-14 15:37:36,565 cisTopic     INFO     Creating coverage matrix\n",
      "\u001b[2m\u001b[36m(pid=4301)\u001b[0m 2021-09-14 15:37:37,396 cisTopic     INFO     Computing TSS profile for Broad_1\n",
      "\u001b[2m\u001b[36m(pid=4301)\u001b[0m 2021-09-14 15:37:38,624 cisTopic     INFO     Formatting annnotation\n",
      "\u001b[2m\u001b[36m(pid=4301)\u001b[0m 2021-09-14 15:37:38,703 cisTopic     INFO     Creating coverage matrix\n",
      "\u001b[2m\u001b[36m(pid=4299)\u001b[0m 2021-09-14 15:37:42,756 cisTopic     INFO     Coverage matrix done\n",
      "\u001b[2m\u001b[36m(pid=4299)\u001b[0m 2021-09-14 15:37:44,421 cisTopic     INFO     Returning normalized TSS coverage matrix per barcode\n",
      "\u001b[2m\u001b[36m(pid=4299)\u001b[0m 2021-09-14 15:37:45,755 cisTopic     INFO     Returning normalized sample TSS enrichment data\n",
      "\u001b[2m\u001b[36m(pid=4299)\u001b[0m 2021-09-14 15:37:45,949 cisTopic     INFO     Computing FRIP profile for CNAG_1\n",
      "\u001b[2m\u001b[36m(pid=4299)\u001b[0m 2021-09-14 15:37:46,404 cisTopic     INFO     Counting fragments\n",
      "\u001b[2m\u001b[36m(pid=4299)\u001b[0m 2021-09-14 15:37:47,460 cisTopic     INFO     Intersecting fragments with regions\n",
      "\u001b[2m\u001b[36m(pid=4297)\u001b[0m 2021-09-14 15:37:49,966 cisTopic     INFO     Coverage matrix done\n",
      "\u001b[2m\u001b[36m(pid=4299)\u001b[0m 2021-09-14 15:37:51,856 cisTopic     INFO     Sample CNAG_1 done!\n",
      "\u001b[2m\u001b[36m(pid=4299)\u001b[0m 2021-09-14 15:37:51,924 cisTopic     INFO     Reading CNAG_2\n",
      "\u001b[2m\u001b[36m(pid=4297)\u001b[0m 2021-09-14 15:37:52,115 cisTopic     INFO     Returning normalized TSS coverage matrix per barcode\n",
      "\u001b[2m\u001b[36m(pid=4297)\u001b[0m 2021-09-14 15:37:53,671 cisTopic     INFO     Returning normalized sample TSS enrichment data\n",
      "\u001b[2m\u001b[36m(pid=4297)\u001b[0m 2021-09-14 15:37:53,754 cisTopic     INFO     Computing FRIP profile for Broad_mito_2\n",
      "\u001b[2m\u001b[36m(pid=4300)\u001b[0m 2021-09-14 15:37:54,035 cisTopic     INFO     Coverage matrix done\n",
      "\u001b[2m\u001b[36m(pid=4297)\u001b[0m 2021-09-14 15:37:54,234 cisTopic     INFO     Counting fragments\n",
      "\u001b[2m\u001b[36m(pid=4297)\u001b[0m 2021-09-14 15:37:55,506 cisTopic     INFO     Intersecting fragments with regions\n",
      "\u001b[2m\u001b[36m(pid=4300)\u001b[0m 2021-09-14 15:37:56,268 cisTopic     INFO     Returning normalized TSS coverage matrix per barcode\n",
      "\u001b[2m\u001b[36m(pid=4300)\u001b[0m 2021-09-14 15:37:57,858 cisTopic     INFO     Returning normalized sample TSS enrichment data\n",
      "\u001b[2m\u001b[36m(pid=4300)\u001b[0m 2021-09-14 15:37:57,952 cisTopic     INFO     Computing FRIP profile for Broad_mito_1\n",
      "\u001b[2m\u001b[36m(pid=4300)\u001b[0m 2021-09-14 15:37:58,449 cisTopic     INFO     Counting fragments\n",
      "\u001b[2m\u001b[36m(pid=4300)\u001b[0m 2021-09-14 15:37:59,788 cisTopic     INFO     Intersecting fragments with regions\n",
      "\u001b[2m\u001b[36m(pid=4297)\u001b[0m 2021-09-14 15:38:00,388 cisTopic     INFO     Sample Broad_mito_2 done!\n",
      "\u001b[2m\u001b[36m(pid=4297)\u001b[0m 2021-09-14 15:38:00,446 cisTopic     INFO     Reading s3atac\n",
      "\u001b[2m\u001b[36m(pid=4299)\u001b[0m 2021-09-14 15:38:05,134 cisTopic     INFO     Computing barcode rank plot for CNAG_2\n",
      "\u001b[2m\u001b[36m(pid=4299)\u001b[0m 2021-09-14 15:38:05,135 cisTopic     INFO     Counting fragments\n",
      "\u001b[2m\u001b[36m(pid=4300)\u001b[0m 2021-09-14 15:38:05,180 cisTopic     INFO     Sample Broad_mito_1 done!\n",
      "\u001b[2m\u001b[36m(pid=4300)\u001b[0m 2021-09-14 15:38:05,242 cisTopic     INFO     Reading Sanger_1\n",
      "\u001b[2m\u001b[36m(pid=4299)\u001b[0m 2021-09-14 15:38:05,714 cisTopic     INFO     Marking barcodes with more than 100\n",
      "\u001b[2m\u001b[36m(pid=4299)\u001b[0m 2021-09-14 15:38:05,715 cisTopic     INFO     Returning plot data\n",
      "\u001b[2m\u001b[36m(pid=4299)\u001b[0m 2021-09-14 15:38:05,715 cisTopic     INFO     Returning valid barcodes\n",
      "\u001b[2m\u001b[36m(pid=4299)\u001b[0m 2021-09-14 15:38:06,587 cisTopic     INFO     Computing duplicate rate plot for CNAG_2\n",
      "\u001b[2m\u001b[36m(pid=4299)\u001b[0m 2021-09-14 15:38:07,462 cisTopic     INFO     Return plot data\n",
      "\u001b[2m\u001b[36m(pid=4299)\u001b[0m 2021-09-14 15:38:07,539 cisTopic     INFO     Computing insert size distribution for CNAG_2\n",
      "\u001b[2m\u001b[36m(pid=4299)\u001b[0m 2021-09-14 15:38:07,539 cisTopic     INFO     Counting fragments\n",
      "\u001b[2m\u001b[36m(pid=4299)\u001b[0m 2021-09-14 15:38:07,904 cisTopic     INFO     Returning plot data\n",
      "\u001b[2m\u001b[36m(pid=4299)\u001b[0m 2021-09-14 15:38:12,434 cisTopic     INFO     Computing TSS profile for CNAG_2\n",
      "\u001b[2m\u001b[36m(pid=4299)\u001b[0m 2021-09-14 15:38:13,119 cisTopic     INFO     Formatting annnotation\n",
      "\u001b[2m\u001b[36m(pid=4299)\u001b[0m 2021-09-14 15:38:13,194 cisTopic     INFO     Creating coverage matrix\n",
      "\u001b[2m\u001b[36m(pid=4297)\u001b[0m 2021-09-14 15:38:20,645 cisTopic     INFO     Computing barcode rank plot for s3atac\n",
      "\u001b[2m\u001b[36m(pid=4297)\u001b[0m 2021-09-14 15:38:20,645 cisTopic     INFO     Counting fragments\n",
      "\u001b[2m\u001b[36m(pid=4297)\u001b[0m 2021-09-14 15:38:21,470 cisTopic     INFO     Marking barcodes with more than 100\n",
      "\u001b[2m\u001b[36m(pid=4297)\u001b[0m 2021-09-14 15:38:21,471 cisTopic     INFO     Returning plot data\n",
      "\u001b[2m\u001b[36m(pid=4297)\u001b[0m 2021-09-14 15:38:21,471 cisTopic     INFO     Returning valid barcodes\n",
      "\u001b[2m\u001b[36m(pid=4297)\u001b[0m 2021-09-14 15:38:22,639 cisTopic     INFO     Computing duplicate rate plot for s3atac\n",
      "\u001b[2m\u001b[36m(pid=4297)\u001b[0m 2021-09-14 15:38:23,676 cisTopic     INFO     Return plot data\n",
      "\u001b[2m\u001b[36m(pid=4297)\u001b[0m 2021-09-14 15:38:23,754 cisTopic     INFO     Computing insert size distribution for s3atac\n",
      "\u001b[2m\u001b[36m(pid=4297)\u001b[0m 2021-09-14 15:38:23,754 cisTopic     INFO     Counting fragments\n",
      "\u001b[2m\u001b[36m(pid=4297)\u001b[0m 2021-09-14 15:38:24,219 cisTopic     INFO     Returning plot data\n",
      "\u001b[2m\u001b[36m(pid=4298)\u001b[0m 2021-09-14 15:38:24,436 cisTopic     INFO     Coverage matrix done\n",
      "\u001b[2m\u001b[36m(pid=4300)\u001b[0m 2021-09-14 15:38:25,575 cisTopic     INFO     Computing barcode rank plot for Sanger_1\n",
      "\u001b[2m\u001b[36m(pid=4300)\u001b[0m 2021-09-14 15:38:25,576 cisTopic     INFO     Counting fragments\n",
      "\u001b[2m\u001b[36m(pid=4300)\u001b[0m 2021-09-14 15:38:26,402 cisTopic     INFO     Marking barcodes with more than 100\n",
      "\u001b[2m\u001b[36m(pid=4300)\u001b[0m 2021-09-14 15:38:26,403 cisTopic     INFO     Returning plot data\n",
      "\u001b[2m\u001b[36m(pid=4300)\u001b[0m 2021-09-14 15:38:26,403 cisTopic     INFO     Returning valid barcodes\n",
      "\u001b[2m\u001b[36m(pid=4298)\u001b[0m 2021-09-14 15:38:27,167 cisTopic     INFO     Returning normalized TSS coverage matrix per barcode\n",
      "\u001b[2m\u001b[36m(pid=4300)\u001b[0m 2021-09-14 15:38:27,718 cisTopic     INFO     Computing duplicate rate plot for Sanger_1\n",
      "\u001b[2m\u001b[36m(pid=4301)\u001b[0m 2021-09-14 15:38:28,062 cisTopic     INFO     Coverage matrix done\n",
      "\u001b[2m\u001b[36m(pid=4300)\u001b[0m 2021-09-14 15:38:28,845 cisTopic     INFO     Return plot data\n",
      "\u001b[2m\u001b[36m(pid=4300)\u001b[0m 2021-09-14 15:38:28,928 cisTopic     INFO     Computing insert size distribution for Sanger_1\n",
      "\u001b[2m\u001b[36m(pid=4300)\u001b[0m 2021-09-14 15:38:28,928 cisTopic     INFO     Counting fragments\n",
      "\u001b[2m\u001b[36m(pid=4298)\u001b[0m 2021-09-14 15:38:29,109 cisTopic     INFO     Returning normalized sample TSS enrichment data\n",
      "\u001b[2m\u001b[36m(pid=4298)\u001b[0m 2021-09-14 15:38:29,192 cisTopic     INFO     Computing FRIP profile for Broad_2\n",
      "\u001b[2m\u001b[36m(pid=4300)\u001b[0m 2021-09-14 15:38:29,452 cisTopic     INFO     Returning plot data\n",
      "\u001b[2m\u001b[36m(pid=4298)\u001b[0m 2021-09-14 15:38:29,620 cisTopic     INFO     Counting fragments\n",
      "\u001b[2m\u001b[36m(pid=4297)\u001b[0m 2021-09-14 15:38:30,533 cisTopic     INFO     Computing TSS profile for s3atac\n",
      "\u001b[2m\u001b[36m(pid=4301)\u001b[0m 2021-09-14 15:38:30,740 cisTopic     INFO     Returning normalized TSS coverage matrix per barcode\n",
      "\u001b[2m\u001b[36m(pid=4298)\u001b[0m 2021-09-14 15:38:31,229 cisTopic     INFO     Intersecting fragments with regions\n",
      "\u001b[2m\u001b[36m(pid=4297)\u001b[0m 2021-09-14 15:38:31,510 cisTopic     INFO     Formatting annnotation\n",
      "\u001b[2m\u001b[36m(pid=4297)\u001b[0m 2021-09-14 15:38:31,585 cisTopic     INFO     Creating coverage matrix\n",
      "\u001b[2m\u001b[36m(pid=4301)\u001b[0m 2021-09-14 15:38:32,700 cisTopic     INFO     Returning normalized sample TSS enrichment data\n",
      "\u001b[2m\u001b[36m(pid=4301)\u001b[0m 2021-09-14 15:38:32,788 cisTopic     INFO     Computing FRIP profile for Broad_1\n",
      "\u001b[2m\u001b[36m(pid=4301)\u001b[0m 2021-09-14 15:38:33,226 cisTopic     INFO     Counting fragments\n",
      "\u001b[2m\u001b[36m(pid=4301)\u001b[0m 2021-09-14 15:38:34,918 cisTopic     INFO     Intersecting fragments with regions\n",
      "\u001b[2m\u001b[36m(pid=4300)\u001b[0m 2021-09-14 15:38:36,673 cisTopic     INFO     Computing TSS profile for Sanger_1\n",
      "\u001b[2m\u001b[36m(pid=4298)\u001b[0m 2021-09-14 15:38:37,661 cisTopic     INFO     Sample Broad_2 done!\n",
      "\u001b[2m\u001b[36m(pid=4298)\u001b[0m 2021-09-14 15:38:37,724 cisTopic     INFO     Reading Sanger_2\n",
      "\u001b[2m\u001b[36m(pid=4300)\u001b[0m 2021-09-14 15:38:37,735 cisTopic     INFO     Formatting annnotation\n",
      "\u001b[2m\u001b[36m(pid=4300)\u001b[0m 2021-09-14 15:38:37,817 cisTopic     INFO     Creating coverage matrix\n",
      "\u001b[2m\u001b[36m(pid=4299)\u001b[0m 2021-09-14 15:38:39,252 cisTopic     INFO     Coverage matrix done\n",
      "\u001b[2m\u001b[36m(pid=4299)\u001b[0m 2021-09-14 15:38:40,936 cisTopic     INFO     Returning normalized TSS coverage matrix per barcode\n",
      "\u001b[2m\u001b[36m(pid=4301)\u001b[0m 2021-09-14 15:38:41,643 cisTopic     INFO     Sample Broad_1 done!\n",
      "\u001b[2m\u001b[36m(pid=4301)\u001b[0m 2021-09-14 15:38:41,709 cisTopic     INFO     Reading Stanford_1\n",
      "\u001b[2m\u001b[36m(pid=4299)\u001b[0m 2021-09-14 15:38:42,087 cisTopic     INFO     Returning normalized sample TSS enrichment data\n",
      "\u001b[2m\u001b[36m(pid=4299)\u001b[0m 2021-09-14 15:38:42,173 cisTopic     INFO     Computing FRIP profile for CNAG_2\n",
      "\u001b[2m\u001b[36m(pid=4299)\u001b[0m 2021-09-14 15:38:42,659 cisTopic     INFO     Counting fragments\n",
      "\u001b[2m\u001b[36m(pid=4299)\u001b[0m 2021-09-14 15:38:43,790 cisTopic     INFO     Intersecting fragments with regions\n",
      "\u001b[2m\u001b[36m(pid=4301)\u001b[0m 2021-09-14 15:38:47,067 cisTopic     INFO     Computing barcode rank plot for Stanford_1\n",
      "\u001b[2m\u001b[36m(pid=4301)\u001b[0m 2021-09-14 15:38:47,067 cisTopic     INFO     Counting fragments\n",
      "\u001b[2m\u001b[36m(pid=4301)\u001b[0m 2021-09-14 15:38:47,257 cisTopic     INFO     Marking barcodes with more than 100\n",
      "\u001b[2m\u001b[36m(pid=4301)\u001b[0m 2021-09-14 15:38:47,257 cisTopic     INFO     Returning plot data\n",
      "\u001b[2m\u001b[36m(pid=4301)\u001b[0m 2021-09-14 15:38:47,257 cisTopic     INFO     Returning valid barcodes\n",
      "\u001b[2m\u001b[36m(pid=4301)\u001b[0m 2021-09-14 15:38:47,580 cisTopic     INFO     Computing duplicate rate plot for Stanford_1\n",
      "\u001b[2m\u001b[36m(pid=4301)\u001b[0m 2021-09-14 15:38:47,864 cisTopic     INFO     Return plot data\n",
      "\u001b[2m\u001b[36m(pid=4301)\u001b[0m 2021-09-14 15:38:47,940 cisTopic     INFO     Computing insert size distribution for Stanford_1\n",
      "\u001b[2m\u001b[36m(pid=4301)\u001b[0m 2021-09-14 15:38:47,940 cisTopic     INFO     Counting fragments\n",
      "\u001b[2m\u001b[36m(pid=4301)\u001b[0m 2021-09-14 15:38:48,056 cisTopic     INFO     Returning plot data\n",
      "\u001b[2m\u001b[36m(pid=4299)\u001b[0m 2021-09-14 15:38:48,304 cisTopic     INFO     Sample CNAG_2 done!\n",
      "\u001b[2m\u001b[36m(pid=4299)\u001b[0m 2021-09-14 15:38:48,353 cisTopic     INFO     Reading Stanford_2\n",
      "\u001b[2m\u001b[36m(pid=4297)\u001b[0m 2021-09-14 15:38:49,119 cisTopic     INFO     Coverage matrix done\n",
      "\u001b[2m\u001b[36m(pid=4301)\u001b[0m 2021-09-14 15:38:49,736 cisTopic     INFO     Computing TSS profile for Stanford_1\n",
      "\u001b[2m\u001b[36m(pid=4301)\u001b[0m 2021-09-14 15:38:49,980 cisTopic     INFO     Formatting annnotation\n",
      "\u001b[2m\u001b[36m(pid=4301)\u001b[0m 2021-09-14 15:38:50,051 cisTopic     INFO     Creating coverage matrix\n",
      "\u001b[2m\u001b[36m(pid=4297)\u001b[0m 2021-09-14 15:38:51,128 cisTopic     INFO     Returning normalized TSS coverage matrix per barcode\n",
      "\u001b[2m\u001b[36m(pid=4297)\u001b[0m 2021-09-14 15:38:52,593 cisTopic     INFO     Returning normalized sample TSS enrichment data\n",
      "\u001b[2m\u001b[36m(pid=4297)\u001b[0m 2021-09-14 15:38:52,674 cisTopic     INFO     Computing FRIP profile for s3atac\n",
      "\u001b[2m\u001b[36m(pid=4297)\u001b[0m 2021-09-14 15:38:53,035 cisTopic     INFO     Counting fragments\n",
      "\u001b[2m\u001b[36m(pid=4297)\u001b[0m 2021-09-14 15:38:54,526 cisTopic     INFO     Intersecting fragments with regions\n",
      "\u001b[2m\u001b[36m(pid=4299)\u001b[0m 2021-09-14 15:38:56,267 cisTopic     INFO     Computing barcode rank plot for Stanford_2\n",
      "\u001b[2m\u001b[36m(pid=4299)\u001b[0m 2021-09-14 15:38:56,267 cisTopic     INFO     Counting fragments\n",
      "\u001b[2m\u001b[36m(pid=4299)\u001b[0m 2021-09-14 15:38:56,556 cisTopic     INFO     Marking barcodes with more than 100\n",
      "\u001b[2m\u001b[36m(pid=4299)\u001b[0m 2021-09-14 15:38:56,557 cisTopic     INFO     Returning plot data\n",
      "\u001b[2m\u001b[36m(pid=4299)\u001b[0m 2021-09-14 15:38:56,558 cisTopic     INFO     Returning valid barcodes\n",
      "\u001b[2m\u001b[36m(pid=4301)\u001b[0m 2021-09-14 15:38:57,082 cisTopic     INFO     Coverage matrix done\n",
      "\u001b[2m\u001b[36m(pid=4299)\u001b[0m 2021-09-14 15:38:57,119 cisTopic     INFO     Computing duplicate rate plot for Stanford_2\n",
      "\u001b[2m\u001b[36m(pid=4297)\u001b[0m 2021-09-14 15:38:57,092 cisTopic     INFO     Sample s3atac done!\n",
      "\u001b[2m\u001b[36m(pid=4297)\u001b[0m 2021-09-14 15:38:57,145 cisTopic     INFO     Reading VIB_1\n",
      "\u001b[2m\u001b[36m(pid=4301)\u001b[0m 2021-09-14 15:38:57,598 cisTopic     INFO     Returning normalized TSS coverage matrix per barcode\n",
      "\u001b[2m\u001b[36m(pid=4299)\u001b[0m 2021-09-14 15:38:57,567 cisTopic     INFO     Return plot data\n",
      "\u001b[2m\u001b[36m(pid=4299)\u001b[0m 2021-09-14 15:38:57,643 cisTopic     INFO     Computing insert size distribution for Stanford_2\n",
      "\u001b[2m\u001b[36m(pid=4299)\u001b[0m 2021-09-14 15:38:57,643 cisTopic     INFO     Counting fragments\n",
      "\u001b[2m\u001b[36m(pid=4299)\u001b[0m 2021-09-14 15:38:57,821 cisTopic     INFO     Returning plot data\n",
      "\u001b[2m\u001b[36m(pid=4301)\u001b[0m 2021-09-14 15:38:57,937 cisTopic     INFO     Returning normalized sample TSS enrichment data\n",
      "\u001b[2m\u001b[36m(pid=4301)\u001b[0m 2021-09-14 15:38:58,013 cisTopic     INFO     Computing FRIP profile for Stanford_1\n",
      "\u001b[2m\u001b[36m(pid=4301)\u001b[0m 2021-09-14 15:38:58,341 cisTopic     INFO     Counting fragments\n",
      "\u001b[2m\u001b[36m(pid=4301)\u001b[0m 2021-09-14 15:38:58,682 cisTopic     INFO     Intersecting fragments with regions\n",
      "\u001b[2m\u001b[36m(pid=4301)\u001b[0m 2021-09-14 15:39:00,099 cisTopic     INFO     Sample Stanford_1 done!\n",
      "\u001b[2m\u001b[36m(pid=4301)\u001b[0m 2021-09-14 15:39:00,127 cisTopic     INFO     Reading VIB_2\n",
      "\u001b[2m\u001b[36m(pid=4299)\u001b[0m 2021-09-14 15:39:00,597 cisTopic     INFO     Computing TSS profile for Stanford_2\n",
      "\u001b[2m\u001b[36m(pid=4299)\u001b[0m 2021-09-14 15:39:00,973 cisTopic     INFO     Formatting annnotation\n",
      "\u001b[2m\u001b[36m(pid=4299)\u001b[0m 2021-09-14 15:39:01,048 cisTopic     INFO     Creating coverage matrix\n",
      "\u001b[2m\u001b[36m(pid=4298)\u001b[0m 2021-09-14 15:39:02,273 cisTopic     INFO     Computing barcode rank plot for Sanger_2\n",
      "\u001b[2m\u001b[36m(pid=4298)\u001b[0m 2021-09-14 15:39:02,274 cisTopic     INFO     Counting fragments\n",
      "\u001b[2m\u001b[36m(pid=4298)\u001b[0m 2021-09-14 15:39:03,253 cisTopic     INFO     Marking barcodes with more than 100\n",
      "\u001b[2m\u001b[36m(pid=4298)\u001b[0m 2021-09-14 15:39:03,254 cisTopic     INFO     Returning plot data\n",
      "\u001b[2m\u001b[36m(pid=4298)\u001b[0m 2021-09-14 15:39:03,254 cisTopic     INFO     Returning valid barcodes\n",
      "\u001b[2m\u001b[36m(pid=4298)\u001b[0m 2021-09-14 15:39:04,742 cisTopic     INFO     Computing duplicate rate plot for Sanger_2\n",
      "\u001b[2m\u001b[36m(pid=4298)\u001b[0m 2021-09-14 15:39:06,039 cisTopic     INFO     Return plot data\n",
      "\u001b[2m\u001b[36m(pid=4298)\u001b[0m 2021-09-14 15:39:06,153 cisTopic     INFO     Computing insert size distribution for Sanger_2\n",
      "\u001b[2m\u001b[36m(pid=4298)\u001b[0m 2021-09-14 15:39:06,153 cisTopic     INFO     Counting fragments\n",
      "\u001b[2m\u001b[36m(pid=4298)\u001b[0m 2021-09-14 15:39:06,779 cisTopic     INFO     Returning plot data\n",
      "\u001b[2m\u001b[36m(pid=4297)\u001b[0m 2021-09-14 15:39:12,853 cisTopic     INFO     Computing barcode rank plot for VIB_1\n",
      "\u001b[2m\u001b[36m(pid=4297)\u001b[0m 2021-09-14 15:39:12,854 cisTopic     INFO     Counting fragments\n",
      "\u001b[2m\u001b[36m(pid=4297)\u001b[0m 2021-09-14 15:39:13,472 cisTopic     INFO     Marking barcodes with more than 100\n",
      "\u001b[2m\u001b[36m(pid=4297)\u001b[0m 2021-09-14 15:39:13,473 cisTopic     INFO     Returning plot data\n",
      "\u001b[2m\u001b[36m(pid=4297)\u001b[0m 2021-09-14 15:39:13,474 cisTopic     INFO     Returning valid barcodes\n",
      "\u001b[2m\u001b[36m(pid=4299)\u001b[0m 2021-09-14 15:39:13,755 cisTopic     INFO     Coverage matrix done\n",
      "\u001b[2m\u001b[36m(pid=4297)\u001b[0m 2021-09-14 15:39:14,413 cisTopic     INFO     Computing duplicate rate plot for VIB_1\n",
      "\u001b[2m\u001b[36m(pid=4299)\u001b[0m 2021-09-14 15:39:14,680 cisTopic     INFO     Returning normalized TSS coverage matrix per barcode\n",
      "\u001b[2m\u001b[36m(pid=4299)\u001b[0m 2021-09-14 15:39:15,283 cisTopic     INFO     Returning normalized sample TSS enrichment data\n",
      "\u001b[2m\u001b[36m(pid=4297)\u001b[0m 2021-09-14 15:39:15,272 cisTopic     INFO     Return plot data\n",
      "\u001b[2m\u001b[36m(pid=4297)\u001b[0m 2021-09-14 15:39:15,346 cisTopic     INFO     Computing insert size distribution for VIB_1\n",
      "\u001b[2m\u001b[36m(pid=4297)\u001b[0m 2021-09-14 15:39:15,346 cisTopic     INFO     Counting fragments\n",
      "\u001b[2m\u001b[36m(pid=4299)\u001b[0m 2021-09-14 15:39:15,402 cisTopic     INFO     Computing FRIP profile for Stanford_2\n",
      "\u001b[2m\u001b[36m(pid=4297)\u001b[0m 2021-09-14 15:39:15,754 cisTopic     INFO     Returning plot data\n",
      "\u001b[2m\u001b[36m(pid=4299)\u001b[0m 2021-09-14 15:39:15,859 cisTopic     INFO     Counting fragments\n",
      "\u001b[2m\u001b[36m(pid=4298)\u001b[0m 2021-09-14 15:39:15,935 cisTopic     INFO     Computing TSS profile for Sanger_2\n",
      "\u001b[2m\u001b[36m(pid=4299)\u001b[0m 2021-09-14 15:39:16,537 cisTopic     INFO     Intersecting fragments with regions\n",
      "\u001b[2m\u001b[36m(pid=4298)\u001b[0m 2021-09-14 15:39:17,222 cisTopic     INFO     Formatting annnotation\n",
      "\u001b[2m\u001b[36m(pid=4298)\u001b[0m 2021-09-14 15:39:17,309 cisTopic     INFO     Creating coverage matrix\n",
      "\u001b[2m\u001b[36m(pid=4299)\u001b[0m 2021-09-14 15:39:19,077 cisTopic     INFO     Sample Stanford_2 done!\n",
      "\u001b[2m\u001b[36m(pid=4300)\u001b[0m 2021-09-14 15:39:20,308 cisTopic     INFO     Coverage matrix done\n",
      "\u001b[2m\u001b[36m(pid=4297)\u001b[0m 2021-09-14 15:39:21,345 cisTopic     INFO     Computing TSS profile for VIB_1\n",
      "\u001b[2m\u001b[36m(pid=4297)\u001b[0m 2021-09-14 15:39:22,130 cisTopic     INFO     Formatting annnotation\n",
      "\u001b[2m\u001b[36m(pid=4297)\u001b[0m 2021-09-14 15:39:22,203 cisTopic     INFO     Creating coverage matrix\n",
      "\u001b[2m\u001b[36m(pid=4300)\u001b[0m 2021-09-14 15:39:22,299 cisTopic     INFO     Returning normalized TSS coverage matrix per barcode\n",
      "\u001b[2m\u001b[36m(pid=4300)\u001b[0m 2021-09-14 15:39:23,797 cisTopic     INFO     Returning normalized sample TSS enrichment data\n",
      "\u001b[2m\u001b[36m(pid=4300)\u001b[0m 2021-09-14 15:39:23,880 cisTopic     INFO     Computing FRIP profile for Sanger_1\n",
      "\u001b[2m\u001b[36m(pid=4300)\u001b[0m 2021-09-14 15:39:24,441 cisTopic     INFO     Counting fragments\n",
      "\u001b[2m\u001b[36m(pid=4300)\u001b[0m 2021-09-14 15:39:26,107 cisTopic     INFO     Intersecting fragments with regions\n",
      "\u001b[2m\u001b[36m(pid=4300)\u001b[0m 2021-09-14 15:39:32,671 cisTopic     INFO     Sample Sanger_1 done!\n",
      "\u001b[2m\u001b[36m(pid=4301)\u001b[0m 2021-09-14 15:39:42,930 cisTopic     INFO     Computing barcode rank plot for VIB_2\n",
      "\u001b[2m\u001b[36m(pid=4301)\u001b[0m 2021-09-14 15:39:42,930 cisTopic     INFO     Counting fragments\n",
      "\u001b[2m\u001b[36m(pid=4301)\u001b[0m 2021-09-14 15:39:44,731 cisTopic     INFO     Marking barcodes with more than 100\n",
      "\u001b[2m\u001b[36m(pid=4301)\u001b[0m 2021-09-14 15:39:44,733 cisTopic     INFO     Returning plot data\n",
      "\u001b[2m\u001b[36m(pid=4301)\u001b[0m 2021-09-14 15:39:44,733 cisTopic     INFO     Returning valid barcodes\n",
      "\u001b[2m\u001b[36m(pid=4301)\u001b[0m 2021-09-14 15:39:47,304 cisTopic     INFO     Computing duplicate rate plot for VIB_2\n",
      "\u001b[2m\u001b[36m(pid=4301)\u001b[0m 2021-09-14 15:39:50,060 cisTopic     INFO     Return plot data\n",
      "\u001b[2m\u001b[36m(pid=4301)\u001b[0m 2021-09-14 15:39:50,143 cisTopic     INFO     Computing insert size distribution for VIB_2\n",
      "\u001b[2m\u001b[36m(pid=4301)\u001b[0m 2021-09-14 15:39:50,144 cisTopic     INFO     Counting fragments\n",
      "\u001b[2m\u001b[36m(pid=4301)\u001b[0m 2021-09-14 15:39:51,312 cisTopic     INFO     Returning plot data\n",
      "\u001b[2m\u001b[36m(pid=4297)\u001b[0m 2021-09-14 15:39:58,544 cisTopic     INFO     Coverage matrix done\n",
      "\u001b[2m\u001b[36m(pid=4297)\u001b[0m 2021-09-14 15:40:00,299 cisTopic     INFO     Returning normalized TSS coverage matrix per barcode\n",
      "\u001b[2m\u001b[36m(pid=4297)\u001b[0m 2021-09-14 15:40:01,497 cisTopic     INFO     Returning normalized sample TSS enrichment data\n",
      "\u001b[2m\u001b[36m(pid=4297)\u001b[0m 2021-09-14 15:40:01,573 cisTopic     INFO     Computing FRIP profile for VIB_1\n",
      "\u001b[2m\u001b[36m(pid=4297)\u001b[0m 2021-09-14 15:40:02,034 cisTopic     INFO     Counting fragments\n",
      "\u001b[2m\u001b[36m(pid=4297)\u001b[0m 2021-09-14 15:40:03,380 cisTopic     INFO     Intersecting fragments with regions\n",
      "\u001b[2m\u001b[36m(pid=4298)\u001b[0m 2021-09-14 15:40:06,090 cisTopic     INFO     Coverage matrix done\n",
      "\u001b[2m\u001b[36m(pid=4301)\u001b[0m 2021-09-14 15:40:07,144 cisTopic     INFO     Computing TSS profile for VIB_2\n",
      "\u001b[2m\u001b[36m(pid=4298)\u001b[0m 2021-09-14 15:40:08,565 cisTopic     INFO     Returning normalized TSS coverage matrix per barcode\n",
      "\u001b[2m\u001b[36m(pid=4297)\u001b[0m 2021-09-14 15:40:09,035 cisTopic     INFO     Sample VIB_1 done!\n",
      "\u001b[2m\u001b[36m(pid=4301)\u001b[0m 2021-09-14 15:40:09,350 cisTopic     INFO     Formatting annnotation\n",
      "\u001b[2m\u001b[36m(pid=4301)\u001b[0m 2021-09-14 15:40:09,431 cisTopic     INFO     Creating coverage matrix\n",
      "\u001b[2m\u001b[36m(pid=4298)\u001b[0m 2021-09-14 15:40:10,359 cisTopic     INFO     Returning normalized sample TSS enrichment data\n",
      "\u001b[2m\u001b[36m(pid=4298)\u001b[0m 2021-09-14 15:40:10,429 cisTopic     INFO     Computing FRIP profile for Sanger_2\n",
      "\u001b[2m\u001b[36m(pid=4298)\u001b[0m 2021-09-14 15:40:11,048 cisTopic     INFO     Counting fragments\n",
      "\u001b[2m\u001b[36m(pid=4298)\u001b[0m 2021-09-14 15:40:13,036 cisTopic     INFO     Intersecting fragments with regions\n",
      "\u001b[2m\u001b[36m(pid=4298)\u001b[0m 2021-09-14 15:40:20,516 cisTopic     INFO     Sample Sanger_2 done!\n",
      "\u001b[2m\u001b[36m(pid=4301)\u001b[0m 2021-09-14 15:41:20,785 cisTopic     INFO     Coverage matrix done\n",
      "\u001b[2m\u001b[36m(pid=4301)\u001b[0m 2021-09-14 15:41:24,768 cisTopic     INFO     Returning normalized TSS coverage matrix per barcode\n",
      "\u001b[2m\u001b[36m(pid=4301)\u001b[0m 2021-09-14 15:41:27,580 cisTopic     INFO     Returning normalized sample TSS enrichment data\n",
      "\u001b[2m\u001b[36m(pid=4301)\u001b[0m 2021-09-14 15:41:27,695 cisTopic     INFO     Computing FRIP profile for VIB_2\n",
      "\u001b[2m\u001b[36m(pid=4301)\u001b[0m 2021-09-14 15:41:28,212 cisTopic     INFO     Counting fragments\n",
      "\u001b[2m\u001b[36m(pid=4301)\u001b[0m 2021-09-14 15:41:31,534 cisTopic     INFO     Intersecting fragments with regions\n",
      "\u001b[2m\u001b[36m(pid=4301)\u001b[0m 2021-09-14 15:41:43,807 cisTopic     INFO     Sample VIB_2 done!\n"
     ]
    }
   ],
   "source": [
    "metadata_bc_dict, profile_data_dict = compute_qc_stats(\n",
    "        fragments_dict=fragments_dict,\n",
    "        tss_annotation=annot,\n",
    "        stats=['barcode_rank_plot', 'duplicate_rate', 'insert_size_distribution', 'profile_tss', 'frip'],\n",
    "        label_list=None,\n",
    "        path_to_regions=consensus_peaks_dict,\n",
    "        n_cpu=5,\n",
    "        valid_bc=None,\n",
    "        n_frag=100,\n",
    "        n_bc=None,\n",
    "        tss_flank_window=2000,\n",
    "        tss_window=50,\n",
    "        tss_minimum_signal_window=100,\n",
    "        tss_rolling_window=10,\n",
    "        min_norm=0.1,\n",
    "        remove_duplicates = True,\n",
    "        )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "57bfb631-a742-4ee7-b457-997533d14d33",
   "metadata": {},
   "outputs": [],
   "source": [
    "f_qc_dir = 'qc__consensus_peaks'\n",
    "if not os.path.exists(os.path.join(wdir, f_qc_dir)):\n",
    "    os.makedirs(os.path.join(wdir, f_qc_dir))\n",
    "    \n",
    "with open(os.path.join(wdir, f_qc_dir, 'metadata.pickle'), 'wb') as f:\n",
    "    pickle.dump(metadata_bc_dict, f)\n",
    "\n",
    "with open(os.path.join(wdir, f_qc_dir, 'profile_data.pickle'), 'wb') as f:\n",
    "    pickle.dump(profile_data_dict, f)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9e0672ba-f15a-4417-b04a-5bd49c95b1a6",
   "metadata": {},
   "source": [
    "## Create the cisTopic objects for each sample"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "81851c34-3012-4476-aad8-dd6a23d85418",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Valid barcodes\n",
    "with open(wdir + 'barcodes_passing_filters2.pkl', 'rb') as f:\n",
    "    bc_passing_filters = pickle.load(f)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "14d33283-bb31-410a-9644-1afefc9143ba",
   "metadata": {},
   "outputs": [],
   "source": [
    "from pycisTopic.cistopic_class import create_cistopic_object_from_fragments"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "ba842107-426d-4eae-b92d-86ff268fd541",
   "metadata": {},
   "outputs": [],
   "source": [
    "#path_to_regions = '/staging/leuven/stg_00002/lcb/cbravo/SCREEN_ENCODE3/V2.hg38-rDHS-Unfiltered.blacklisted.bed'\n",
    "path_to_blacklist = '/staging/leuven/stg_00002/lcb/cflerin/testruns/pycisTopic/pycisTopic/blacklist/hg38-blacklist.v2.bed'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "e854fba5-fcf1-4c4d-a3c6-c72e51f5da0f",
   "metadata": {},
   "outputs": [],
   "source": [
    "f_cto_dir = 'cistopic_objs__consensus_peaks'\n",
    "if not os.path.exists(os.path.join(wdir, f_cto_dir)):\n",
    "    os.makedirs(os.path.join(wdir, f_cto_dir))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "e0ce8585-a38d-46e6-9606-0b455996f08f",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Skipping Broad_1\n",
      "Skipping Broad_2\n",
      "Skipping Broad_mito_1\n",
      "Skipping Broad_mito_2\n",
      "Skipping CNAG_1\n",
      "Skipping CNAG_2\n",
      "Skipping s3atac\n",
      "Skipping Sanger_1\n",
      "Skipping Sanger_2\n",
      "Skipping Stanford_1\n",
      "Skipping Stanford_2\n",
      "Skipping VIB_1\n",
      "Skipping VIB_2\n"
     ]
    }
   ],
   "source": [
    "#Create all objects\n",
    "\n",
    "for key in bc_passing_filters.keys():\n",
    "    f_out = os.path.join(wdir, f_cto_dir, key + '__cistopic_obj.pkl')\n",
    "    if os.path.isfile(f_out):\n",
    "        print(f\"Skipping {key}\")\n",
    "        continue\n",
    "        \n",
    "    tmp_cto = create_cistopic_object_from_fragments(path_to_fragments=fragments_dict[key],\n",
    "                                                    path_to_regions=consensus_peaks_dict[key],\n",
    "                                                    path_to_blacklist=path_to_blacklist,\n",
    "                                                    #metrics=metadata_bc_dict[key],\n",
    "                                                    valid_bc=bc_passing_filters[key],\n",
    "                                                    n_cpu=1,\n",
    "                                                    partition=20,\n",
    "                                                    project=key)\n",
    "    \n",
    "    with open(f_out, 'wb') as f:\n",
    "        pickle.dump(tmp_cto, f)\n",
    "    \n",
    "    print(f\"DONE {key}\")\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e5f9c01a-ab62-4e67-9e16-5aa319bd357c",
   "metadata": {},
   "source": [
    "## Create a merged object\n",
    "\n",
    "Here, use a common set of peaks for each sample, then merge them"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "420bf6a4-ca4f-462e-b5ee-1d368faf2367",
   "metadata": {},
   "outputs": [],
   "source": [
    "f_cto_merged_dir = 'cistopic_objs__consensus_peaks_merged'\n",
    "if not os.path.exists(os.path.join(wdir, f_cto_merged_dir)):\n",
    "    os.makedirs(os.path.join(wdir, f_cto_merged_dir))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "bac432f1-dbdc-4c45-a2c3-ee1b838a0cb4",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Skipping Broad_1\n",
      "Skipping Broad_2\n",
      "Skipping Broad_mito_1\n",
      "Skipping Broad_mito_2\n",
      "Skipping CNAG_1\n",
      "Skipping CNAG_2\n",
      "Skipping s3atac\n",
      "Skipping Sanger_1\n",
      "Skipping Sanger_2\n",
      "Skipping Stanford_1\n",
      "Skipping Stanford_2\n",
      "Skipping VIB_1\n",
      "Skipping VIB_2\n"
     ]
    }
   ],
   "source": [
    "#Create all objects\n",
    "\n",
    "for key in bc_passing_filters.keys():\n",
    "    f_out = os.path.join(wdir, f_cto_merged_dir, key + '__cistopic_obj.pkl')\n",
    "    if os.path.isfile(f_out):\n",
    "        print(f\"Skipping {key}\")\n",
    "        continue\n",
    "        \n",
    "    tmp_cto = create_cistopic_object_from_fragments(path_to_fragments=fragments_dict[key],\n",
    "                                                    path_to_regions=consensus_peaks_dict['merged'],\n",
    "                                                    path_to_blacklist=path_to_blacklist,\n",
    "                                                    #metrics=metadata_bc_dict[key],\n",
    "                                                    valid_bc=bc_passing_filters[key],\n",
    "                                                    n_cpu=1,\n",
    "                                                    partition=20,\n",
    "                                                    project=key)\n",
    "    \n",
    "    with open(f_out, 'wb') as f:\n",
    "        pickle.dump(tmp_cto, f)\n",
    "    \n",
    "    print(f\"DONE {key}\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "785f24d3-a7cd-421d-803c-6d0044fc96a4",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Loaded Broad_1\n",
      "Loaded Broad_2\n",
      "Loaded Broad_mito_1\n",
      "Loaded Broad_mito_2\n",
      "Loaded CNAG_1\n",
      "Loaded CNAG_2\n",
      "Loaded s3atac\n",
      "Loaded Sanger_1\n",
      "Loaded Sanger_2\n",
      "Loaded Stanford_1\n",
      "Loaded Stanford_2\n",
      "Loaded VIB_1\n",
      "Loaded VIB_2\n"
     ]
    }
   ],
   "source": [
    "# load objects into dict:\n",
    "cistopic_obj_dict = {}\n",
    "for key in bc_passing_filters.keys():\n",
    "    f_out = os.path.join(wdir, f_cto_merged_dir, key + '__cistopic_obj.pkl')\n",
    "    with open(f_out, 'rb') as f:\n",
    "        cistopic_obj_dict[key] = pickle.load(f)\n",
    "    print(f\"Loaded {key}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "98db7b20-3a2d-496c-a555-f73f8cfae8f8",
   "metadata": {},
   "outputs": [],
   "source": [
    "import copy"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "2cf9b598-5921-49b9-ad81-5be4a3000082",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2021-09-15 11:49:43,074 cisTopic     INFO     cisTopic object 1 merged\n",
      "2021-09-15 11:49:48,865 cisTopic     INFO     cisTopic object 2 merged\n",
      "2021-09-15 11:49:55,561 cisTopic     INFO     cisTopic object 3 merged\n",
      "2021-09-15 11:50:03,457 cisTopic     INFO     cisTopic object 4 merged\n",
      "2021-09-15 11:50:11,848 cisTopic     INFO     cisTopic object 5 merged\n",
      "2021-09-15 11:50:20,721 cisTopic     INFO     cisTopic object 6 merged\n",
      "2021-09-15 11:50:30,925 cisTopic     INFO     cisTopic object 7 merged\n",
      "2021-09-15 11:50:42,607 cisTopic     INFO     cisTopic object 8 merged\n",
      "2021-09-15 11:50:54,504 cisTopic     INFO     cisTopic object 9 merged\n",
      "2021-09-15 11:51:06,880 cisTopic     INFO     cisTopic object 10 merged\n",
      "2021-09-15 11:51:20,550 cisTopic     INFO     cisTopic object 11 merged\n",
      "2021-09-15 11:51:36,110 cisTopic     INFO     cisTopic object 12 merged\n"
     ]
    }
   ],
   "source": [
    "f_out = os.path.join(wdir, f_cto_merged_dir, 'cellds_merged__cistopic_obj.pkl')\n",
    "if os.path.isfile(f_out):\n",
    "    print(f\"Loading {f_out}\")\n",
    "    with open(f_out, 'rb') as f:\n",
    "        cistopic_obj_dict['merged'] = pickle.load(f)\n",
    "else:\n",
    "    # merge:\n",
    "    ct_keys = list(cistopic_obj_dict.keys())\n",
    "    cto_merged = copy.copy(cistopic_obj_dict[ct_keys[0]])\n",
    "    cto_merged.merge([ cistopic_obj_dict[k] for k in ct_keys[1:] ], project='scATAC-seq_benchmark')\n",
    "    with open(f_out, 'wb') as f:\n",
    "        pickle.dump(cto_merged, f)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2c433422-4d98-487e-906f-333e86f1cbc6",
   "metadata": {},
   "source": [
    "## Create a set of cisTopic objects for downstream analysis\n",
    "\n",
    "Include each sample individually (using the sample-specific consensus peaks), plus the merged sample (using the merged consensus peaks)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "3a7f52cd-1f73-44d6-af2c-6b7de7ac6d89",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Loaded Broad_1\n",
      "Loaded Broad_2\n",
      "Loaded Broad_mito_1\n",
      "Loaded Broad_mito_2\n",
      "Loaded CNAG_1\n",
      "Loaded CNAG_2\n",
      "Loaded s3atac\n",
      "Loaded Sanger_1\n",
      "Loaded Sanger_2\n",
      "Loaded Stanford_1\n",
      "Loaded Stanford_2\n",
      "Loaded VIB_1\n",
      "Loaded VIB_2\n",
      "Loading /staging/leuven/stg_00002/lcb/cflerin/analysis/pbmc_atac/analysis2/data_freeze_Jun2021_cell_downsampled/atac_qc/jupyter/cistopic_objs__consensus_peaks_merged/cellds_merged__cistopic_obj.pkl\n"
     ]
    }
   ],
   "source": [
    "del cistopic_obj_dict\n",
    "\n",
    "cistopic_obj_dict = {}\n",
    "# load sample objects into dict:\n",
    "for key in bc_passing_filters.keys():\n",
    "    f_out = os.path.join(wdir, f_cto_dir, key + '__cistopic_obj.pkl')\n",
    "    with open(f_out, 'rb') as f:\n",
    "        cistopic_obj_dict[key] = pickle.load(f)\n",
    "    print(f\"Loaded {key}\")\n",
    "\n",
    "# add the merged object:\n",
    "f_out = os.path.join(wdir, f_cto_merged_dir, 'cellds_merged__cistopic_obj.pkl')\n",
    "if os.path.isfile(f_out):\n",
    "    print(f\"Loading {f_out}\")\n",
    "    with open(f_out, 'rb') as f:\n",
    "        cistopic_obj_dict['merged'] = pickle.load(f)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "969fa2cd-f422-4c3d-9450-56c70c0f4a53",
   "metadata": {},
   "source": [
    "## Add cell annotations"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "a09c75f8-5ada-4abe-8f37-3b0bc36a1c1a",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Doublet_scores_fragments</th>\n",
       "      <th>Total_nr_frag_in_regions</th>\n",
       "      <th>pycisTopic_leiden_10_1.2</th>\n",
       "      <th>Dupl_nr_frag</th>\n",
       "      <th>barcode</th>\n",
       "      <th>seurat_cell_type_pred_score</th>\n",
       "      <th>pycisTopic_leiden_10_1.0</th>\n",
       "      <th>pycisTopic_leiden_10_0.6</th>\n",
       "      <th>Unique_nr_frag_in_regions</th>\n",
       "      <th>Predicted_doublets_fragments</th>\n",
       "      <th>...</th>\n",
       "      <th>Dupl_rate</th>\n",
       "      <th>pycisTopic_leiden_10_0.8</th>\n",
       "      <th>Log_total_nr_frag</th>\n",
       "      <th>Total_nr_frag</th>\n",
       "      <th>cisTopic_nr_frag</th>\n",
       "      <th>Unique_nr_frag</th>\n",
       "      <th>pycisTopic_leiden_10_1.4</th>\n",
       "      <th>pycisTopic_leiden_10_1.6</th>\n",
       "      <th>pycisTopic_leiden_10_1.8</th>\n",
       "      <th>consensus_cell_type</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>TTCCTCTATGGTGTTTATCAT_GCGTAGACACGCCACTTCATC_ATAGGCACGAGTGGCGCCTAA-Broad_1</th>\n",
       "      <td>0.127469</td>\n",
       "      <td>6949</td>\n",
       "      <td>8</td>\n",
       "      <td>3440</td>\n",
       "      <td>TTCCTCTATGGTGTTTATCAT_GCGTAGACACGCCACTTCATC_AT...</td>\n",
       "      <td>0.682683</td>\n",
       "      <td>8</td>\n",
       "      <td>10</td>\n",
       "      <td>4268</td>\n",
       "      <td>False</td>\n",
       "      <td>...</td>\n",
       "      <td>0.370650</td>\n",
       "      <td>10</td>\n",
       "      <td>3.967595</td>\n",
       "      <td>9281</td>\n",
       "      <td>6019</td>\n",
       "      <td>5841</td>\n",
       "      <td>7</td>\n",
       "      <td>25</td>\n",
       "      <td>26</td>\n",
       "      <td>CD14+ monocyte</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GCGACTCAGACCATTTATCAT_CGACACTTCGCGCATTCCTCT_CGCAATCGAGGACACCGAACC-Broad_1</th>\n",
       "      <td>0.354582</td>\n",
       "      <td>7652</td>\n",
       "      <td>2</td>\n",
       "      <td>3538</td>\n",
       "      <td>GCGACTCAGACCATTTATCAT_CGACACTTCGCGCATTCCTCT_CG...</td>\n",
       "      <td>0.516413</td>\n",
       "      <td>2</td>\n",
       "      <td>3</td>\n",
       "      <td>4554</td>\n",
       "      <td>False</td>\n",
       "      <td>...</td>\n",
       "      <td>0.391026</td>\n",
       "      <td>2</td>\n",
       "      <td>3.956553</td>\n",
       "      <td>9048</td>\n",
       "      <td>5884</td>\n",
       "      <td>5510</td>\n",
       "      <td>4</td>\n",
       "      <td>6</td>\n",
       "      <td>7</td>\n",
       "      <td>Cytotoxic T cell</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TACCGAATGACCGCATTCGTT_CTCATTTACTCAATAGCAACG_TGTCGCTCGATTACCACATGA-Broad_1</th>\n",
       "      <td>0.235127</td>\n",
       "      <td>7275</td>\n",
       "      <td>0</td>\n",
       "      <td>4158</td>\n",
       "      <td>TACCGAATGACCGCATTCGTT_CTCATTTACTCAATAGCAACG_TG...</td>\n",
       "      <td>0.612734</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>3891</td>\n",
       "      <td>False</td>\n",
       "      <td>...</td>\n",
       "      <td>0.443615</td>\n",
       "      <td>0</td>\n",
       "      <td>3.971879</td>\n",
       "      <td>9373</td>\n",
       "      <td>5470</td>\n",
       "      <td>5215</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>5</td>\n",
       "      <td>CD4+ T cell</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ACGTTGGGAACCGTCCTCCTT_AACTCTTGGTTAGTCTCTTGC_GATCACCTTATCATGAATATG_TGGTGAATCCTTAAGGAGCCT-Broad_1</th>\n",
       "      <td>0.284768</td>\n",
       "      <td>8422</td>\n",
       "      <td>0</td>\n",
       "      <td>4752</td>\n",
       "      <td>ACGTTGGGAACCGTCCTCCTT_AACTCTTGGTTAGTCTCTTGC_GA...</td>\n",
       "      <td>0.783805</td>\n",
       "      <td>18</td>\n",
       "      <td>1</td>\n",
       "      <td>4193</td>\n",
       "      <td>False</td>\n",
       "      <td>...</td>\n",
       "      <td>0.480631</td>\n",
       "      <td>0</td>\n",
       "      <td>3.995065</td>\n",
       "      <td>9887</td>\n",
       "      <td>5322</td>\n",
       "      <td>5135</td>\n",
       "      <td>10</td>\n",
       "      <td>7</td>\n",
       "      <td>6</td>\n",
       "      <td>CD4+ T cell</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TCATACACCAAGCTCATGTAT_AGGATACCCGCGATATGTTCC_TAACGCCGAATCAAGAATCAA-Broad_1</th>\n",
       "      <td>0.508197</td>\n",
       "      <td>6925</td>\n",
       "      <td>8</td>\n",
       "      <td>3824</td>\n",
       "      <td>TCATACACCAAGCTCATGTAT_AGGATACCCGCGATATGTTCC_TA...</td>\n",
       "      <td>0.616381</td>\n",
       "      <td>8</td>\n",
       "      <td>10</td>\n",
       "      <td>3700</td>\n",
       "      <td>False</td>\n",
       "      <td>...</td>\n",
       "      <td>0.440705</td>\n",
       "      <td>10</td>\n",
       "      <td>3.938370</td>\n",
       "      <td>8677</td>\n",
       "      <td>4584</td>\n",
       "      <td>4853</td>\n",
       "      <td>7</td>\n",
       "      <td>25</td>\n",
       "      <td>26</td>\n",
       "      <td>CD14+ monocyte</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CTTCCAGCAAGGTCGA-VIB_2</th>\n",
       "      <td>0.117647</td>\n",
       "      <td>5697</td>\n",
       "      <td>21</td>\n",
       "      <td>1166</td>\n",
       "      <td>CTTCCAGCAAGGTCGA</td>\n",
       "      <td>0.961163</td>\n",
       "      <td>16</td>\n",
       "      <td>0</td>\n",
       "      <td>4843</td>\n",
       "      <td>False</td>\n",
       "      <td>...</td>\n",
       "      <td>0.133425</td>\n",
       "      <td>20</td>\n",
       "      <td>3.941462</td>\n",
       "      <td>8739</td>\n",
       "      <td>6498</td>\n",
       "      <td>7573</td>\n",
       "      <td>14</td>\n",
       "      <td>14</td>\n",
       "      <td>13</td>\n",
       "      <td>CD14+ monocyte</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GTCCATCAGTACCTCA-VIB_2</th>\n",
       "      <td>0.131222</td>\n",
       "      <td>5164</td>\n",
       "      <td>1</td>\n",
       "      <td>1389</td>\n",
       "      <td>GTCCATCAGTACCTCA</td>\n",
       "      <td>0.979147</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>4177</td>\n",
       "      <td>False</td>\n",
       "      <td>...</td>\n",
       "      <td>0.159216</td>\n",
       "      <td>1</td>\n",
       "      <td>3.940716</td>\n",
       "      <td>8724</td>\n",
       "      <td>5883</td>\n",
       "      <td>7335</td>\n",
       "      <td>3</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>CD14+ monocyte</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TTGTTCACAACTCGAT-VIB_2</th>\n",
       "      <td>0.278075</td>\n",
       "      <td>5062</td>\n",
       "      <td>21</td>\n",
       "      <td>2031</td>\n",
       "      <td>TTGTTCACAACTCGAT</td>\n",
       "      <td>0.782794</td>\n",
       "      <td>16</td>\n",
       "      <td>0</td>\n",
       "      <td>3726</td>\n",
       "      <td>False</td>\n",
       "      <td>...</td>\n",
       "      <td>0.225441</td>\n",
       "      <td>20</td>\n",
       "      <td>3.954677</td>\n",
       "      <td>9009</td>\n",
       "      <td>5464</td>\n",
       "      <td>6978</td>\n",
       "      <td>14</td>\n",
       "      <td>14</td>\n",
       "      <td>13</td>\n",
       "      <td>CD14+ monocyte</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CTTCCAGTCAGTGTGT-VIB_2</th>\n",
       "      <td>0.165775</td>\n",
       "      <td>6068</td>\n",
       "      <td>1</td>\n",
       "      <td>1931</td>\n",
       "      <td>CTTCCAGTCAGTGTGT</td>\n",
       "      <td>0.859730</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>4579</td>\n",
       "      <td>False</td>\n",
       "      <td>...</td>\n",
       "      <td>0.201230</td>\n",
       "      <td>1</td>\n",
       "      <td>3.982090</td>\n",
       "      <td>9596</td>\n",
       "      <td>6208</td>\n",
       "      <td>7665</td>\n",
       "      <td>14</td>\n",
       "      <td>2</td>\n",
       "      <td>13</td>\n",
       "      <td>CD14+ monocyte</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GGAGAACAGTTACACC-VIB_2</th>\n",
       "      <td>0.102302</td>\n",
       "      <td>3807</td>\n",
       "      <td>1</td>\n",
       "      <td>824</td>\n",
       "      <td>GGAGAACAGTTACACC</td>\n",
       "      <td>0.864884</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>3298</td>\n",
       "      <td>False</td>\n",
       "      <td>...</td>\n",
       "      <td>0.091892</td>\n",
       "      <td>1</td>\n",
       "      <td>3.952647</td>\n",
       "      <td>8967</td>\n",
       "      <td>5396</td>\n",
       "      <td>8143</td>\n",
       "      <td>3</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>CD14+ monocyte</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>39580 rows × 30 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                    Doublet_scores_fragments  \\\n",
       "TTCCTCTATGGTGTTTATCAT_GCGTAGACACGCCACTTCATC_ATA...                  0.127469   \n",
       "GCGACTCAGACCATTTATCAT_CGACACTTCGCGCATTCCTCT_CGC...                  0.354582   \n",
       "TACCGAATGACCGCATTCGTT_CTCATTTACTCAATAGCAACG_TGT...                  0.235127   \n",
       "ACGTTGGGAACCGTCCTCCTT_AACTCTTGGTTAGTCTCTTGC_GAT...                  0.284768   \n",
       "TCATACACCAAGCTCATGTAT_AGGATACCCGCGATATGTTCC_TAA...                  0.508197   \n",
       "...                                                                      ...   \n",
       "CTTCCAGCAAGGTCGA-VIB_2                                              0.117647   \n",
       "GTCCATCAGTACCTCA-VIB_2                                              0.131222   \n",
       "TTGTTCACAACTCGAT-VIB_2                                              0.278075   \n",
       "CTTCCAGTCAGTGTGT-VIB_2                                              0.165775   \n",
       "GGAGAACAGTTACACC-VIB_2                                              0.102302   \n",
       "\n",
       "                                                    Total_nr_frag_in_regions  \\\n",
       "TTCCTCTATGGTGTTTATCAT_GCGTAGACACGCCACTTCATC_ATA...                      6949   \n",
       "GCGACTCAGACCATTTATCAT_CGACACTTCGCGCATTCCTCT_CGC...                      7652   \n",
       "TACCGAATGACCGCATTCGTT_CTCATTTACTCAATAGCAACG_TGT...                      7275   \n",
       "ACGTTGGGAACCGTCCTCCTT_AACTCTTGGTTAGTCTCTTGC_GAT...                      8422   \n",
       "TCATACACCAAGCTCATGTAT_AGGATACCCGCGATATGTTCC_TAA...                      6925   \n",
       "...                                                                      ...   \n",
       "CTTCCAGCAAGGTCGA-VIB_2                                                  5697   \n",
       "GTCCATCAGTACCTCA-VIB_2                                                  5164   \n",
       "TTGTTCACAACTCGAT-VIB_2                                                  5062   \n",
       "CTTCCAGTCAGTGTGT-VIB_2                                                  6068   \n",
       "GGAGAACAGTTACACC-VIB_2                                                  3807   \n",
       "\n",
       "                                                    pycisTopic_leiden_10_1.2  \\\n",
       "TTCCTCTATGGTGTTTATCAT_GCGTAGACACGCCACTTCATC_ATA...                         8   \n",
       "GCGACTCAGACCATTTATCAT_CGACACTTCGCGCATTCCTCT_CGC...                         2   \n",
       "TACCGAATGACCGCATTCGTT_CTCATTTACTCAATAGCAACG_TGT...                         0   \n",
       "ACGTTGGGAACCGTCCTCCTT_AACTCTTGGTTAGTCTCTTGC_GAT...                         0   \n",
       "TCATACACCAAGCTCATGTAT_AGGATACCCGCGATATGTTCC_TAA...                         8   \n",
       "...                                                                      ...   \n",
       "CTTCCAGCAAGGTCGA-VIB_2                                                    21   \n",
       "GTCCATCAGTACCTCA-VIB_2                                                     1   \n",
       "TTGTTCACAACTCGAT-VIB_2                                                    21   \n",
       "CTTCCAGTCAGTGTGT-VIB_2                                                     1   \n",
       "GGAGAACAGTTACACC-VIB_2                                                     1   \n",
       "\n",
       "                                                    Dupl_nr_frag  \\\n",
       "TTCCTCTATGGTGTTTATCAT_GCGTAGACACGCCACTTCATC_ATA...          3440   \n",
       "GCGACTCAGACCATTTATCAT_CGACACTTCGCGCATTCCTCT_CGC...          3538   \n",
       "TACCGAATGACCGCATTCGTT_CTCATTTACTCAATAGCAACG_TGT...          4158   \n",
       "ACGTTGGGAACCGTCCTCCTT_AACTCTTGGTTAGTCTCTTGC_GAT...          4752   \n",
       "TCATACACCAAGCTCATGTAT_AGGATACCCGCGATATGTTCC_TAA...          3824   \n",
       "...                                                          ...   \n",
       "CTTCCAGCAAGGTCGA-VIB_2                                      1166   \n",
       "GTCCATCAGTACCTCA-VIB_2                                      1389   \n",
       "TTGTTCACAACTCGAT-VIB_2                                      2031   \n",
       "CTTCCAGTCAGTGTGT-VIB_2                                      1931   \n",
       "GGAGAACAGTTACACC-VIB_2                                       824   \n",
       "\n",
       "                                                                                              barcode  \\\n",
       "TTCCTCTATGGTGTTTATCAT_GCGTAGACACGCCACTTCATC_ATA...  TTCCTCTATGGTGTTTATCAT_GCGTAGACACGCCACTTCATC_AT...   \n",
       "GCGACTCAGACCATTTATCAT_CGACACTTCGCGCATTCCTCT_CGC...  GCGACTCAGACCATTTATCAT_CGACACTTCGCGCATTCCTCT_CG...   \n",
       "TACCGAATGACCGCATTCGTT_CTCATTTACTCAATAGCAACG_TGT...  TACCGAATGACCGCATTCGTT_CTCATTTACTCAATAGCAACG_TG...   \n",
       "ACGTTGGGAACCGTCCTCCTT_AACTCTTGGTTAGTCTCTTGC_GAT...  ACGTTGGGAACCGTCCTCCTT_AACTCTTGGTTAGTCTCTTGC_GA...   \n",
       "TCATACACCAAGCTCATGTAT_AGGATACCCGCGATATGTTCC_TAA...  TCATACACCAAGCTCATGTAT_AGGATACCCGCGATATGTTCC_TA...   \n",
       "...                                                                                               ...   \n",
       "CTTCCAGCAAGGTCGA-VIB_2                                                               CTTCCAGCAAGGTCGA   \n",
       "GTCCATCAGTACCTCA-VIB_2                                                               GTCCATCAGTACCTCA   \n",
       "TTGTTCACAACTCGAT-VIB_2                                                               TTGTTCACAACTCGAT   \n",
       "CTTCCAGTCAGTGTGT-VIB_2                                                               CTTCCAGTCAGTGTGT   \n",
       "GGAGAACAGTTACACC-VIB_2                                                               GGAGAACAGTTACACC   \n",
       "\n",
       "                                                    seurat_cell_type_pred_score  \\\n",
       "TTCCTCTATGGTGTTTATCAT_GCGTAGACACGCCACTTCATC_ATA...                     0.682683   \n",
       "GCGACTCAGACCATTTATCAT_CGACACTTCGCGCATTCCTCT_CGC...                     0.516413   \n",
       "TACCGAATGACCGCATTCGTT_CTCATTTACTCAATAGCAACG_TGT...                     0.612734   \n",
       "ACGTTGGGAACCGTCCTCCTT_AACTCTTGGTTAGTCTCTTGC_GAT...                     0.783805   \n",
       "TCATACACCAAGCTCATGTAT_AGGATACCCGCGATATGTTCC_TAA...                     0.616381   \n",
       "...                                                                         ...   \n",
       "CTTCCAGCAAGGTCGA-VIB_2                                                 0.961163   \n",
       "GTCCATCAGTACCTCA-VIB_2                                                 0.979147   \n",
       "TTGTTCACAACTCGAT-VIB_2                                                 0.782794   \n",
       "CTTCCAGTCAGTGTGT-VIB_2                                                 0.859730   \n",
       "GGAGAACAGTTACACC-VIB_2                                                 0.864884   \n",
       "\n",
       "                                                    pycisTopic_leiden_10_1.0  \\\n",
       "TTCCTCTATGGTGTTTATCAT_GCGTAGACACGCCACTTCATC_ATA...                         8   \n",
       "GCGACTCAGACCATTTATCAT_CGACACTTCGCGCATTCCTCT_CGC...                         2   \n",
       "TACCGAATGACCGCATTCGTT_CTCATTTACTCAATAGCAACG_TGT...                         1   \n",
       "ACGTTGGGAACCGTCCTCCTT_AACTCTTGGTTAGTCTCTTGC_GAT...                        18   \n",
       "TCATACACCAAGCTCATGTAT_AGGATACCCGCGATATGTTCC_TAA...                         8   \n",
       "...                                                                      ...   \n",
       "CTTCCAGCAAGGTCGA-VIB_2                                                    16   \n",
       "GTCCATCAGTACCTCA-VIB_2                                                     0   \n",
       "TTGTTCACAACTCGAT-VIB_2                                                    16   \n",
       "CTTCCAGTCAGTGTGT-VIB_2                                                     0   \n",
       "GGAGAACAGTTACACC-VIB_2                                                     0   \n",
       "\n",
       "                                                    pycisTopic_leiden_10_0.6  \\\n",
       "TTCCTCTATGGTGTTTATCAT_GCGTAGACACGCCACTTCATC_ATA...                        10   \n",
       "GCGACTCAGACCATTTATCAT_CGACACTTCGCGCATTCCTCT_CGC...                         3   \n",
       "TACCGAATGACCGCATTCGTT_CTCATTTACTCAATAGCAACG_TGT...                         1   \n",
       "ACGTTGGGAACCGTCCTCCTT_AACTCTTGGTTAGTCTCTTGC_GAT...                         1   \n",
       "TCATACACCAAGCTCATGTAT_AGGATACCCGCGATATGTTCC_TAA...                        10   \n",
       "...                                                                      ...   \n",
       "CTTCCAGCAAGGTCGA-VIB_2                                                     0   \n",
       "GTCCATCAGTACCTCA-VIB_2                                                     0   \n",
       "TTGTTCACAACTCGAT-VIB_2                                                     0   \n",
       "CTTCCAGTCAGTGTGT-VIB_2                                                     0   \n",
       "GGAGAACAGTTACACC-VIB_2                                                     0   \n",
       "\n",
       "                                                    Unique_nr_frag_in_regions  \\\n",
       "TTCCTCTATGGTGTTTATCAT_GCGTAGACACGCCACTTCATC_ATA...                       4268   \n",
       "GCGACTCAGACCATTTATCAT_CGACACTTCGCGCATTCCTCT_CGC...                       4554   \n",
       "TACCGAATGACCGCATTCGTT_CTCATTTACTCAATAGCAACG_TGT...                       3891   \n",
       "ACGTTGGGAACCGTCCTCCTT_AACTCTTGGTTAGTCTCTTGC_GAT...                       4193   \n",
       "TCATACACCAAGCTCATGTAT_AGGATACCCGCGATATGTTCC_TAA...                       3700   \n",
       "...                                                                       ...   \n",
       "CTTCCAGCAAGGTCGA-VIB_2                                                   4843   \n",
       "GTCCATCAGTACCTCA-VIB_2                                                   4177   \n",
       "TTGTTCACAACTCGAT-VIB_2                                                   3726   \n",
       "CTTCCAGTCAGTGTGT-VIB_2                                                   4579   \n",
       "GGAGAACAGTTACACC-VIB_2                                                   3298   \n",
       "\n",
       "                                                    Predicted_doublets_fragments  \\\n",
       "TTCCTCTATGGTGTTTATCAT_GCGTAGACACGCCACTTCATC_ATA...                         False   \n",
       "GCGACTCAGACCATTTATCAT_CGACACTTCGCGCATTCCTCT_CGC...                         False   \n",
       "TACCGAATGACCGCATTCGTT_CTCATTTACTCAATAGCAACG_TGT...                         False   \n",
       "ACGTTGGGAACCGTCCTCCTT_AACTCTTGGTTAGTCTCTTGC_GAT...                         False   \n",
       "TCATACACCAAGCTCATGTAT_AGGATACCCGCGATATGTTCC_TAA...                         False   \n",
       "...                                                                          ...   \n",
       "CTTCCAGCAAGGTCGA-VIB_2                                                     False   \n",
       "GTCCATCAGTACCTCA-VIB_2                                                     False   \n",
       "TTGTTCACAACTCGAT-VIB_2                                                     False   \n",
       "CTTCCAGTCAGTGTGT-VIB_2                                                     False   \n",
       "GGAGAACAGTTACACC-VIB_2                                                     False   \n",
       "\n",
       "                                                    ... Dupl_rate  \\\n",
       "TTCCTCTATGGTGTTTATCAT_GCGTAGACACGCCACTTCATC_ATA...  ...  0.370650   \n",
       "GCGACTCAGACCATTTATCAT_CGACACTTCGCGCATTCCTCT_CGC...  ...  0.391026   \n",
       "TACCGAATGACCGCATTCGTT_CTCATTTACTCAATAGCAACG_TGT...  ...  0.443615   \n",
       "ACGTTGGGAACCGTCCTCCTT_AACTCTTGGTTAGTCTCTTGC_GAT...  ...  0.480631   \n",
       "TCATACACCAAGCTCATGTAT_AGGATACCCGCGATATGTTCC_TAA...  ...  0.440705   \n",
       "...                                                 ...       ...   \n",
       "CTTCCAGCAAGGTCGA-VIB_2                              ...  0.133425   \n",
       "GTCCATCAGTACCTCA-VIB_2                              ...  0.159216   \n",
       "TTGTTCACAACTCGAT-VIB_2                              ...  0.225441   \n",
       "CTTCCAGTCAGTGTGT-VIB_2                              ...  0.201230   \n",
       "GGAGAACAGTTACACC-VIB_2                              ...  0.091892   \n",
       "\n",
       "                                                    pycisTopic_leiden_10_0.8  \\\n",
       "TTCCTCTATGGTGTTTATCAT_GCGTAGACACGCCACTTCATC_ATA...                        10   \n",
       "GCGACTCAGACCATTTATCAT_CGACACTTCGCGCATTCCTCT_CGC...                         2   \n",
       "TACCGAATGACCGCATTCGTT_CTCATTTACTCAATAGCAACG_TGT...                         0   \n",
       "ACGTTGGGAACCGTCCTCCTT_AACTCTTGGTTAGTCTCTTGC_GAT...                         0   \n",
       "TCATACACCAAGCTCATGTAT_AGGATACCCGCGATATGTTCC_TAA...                        10   \n",
       "...                                                                      ...   \n",
       "CTTCCAGCAAGGTCGA-VIB_2                                                    20   \n",
       "GTCCATCAGTACCTCA-VIB_2                                                     1   \n",
       "TTGTTCACAACTCGAT-VIB_2                                                    20   \n",
       "CTTCCAGTCAGTGTGT-VIB_2                                                     1   \n",
       "GGAGAACAGTTACACC-VIB_2                                                     1   \n",
       "\n",
       "                                                    Log_total_nr_frag  \\\n",
       "TTCCTCTATGGTGTTTATCAT_GCGTAGACACGCCACTTCATC_ATA...           3.967595   \n",
       "GCGACTCAGACCATTTATCAT_CGACACTTCGCGCATTCCTCT_CGC...           3.956553   \n",
       "TACCGAATGACCGCATTCGTT_CTCATTTACTCAATAGCAACG_TGT...           3.971879   \n",
       "ACGTTGGGAACCGTCCTCCTT_AACTCTTGGTTAGTCTCTTGC_GAT...           3.995065   \n",
       "TCATACACCAAGCTCATGTAT_AGGATACCCGCGATATGTTCC_TAA...           3.938370   \n",
       "...                                                               ...   \n",
       "CTTCCAGCAAGGTCGA-VIB_2                                       3.941462   \n",
       "GTCCATCAGTACCTCA-VIB_2                                       3.940716   \n",
       "TTGTTCACAACTCGAT-VIB_2                                       3.954677   \n",
       "CTTCCAGTCAGTGTGT-VIB_2                                       3.982090   \n",
       "GGAGAACAGTTACACC-VIB_2                                       3.952647   \n",
       "\n",
       "                                                   Total_nr_frag  \\\n",
       "TTCCTCTATGGTGTTTATCAT_GCGTAGACACGCCACTTCATC_ATA...          9281   \n",
       "GCGACTCAGACCATTTATCAT_CGACACTTCGCGCATTCCTCT_CGC...          9048   \n",
       "TACCGAATGACCGCATTCGTT_CTCATTTACTCAATAGCAACG_TGT...          9373   \n",
       "ACGTTGGGAACCGTCCTCCTT_AACTCTTGGTTAGTCTCTTGC_GAT...          9887   \n",
       "TCATACACCAAGCTCATGTAT_AGGATACCCGCGATATGTTCC_TAA...          8677   \n",
       "...                                                          ...   \n",
       "CTTCCAGCAAGGTCGA-VIB_2                                      8739   \n",
       "GTCCATCAGTACCTCA-VIB_2                                      8724   \n",
       "TTGTTCACAACTCGAT-VIB_2                                      9009   \n",
       "CTTCCAGTCAGTGTGT-VIB_2                                      9596   \n",
       "GGAGAACAGTTACACC-VIB_2                                      8967   \n",
       "\n",
       "                                                   cisTopic_nr_frag  \\\n",
       "TTCCTCTATGGTGTTTATCAT_GCGTAGACACGCCACTTCATC_ATA...             6019   \n",
       "GCGACTCAGACCATTTATCAT_CGACACTTCGCGCATTCCTCT_CGC...             5884   \n",
       "TACCGAATGACCGCATTCGTT_CTCATTTACTCAATAGCAACG_TGT...             5470   \n",
       "ACGTTGGGAACCGTCCTCCTT_AACTCTTGGTTAGTCTCTTGC_GAT...             5322   \n",
       "TCATACACCAAGCTCATGTAT_AGGATACCCGCGATATGTTCC_TAA...             4584   \n",
       "...                                                             ...   \n",
       "CTTCCAGCAAGGTCGA-VIB_2                                         6498   \n",
       "GTCCATCAGTACCTCA-VIB_2                                         5883   \n",
       "TTGTTCACAACTCGAT-VIB_2                                         5464   \n",
       "CTTCCAGTCAGTGTGT-VIB_2                                         6208   \n",
       "GGAGAACAGTTACACC-VIB_2                                         5396   \n",
       "\n",
       "                                                    Unique_nr_frag  \\\n",
       "TTCCTCTATGGTGTTTATCAT_GCGTAGACACGCCACTTCATC_ATA...            5841   \n",
       "GCGACTCAGACCATTTATCAT_CGACACTTCGCGCATTCCTCT_CGC...            5510   \n",
       "TACCGAATGACCGCATTCGTT_CTCATTTACTCAATAGCAACG_TGT...            5215   \n",
       "ACGTTGGGAACCGTCCTCCTT_AACTCTTGGTTAGTCTCTTGC_GAT...            5135   \n",
       "TCATACACCAAGCTCATGTAT_AGGATACCCGCGATATGTTCC_TAA...            4853   \n",
       "...                                                            ...   \n",
       "CTTCCAGCAAGGTCGA-VIB_2                                        7573   \n",
       "GTCCATCAGTACCTCA-VIB_2                                        7335   \n",
       "TTGTTCACAACTCGAT-VIB_2                                        6978   \n",
       "CTTCCAGTCAGTGTGT-VIB_2                                        7665   \n",
       "GGAGAACAGTTACACC-VIB_2                                        8143   \n",
       "\n",
       "                                                    pycisTopic_leiden_10_1.4  \\\n",
       "TTCCTCTATGGTGTTTATCAT_GCGTAGACACGCCACTTCATC_ATA...                         7   \n",
       "GCGACTCAGACCATTTATCAT_CGACACTTCGCGCATTCCTCT_CGC...                         4   \n",
       "TACCGAATGACCGCATTCGTT_CTCATTTACTCAATAGCAACG_TGT...                         0   \n",
       "ACGTTGGGAACCGTCCTCCTT_AACTCTTGGTTAGTCTCTTGC_GAT...                        10   \n",
       "TCATACACCAAGCTCATGTAT_AGGATACCCGCGATATGTTCC_TAA...                         7   \n",
       "...                                                                      ...   \n",
       "CTTCCAGCAAGGTCGA-VIB_2                                                    14   \n",
       "GTCCATCAGTACCTCA-VIB_2                                                     3   \n",
       "TTGTTCACAACTCGAT-VIB_2                                                    14   \n",
       "CTTCCAGTCAGTGTGT-VIB_2                                                    14   \n",
       "GGAGAACAGTTACACC-VIB_2                                                     3   \n",
       "\n",
       "                                                    pycisTopic_leiden_10_1.6  \\\n",
       "TTCCTCTATGGTGTTTATCAT_GCGTAGACACGCCACTTCATC_ATA...                        25   \n",
       "GCGACTCAGACCATTTATCAT_CGACACTTCGCGCATTCCTCT_CGC...                         6   \n",
       "TACCGAATGACCGCATTCGTT_CTCATTTACTCAATAGCAACG_TGT...                         0   \n",
       "ACGTTGGGAACCGTCCTCCTT_AACTCTTGGTTAGTCTCTTGC_GAT...                         7   \n",
       "TCATACACCAAGCTCATGTAT_AGGATACCCGCGATATGTTCC_TAA...                        25   \n",
       "...                                                                      ...   \n",
       "CTTCCAGCAAGGTCGA-VIB_2                                                    14   \n",
       "GTCCATCAGTACCTCA-VIB_2                                                     2   \n",
       "TTGTTCACAACTCGAT-VIB_2                                                    14   \n",
       "CTTCCAGTCAGTGTGT-VIB_2                                                     2   \n",
       "GGAGAACAGTTACACC-VIB_2                                                     2   \n",
       "\n",
       "                                                   pycisTopic_leiden_10_1.8  \\\n",
       "TTCCTCTATGGTGTTTATCAT_GCGTAGACACGCCACTTCATC_ATA...                       26   \n",
       "GCGACTCAGACCATTTATCAT_CGACACTTCGCGCATTCCTCT_CGC...                        7   \n",
       "TACCGAATGACCGCATTCGTT_CTCATTTACTCAATAGCAACG_TGT...                        5   \n",
       "ACGTTGGGAACCGTCCTCCTT_AACTCTTGGTTAGTCTCTTGC_GAT...                        6   \n",
       "TCATACACCAAGCTCATGTAT_AGGATACCCGCGATATGTTCC_TAA...                       26   \n",
       "...                                                                     ...   \n",
       "CTTCCAGCAAGGTCGA-VIB_2                                                   13   \n",
       "GTCCATCAGTACCTCA-VIB_2                                                    2   \n",
       "TTGTTCACAACTCGAT-VIB_2                                                   13   \n",
       "CTTCCAGTCAGTGTGT-VIB_2                                                   13   \n",
       "GGAGAACAGTTACACC-VIB_2                                                    2   \n",
       "\n",
       "                                                    consensus_cell_type  \n",
       "TTCCTCTATGGTGTTTATCAT_GCGTAGACACGCCACTTCATC_ATA...       CD14+ monocyte  \n",
       "GCGACTCAGACCATTTATCAT_CGACACTTCGCGCATTCCTCT_CGC...     Cytotoxic T cell  \n",
       "TACCGAATGACCGCATTCGTT_CTCATTTACTCAATAGCAACG_TGT...          CD4+ T cell  \n",
       "ACGTTGGGAACCGTCCTCCTT_AACTCTTGGTTAGTCTCTTGC_GAT...          CD4+ T cell  \n",
       "TCATACACCAAGCTCATGTAT_AGGATACCCGCGATATGTTCC_TAA...       CD14+ monocyte  \n",
       "...                                                                 ...  \n",
       "CTTCCAGCAAGGTCGA-VIB_2                                   CD14+ monocyte  \n",
       "GTCCATCAGTACCTCA-VIB_2                                   CD14+ monocyte  \n",
       "TTGTTCACAACTCGAT-VIB_2                                   CD14+ monocyte  \n",
       "CTTCCAGTCAGTGTGT-VIB_2                                   CD14+ monocyte  \n",
       "GGAGAACAGTTACACC-VIB_2                                   CD14+ monocyte  \n",
       "\n",
       "[39580 rows x 30 columns]"
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "cellannot = pd.read_csv(\n",
    "    '/staging/leuven/stg_00002/lcb/cflerin/analysis/pbmc_atac/analysis2/data_freeze_Jun2021_cell_downsampled/atac_qc/jupyter/cellds_filtered_cell_data.tsv',\n",
    "    sep='\\t', index_col=0)\n",
    "cellannot"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "08df0b7b-3f66-454e-86aa-81bdfd4d15e8",
   "metadata": {},
   "outputs": [],
   "source": [
    "for key in cistopic_obj_dict.keys():\n",
    "    cistopic_obj_dict[key].add_cell_data(\n",
    "        cellannot[['consensus_cell_type']]\n",
    "    )"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "58d893da-50cf-4b55-94c2-42cbe304b31c",
   "metadata": {},
   "source": [
    "### Run models for quick visualize/clustering"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "c92e38c4-5d8f-4946-8ea1-b1997d713b7d",
   "metadata": {},
   "outputs": [],
   "source": [
    "from pycisTopic.lda_models import run_cgs_models_mallet"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "96a50b14-241d-46c0-9d53-c3cac694370d",
   "metadata": {},
   "source": [
    "#### mallet models"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "6c0a420b-fec6-4de2-9473-9a99ec92da26",
   "metadata": {},
   "outputs": [],
   "source": [
    "# set the memory limit for mallet (1GB default is not enough)\n",
    "os.environ['MALLET_MEMORY'] = '100G'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "d6e970ec-1589-4afd-98a0-66eaf3ada1c9",
   "metadata": {},
   "outputs": [],
   "source": [
    "f_mod_dir = 'models__consensus__mallet'\n",
    "if not os.path.exists(os.path.join(wdir, f_mod_dir)):\n",
    "    os.makedirs(os.path.join(wdir, f_mod_dir))\n",
    "\n",
    "f_mod_tmpdir = '/scratch/leuven/325/vsc32528/tmp/mallet_cellds'\n",
    "if not os.path.exists(f_mod_tmpdir):\n",
    "    os.makedirs(f_mod_tmpdir)\n",
    "\n",
    "n_topics = [2, 5, 10, 15, 20, 25, 30, 35, 40, 45, 50]\n",
    "n_topics_merged = [2, 5, 10, 20, 30, 35, 40, 45, 50, 60, 70, 80, 90, 100]\n",
    "n_iter = 500"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b2ff4242-19d1-4275-9d2e-a6741e4334f5",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Loading /staging/leuven/stg_00002/lcb/cflerin/analysis/pbmc_atac/analysis2/data_freeze_Jun2021_cell_downsampled/atac_qc/jupyter/models__consensus__mallet/Broad_1__models_500_iter.pkl\n",
      "Loading /staging/leuven/stg_00002/lcb/cflerin/analysis/pbmc_atac/analysis2/data_freeze_Jun2021_cell_downsampled/atac_qc/jupyter/models__consensus__mallet/Broad_2__models_500_iter.pkl\n",
      "Loading /staging/leuven/stg_00002/lcb/cflerin/analysis/pbmc_atac/analysis2/data_freeze_Jun2021_cell_downsampled/atac_qc/jupyter/models__consensus__mallet/Broad_mito_1__models_500_iter.pkl\n",
      "Loading /staging/leuven/stg_00002/lcb/cflerin/analysis/pbmc_atac/analysis2/data_freeze_Jun2021_cell_downsampled/atac_qc/jupyter/models__consensus__mallet/Broad_mito_2__models_500_iter.pkl\n",
      "Loading /staging/leuven/stg_00002/lcb/cflerin/analysis/pbmc_atac/analysis2/data_freeze_Jun2021_cell_downsampled/atac_qc/jupyter/models__consensus__mallet/CNAG_1__models_500_iter.pkl\n",
      "Loading /staging/leuven/stg_00002/lcb/cflerin/analysis/pbmc_atac/analysis2/data_freeze_Jun2021_cell_downsampled/atac_qc/jupyter/models__consensus__mallet/CNAG_2__models_500_iter.pkl\n",
      "Loading /staging/leuven/stg_00002/lcb/cflerin/analysis/pbmc_atac/analysis2/data_freeze_Jun2021_cell_downsampled/atac_qc/jupyter/models__consensus__mallet/s3atac__models_500_iter.pkl\n",
      "Loading /staging/leuven/stg_00002/lcb/cflerin/analysis/pbmc_atac/analysis2/data_freeze_Jun2021_cell_downsampled/atac_qc/jupyter/models__consensus__mallet/Sanger_1__models_500_iter.pkl\n",
      "Loading /staging/leuven/stg_00002/lcb/cflerin/analysis/pbmc_atac/analysis2/data_freeze_Jun2021_cell_downsampled/atac_qc/jupyter/models__consensus__mallet/Sanger_2__models_500_iter.pkl\n",
      "Loading /staging/leuven/stg_00002/lcb/cflerin/analysis/pbmc_atac/analysis2/data_freeze_Jun2021_cell_downsampled/atac_qc/jupyter/models__consensus__mallet/Stanford_1__models_500_iter.pkl\n",
      "Loading /staging/leuven/stg_00002/lcb/cflerin/analysis/pbmc_atac/analysis2/data_freeze_Jun2021_cell_downsampled/atac_qc/jupyter/models__consensus__mallet/Stanford_2__models_500_iter.pkl\n",
      "Loading /staging/leuven/stg_00002/lcb/cflerin/analysis/pbmc_atac/analysis2/data_freeze_Jun2021_cell_downsampled/atac_qc/jupyter/models__consensus__mallet/VIB_1__models_500_iter.pkl\n",
      "Loading /staging/leuven/stg_00002/lcb/cflerin/analysis/pbmc_atac/analysis2/data_freeze_Jun2021_cell_downsampled/atac_qc/jupyter/models__consensus__mallet/VIB_2__models_500_iter.pkl\n",
      "Running merged\n",
      "2021-09-15 11:56:19,199 cisTopic     INFO     Formatting input to corpus\n",
      "2021-09-15 11:56:25,962 gensim.corpora.dictionary INFO     adding document #0 to Dictionary(0 unique tokens: [])\n",
      "2021-09-15 11:56:58,272 gensim.corpora.dictionary INFO     adding document #10000 to Dictionary(0 unique tokens: [])\n",
      "2021-09-15 11:57:29,291 gensim.corpora.dictionary INFO     adding document #20000 to Dictionary(0 unique tokens: [])\n",
      "2021-09-15 11:58:07,062 gensim.corpora.dictionary INFO     adding document #30000 to Dictionary(0 unique tokens: [])\n",
      "2021-09-15 11:58:45,839 gensim.corpora.dictionary INFO     built Dictionary(298464 unique tokens: ['0', '1', '2', '3', '4']...) from 39580 documents (total 155187486 corpus positions)\n",
      "2021-09-15 11:58:45,840 cisTopic     INFO     Running model with 2 topics\n",
      "2021-09-15 11:58:45,849 LDAMalletWrapper INFO     Serializing temporary corpus to /scratch/leuven/325/vsc32528/tmp/mallet_cellds/merged_corpus.txt\n",
      "2021-09-15 12:05:01,700 LDAMalletWrapper INFO     Converting temporary corpus to MALLET format with mallet import-file --preserve-case --keep-sequence --remove-stopwords --token-regex \"\\S+\" --input /scratch/leuven/325/vsc32528/tmp/mallet_cellds/merged_corpus.txt --output /scratch/leuven/325/vsc32528/tmp/mallet_cellds/merged_corpus.mallet\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/bin/bash: warning: setlocale: LC_ALL: cannot change locale (en_US.UTF-8)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2021-09-15 12:06:31,003 LDAMalletWrapper INFO     Training MALLET LDA with mallet train-topics --input /scratch/leuven/325/vsc32528/tmp/mallet_cellds/merged_corpus.mallet --num-topics 2  --alpha 50 --beta 0.1 --optimize-interval 0 --num-threads 16 --output-state /scratch/leuven/325/vsc32528/tmp/mallet_cellds/merged_44b1e6_state.mallet.gz --output-doc-topics /scratch/leuven/325/vsc32528/tmp/mallet_cellds/merged_44b1e6_doctopics.txt --output-topic-keys /scratch/leuven/325/vsc32528/tmp/mallet_cellds/merged_44b1e6_topickeys.txt --num-iterations 500 --inferencer-filename /scratch/leuven/325/vsc32528/tmp/mallet_cellds/merged_44b1e6_inferencer.mallet --doc-topics-threshold 0.0  --random-seed 555\n",
      "2021-09-15 12:47:50,158 LDAMalletWrapper INFO     loading assigned topics from /scratch/leuven/325/vsc32528/tmp/mallet_cellds/merged_44b1e6_state.mallet.gz\n",
      "2021-09-15 12:53:44,245 cisTopic     INFO     Model with 2 topics done!\n",
      "2021-09-15 12:53:44,247 cisTopic     INFO     Running model with 5 topics\n",
      "2021-09-15 12:53:44,260 LDAMalletWrapper INFO     MALLET corpus already exists, training model\n",
      "2021-09-15 12:53:44,261 LDAMalletWrapper INFO     Training MALLET LDA with mallet train-topics --input /scratch/leuven/325/vsc32528/tmp/mallet_cellds/merged_corpus.mallet --num-topics 5  --alpha 50 --beta 0.1 --optimize-interval 0 --num-threads 16 --output-state /scratch/leuven/325/vsc32528/tmp/mallet_cellds/merged_efb50a_state.mallet.gz --output-doc-topics /scratch/leuven/325/vsc32528/tmp/mallet_cellds/merged_efb50a_doctopics.txt --output-topic-keys /scratch/leuven/325/vsc32528/tmp/mallet_cellds/merged_efb50a_topickeys.txt --num-iterations 500 --inferencer-filename /scratch/leuven/325/vsc32528/tmp/mallet_cellds/merged_efb50a_inferencer.mallet --doc-topics-threshold 0.0  --random-seed 555\n",
      "2021-09-15 13:49:20,620 LDAMalletWrapper INFO     loading assigned topics from /scratch/leuven/325/vsc32528/tmp/mallet_cellds/merged_efb50a_state.mallet.gz\n",
      "2021-09-15 13:56:22,223 cisTopic     INFO     Model with 5 topics done!\n",
      "2021-09-15 13:56:22,224 cisTopic     INFO     Running model with 10 topics\n",
      "2021-09-15 13:56:22,236 LDAMalletWrapper INFO     MALLET corpus already exists, training model\n",
      "2021-09-15 13:56:22,236 LDAMalletWrapper INFO     Training MALLET LDA with mallet train-topics --input /scratch/leuven/325/vsc32528/tmp/mallet_cellds/merged_corpus.mallet --num-topics 10  --alpha 50 --beta 0.1 --optimize-interval 0 --num-threads 16 --output-state /scratch/leuven/325/vsc32528/tmp/mallet_cellds/merged_e4e49b_state.mallet.gz --output-doc-topics /scratch/leuven/325/vsc32528/tmp/mallet_cellds/merged_e4e49b_doctopics.txt --output-topic-keys /scratch/leuven/325/vsc32528/tmp/mallet_cellds/merged_e4e49b_topickeys.txt --num-iterations 500 --inferencer-filename /scratch/leuven/325/vsc32528/tmp/mallet_cellds/merged_e4e49b_inferencer.mallet --doc-topics-threshold 0.0  --random-seed 555\n",
      "2021-09-15 14:46:49,874 LDAMalletWrapper INFO     loading assigned topics from /scratch/leuven/325/vsc32528/tmp/mallet_cellds/merged_e4e49b_state.mallet.gz\n",
      "2021-09-15 14:53:54,590 cisTopic     INFO     Model with 10 topics done!\n",
      "2021-09-15 14:53:54,592 cisTopic     INFO     Running model with 20 topics\n",
      "2021-09-15 14:53:54,607 LDAMalletWrapper INFO     MALLET corpus already exists, training model\n",
      "2021-09-15 14:53:54,607 LDAMalletWrapper INFO     Training MALLET LDA with mallet train-topics --input /scratch/leuven/325/vsc32528/tmp/mallet_cellds/merged_corpus.mallet --num-topics 20  --alpha 50 --beta 0.1 --optimize-interval 0 --num-threads 16 --output-state /scratch/leuven/325/vsc32528/tmp/mallet_cellds/merged_b32e24_state.mallet.gz --output-doc-topics /scratch/leuven/325/vsc32528/tmp/mallet_cellds/merged_b32e24_doctopics.txt --output-topic-keys /scratch/leuven/325/vsc32528/tmp/mallet_cellds/merged_b32e24_topickeys.txt --num-iterations 500 --inferencer-filename /scratch/leuven/325/vsc32528/tmp/mallet_cellds/merged_b32e24_inferencer.mallet --doc-topics-threshold 0.0  --random-seed 555\n",
      "2021-09-15 15:48:30,993 LDAMalletWrapper INFO     loading assigned topics from /scratch/leuven/325/vsc32528/tmp/mallet_cellds/merged_b32e24_state.mallet.gz\n",
      "2021-09-15 15:55:47,116 cisTopic     INFO     Model with 20 topics done!\n",
      "2021-09-15 15:55:47,122 cisTopic     INFO     Running model with 30 topics\n",
      "2021-09-15 15:55:47,136 LDAMalletWrapper INFO     MALLET corpus already exists, training model\n",
      "2021-09-15 15:55:47,137 LDAMalletWrapper INFO     Training MALLET LDA with mallet train-topics --input /scratch/leuven/325/vsc32528/tmp/mallet_cellds/merged_corpus.mallet --num-topics 30  --alpha 50 --beta 0.1 --optimize-interval 0 --num-threads 16 --output-state /scratch/leuven/325/vsc32528/tmp/mallet_cellds/merged_65e0fa_state.mallet.gz --output-doc-topics /scratch/leuven/325/vsc32528/tmp/mallet_cellds/merged_65e0fa_doctopics.txt --output-topic-keys /scratch/leuven/325/vsc32528/tmp/mallet_cellds/merged_65e0fa_topickeys.txt --num-iterations 500 --inferencer-filename /scratch/leuven/325/vsc32528/tmp/mallet_cellds/merged_65e0fa_inferencer.mallet --doc-topics-threshold 0.0  --random-seed 555\n",
      "2021-09-15 16:56:38,414 LDAMalletWrapper INFO     loading assigned topics from /scratch/leuven/325/vsc32528/tmp/mallet_cellds/merged_65e0fa_state.mallet.gz\n",
      "2021-09-15 17:04:01,964 cisTopic     INFO     Model with 30 topics done!\n",
      "2021-09-15 17:04:01,986 cisTopic     INFO     Running model with 35 topics\n",
      "2021-09-15 17:04:02,001 LDAMalletWrapper INFO     MALLET corpus already exists, training model\n",
      "2021-09-15 17:04:02,002 LDAMalletWrapper INFO     Training MALLET LDA with mallet train-topics --input /scratch/leuven/325/vsc32528/tmp/mallet_cellds/merged_corpus.mallet --num-topics 35  --alpha 50 --beta 0.1 --optimize-interval 0 --num-threads 16 --output-state /scratch/leuven/325/vsc32528/tmp/mallet_cellds/merged_2cce61_state.mallet.gz --output-doc-topics /scratch/leuven/325/vsc32528/tmp/mallet_cellds/merged_2cce61_doctopics.txt --output-topic-keys /scratch/leuven/325/vsc32528/tmp/mallet_cellds/merged_2cce61_topickeys.txt --num-iterations 500 --inferencer-filename /scratch/leuven/325/vsc32528/tmp/mallet_cellds/merged_2cce61_inferencer.mallet --doc-topics-threshold 0.0  --random-seed 555\n",
      "2021-09-15 18:18:51,162 LDAMalletWrapper INFO     loading assigned topics from /scratch/leuven/325/vsc32528/tmp/mallet_cellds/merged_2cce61_state.mallet.gz\n",
      "2021-09-15 18:26:11,687 cisTopic     INFO     Model with 35 topics done!\n",
      "2021-09-15 18:26:11,709 cisTopic     INFO     Running model with 40 topics\n",
      "2021-09-15 18:26:11,724 LDAMalletWrapper INFO     MALLET corpus already exists, training model\n",
      "2021-09-15 18:26:11,725 LDAMalletWrapper INFO     Training MALLET LDA with mallet train-topics --input /scratch/leuven/325/vsc32528/tmp/mallet_cellds/merged_corpus.mallet --num-topics 40  --alpha 50 --beta 0.1 --optimize-interval 0 --num-threads 16 --output-state /scratch/leuven/325/vsc32528/tmp/mallet_cellds/merged_4888e7_state.mallet.gz --output-doc-topics /scratch/leuven/325/vsc32528/tmp/mallet_cellds/merged_4888e7_doctopics.txt --output-topic-keys /scratch/leuven/325/vsc32528/tmp/mallet_cellds/merged_4888e7_topickeys.txt --num-iterations 500 --inferencer-filename /scratch/leuven/325/vsc32528/tmp/mallet_cellds/merged_4888e7_inferencer.mallet --doc-topics-threshold 0.0  --random-seed 555\n"
     ]
    }
   ],
   "source": [
    "# mallet implementation\n",
    "models_dict = {}\n",
    "for key in cistopic_obj_dict.keys():\n",
    "    f_mod = os.path.join(wdir, f_mod_dir, key + '__models_' + str(n_iter) + '_iter.pkl')\n",
    "    if os.path.isfile(f_mod):\n",
    "        print(f\"Loading {f_mod}\")\n",
    "        with open(f_mod, 'rb') as f:\n",
    "            models_dict[key] = pickle.load(f)\n",
    "    else:\n",
    "        print(f\"Running {key}\")\n",
    "        model = run_cgs_models_mallet(\n",
    "                            'mallet',\n",
    "                            cistopic_obj_dict[key],\n",
    "                            n_topics=n_topics_merged if key=='merged' else n_topics ,\n",
    "                            n_cpu=16,\n",
    "                            n_iter=n_iter,\n",
    "                            random_state=555,\n",
    "                            alpha=50,\n",
    "                            alpha_by_topic=True,\n",
    "                            eta=0.1,\n",
    "                            eta_by_topic=False,\n",
    "                            tmp_path=os.path.join(f_mod_tmpdir, key + '_')\n",
    "                            )\n",
    "        # Save\n",
    "        with open(f_mod, 'wb') as f:\n",
    "            pickle.dump(model, f)\n",
    "        print(f\"Finished {key}\")\n",
    "        models_dict[key] = model"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b85d98eb-a112-4beb-9b83-38794ab77845",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "pycisTopic singularity latest",
   "language": "python",
   "name": "pycistopic_singularity"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.6"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
