{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "37960020",
   "metadata": {},
   "source": [
    "# pycisTopic analysis\n",
    "\n",
    "Library downsampled dataset, using consensus peak regions."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "dd6b2b2a",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'0.1.dev295+gbd4bf4d.d20210830'"
      ]
     },
     "execution_count": 1,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import pycisTopic\n",
    "pycisTopic.__version__"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "765356fc",
   "metadata": {},
   "outputs": [],
   "source": [
    "import warnings\n",
    "warnings.filterwarnings('ignore')\n",
    "warnings.simplefilter('ignore')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "f4c1284d",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pickle\n",
    "import pandas as pd"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "0039a453",
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "wdir = '/staging/leuven/stg_00002/lcb/cflerin/analysis/pbmc_atac/analysis2/data_freeze_Jun2021_library_downsampled/atac_qc_multiplet_merged/jupyter/'\n",
    "os.chdir( wdir )"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "24b6a61a",
   "metadata": {},
   "source": [
    "---"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "747ca1b6",
   "metadata": {},
   "outputs": [],
   "source": [
    "fragments_dict = {\n",
    "'Broad_1':      '/staging/leuven/stg_00002/lcb/cflerin/analysis/pbmc_atac/analysis2/data_freeze_Jun2021_library_downsampled/atac_preprocess/multiplet_tagged/fragments/Broad_1.sinto.mm.fragments.tsv.gz',\n",
    "'Broad_2':      '/staging/leuven/stg_00002/lcb/cflerin/analysis/pbmc_atac/analysis2/data_freeze_Jun2021_library_downsampled/atac_preprocess/multiplet_tagged/fragments/Broad_2.sinto.mm.fragments.tsv.gz',\n",
    "## reference sample:\n",
    "'Broad_mito_1': '/staging/leuven/stg_00002/lcb/cflerin/analysis/pbmc_atac/analysis2/data_freeze_Jun2021/atac_preprocess/multiplet_tagged/fragments/Broad_mito_1.sinto.mm.fragments.tsv.gz',\n",
    "#\n",
    "'Broad_mito_2': '/staging/leuven/stg_00002/lcb/cflerin/analysis/pbmc_atac/analysis2/data_freeze_Jun2021_library_downsampled/atac_preprocess/multiplet_tagged/fragments/Broad_mito_2.sinto.mm.fragments.tsv.gz',\n",
    "'CNAG_1':       '/staging/leuven/stg_00002/lcb/cflerin/analysis/pbmc_atac/analysis2/data_freeze_Jun2021_library_downsampled/atac_preprocess/multiplet_tagged/fragments/CNAG_1.sinto.mm.fragments.tsv.gz',\n",
    "'CNAG_2':       '/staging/leuven/stg_00002/lcb/cflerin/analysis/pbmc_atac/analysis2/data_freeze_Jun2021_library_downsampled/atac_preprocess/multiplet_tagged/fragments/CNAG_2.sinto.mm.fragments.tsv.gz',\n",
    "'s3atac':       '/staging/leuven/stg_00002/lcb/cflerin/analysis/pbmc_atac/analysis2/data_freeze_Jun2021_library_downsampled/atac_preprocess/s3atac/05b_sinto_fragments/s3atac.sinto.fragments.tsv.gz',\n",
    "'Sanger_1':     '/staging/leuven/stg_00002/lcb/cflerin/analysis/pbmc_atac/analysis2/data_freeze_Jun2021_library_downsampled/atac_preprocess/multiplet_tagged/fragments/Sanger_1.sinto.mm.fragments.tsv.gz',\n",
    "'Sanger_2':     '/staging/leuven/stg_00002/lcb/cflerin/analysis/pbmc_atac/analysis2/data_freeze_Jun2021_library_downsampled/atac_preprocess/multiplet_tagged/fragments/Sanger_2.sinto.mm.fragments.tsv.gz',\n",
    "'Stanford_1':   '/staging/leuven/stg_00002/lcb/cflerin/analysis/pbmc_atac/analysis2/data_freeze_Jun2021_library_downsampled/atac_preprocess/multiplet_tagged/fragments/Stanford_1.sinto.mm.fragments.tsv.gz',\n",
    "'Stanford_2':   '/staging/leuven/stg_00002/lcb/cflerin/analysis/pbmc_atac/analysis2/data_freeze_Jun2021_library_downsampled/atac_preprocess/multiplet_tagged/fragments/Stanford_2.sinto.mm.fragments.tsv.gz',\n",
    "'VIB_1':        '/staging/leuven/stg_00002/lcb/cflerin/analysis/pbmc_atac/analysis2/data_freeze_Jun2021_library_downsampled/atac_preprocess/multiplet_tagged/fragments/VIB_1.sinto.mm.fragments.tsv.gz',\n",
    "'VIB_2':        '/staging/leuven/stg_00002/lcb/cflerin/analysis/pbmc_atac/analysis2/data_freeze_Jun2021_library_downsampled/atac_preprocess/multiplet_tagged/fragments/VIB_2.sinto.mm.fragments.tsv.gz',\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "4b9fa5d4",
   "metadata": {},
   "outputs": [],
   "source": [
    "consensus_peaks_dict = {\n",
    "'Broad_1':      '/staging/leuven/stg_00002/lcb/cflerin/analysis/pbmc_atac/analysis2/data_freeze_Jun2021_library_downsampled/atac_qc_multiplet_merged/jupyter/consensus_peak_calling/Broad_1/Broad_1__consensus_regions.bed',\n",
    "'Broad_2':      '/staging/leuven/stg_00002/lcb/cflerin/analysis/pbmc_atac/analysis2/data_freeze_Jun2021_library_downsampled/atac_qc_multiplet_merged/jupyter/consensus_peak_calling/Broad_2/Broad_2__consensus_regions.bed',\n",
    "'Broad_mito_1': '/staging/leuven/stg_00002/lcb/cflerin/analysis/pbmc_atac/analysis2/data_freeze_Jun2021_library_downsampled/atac_qc_multiplet_merged/jupyter/consensus_peak_calling/Broad_mito_1/Broad_mito_1__consensus_regions.bed',\n",
    "'Broad_mito_2': '/staging/leuven/stg_00002/lcb/cflerin/analysis/pbmc_atac/analysis2/data_freeze_Jun2021_library_downsampled/atac_qc_multiplet_merged/jupyter/consensus_peak_calling/Broad_mito_2/Broad_mito_2__consensus_regions.bed',\n",
    "'CNAG_1':       '/staging/leuven/stg_00002/lcb/cflerin/analysis/pbmc_atac/analysis2/data_freeze_Jun2021_library_downsampled/atac_qc_multiplet_merged/jupyter/consensus_peak_calling/CNAG_1/CNAG_1__consensus_regions.bed',\n",
    "'CNAG_2':       '/staging/leuven/stg_00002/lcb/cflerin/analysis/pbmc_atac/analysis2/data_freeze_Jun2021_library_downsampled/atac_qc_multiplet_merged/jupyter/consensus_peak_calling/CNAG_2/CNAG_2__consensus_regions.bed',\n",
    "'merged':       '/staging/leuven/stg_00002/lcb/cflerin/analysis/pbmc_atac/analysis2/data_freeze_Jun2021_library_downsampled/atac_qc_multiplet_merged/jupyter/consensus_peak_calling/merged/merged__consensus_regions.bed',\n",
    "'s3atac':       '/staging/leuven/stg_00002/lcb/cflerin/analysis/pbmc_atac/analysis2/data_freeze_Jun2021_library_downsampled/atac_qc_multiplet_merged/jupyter/consensus_peak_calling/s3atac/s3atac__consensus_regions.bed',\n",
    "'Sanger_1':     '/staging/leuven/stg_00002/lcb/cflerin/analysis/pbmc_atac/analysis2/data_freeze_Jun2021_library_downsampled/atac_qc_multiplet_merged/jupyter/consensus_peak_calling/Sanger_1/Sanger_1__consensus_regions.bed',\n",
    "'Sanger_2':     '/staging/leuven/stg_00002/lcb/cflerin/analysis/pbmc_atac/analysis2/data_freeze_Jun2021_library_downsampled/atac_qc_multiplet_merged/jupyter/consensus_peak_calling/Sanger_2/Sanger_2__consensus_regions.bed',\n",
    "'Stanford_1':   '/staging/leuven/stg_00002/lcb/cflerin/analysis/pbmc_atac/analysis2/data_freeze_Jun2021_library_downsampled/atac_qc_multiplet_merged/jupyter/consensus_peak_calling/Stanford_1/Stanford_1__consensus_regions.bed',\n",
    "'Stanford_2':   '/staging/leuven/stg_00002/lcb/cflerin/analysis/pbmc_atac/analysis2/data_freeze_Jun2021_library_downsampled/atac_qc_multiplet_merged/jupyter/consensus_peak_calling/Stanford_2/Stanford_2__consensus_regions.bed',\n",
    "'VIB_1':        '/staging/leuven/stg_00002/lcb/cflerin/analysis/pbmc_atac/analysis2/data_freeze_Jun2021_library_downsampled/atac_qc_multiplet_merged/jupyter/consensus_peak_calling/VIB_1/VIB_1__consensus_regions.bed',\n",
    "'VIB_2':        '/staging/leuven/stg_00002/lcb/cflerin/analysis/pbmc_atac/analysis2/data_freeze_Jun2021_library_downsampled/atac_qc_multiplet_merged/jupyter/consensus_peak_calling/VIB_2/VIB_2__consensus_regions.bed',\n",
    "'merged': '/staging/leuven/stg_00002/lcb/cflerin/analysis/pbmc_atac/analysis2/data_freeze_Jun2021_library_downsampled/atac_qc_multiplet_merged/jupyter/consensus_peak_calling/merged/merged__consensus_regions.bed', \n",
    "}"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "704a3a40",
   "metadata": {},
   "source": [
    "## Library QC metrics"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "ae703714",
   "metadata": {},
   "outputs": [],
   "source": [
    "from pycisTopic.qc import compute_qc_stats"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "cb66f9a6",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Start</th>\n",
       "      <th>Strand</th>\n",
       "      <th>Gene</th>\n",
       "      <th>Transcript_type</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>8986</th>\n",
       "      <td>chrY</td>\n",
       "      <td>9337464</td>\n",
       "      <td>1</td>\n",
       "      <td>TSPY4</td>\n",
       "      <td>protein_coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8987</th>\n",
       "      <td>chrY</td>\n",
       "      <td>9337510</td>\n",
       "      <td>1</td>\n",
       "      <td>TSPY4</td>\n",
       "      <td>protein_coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9024</th>\n",
       "      <td>chrY</td>\n",
       "      <td>22490397</td>\n",
       "      <td>1</td>\n",
       "      <td>PRY</td>\n",
       "      <td>protein_coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9100</th>\n",
       "      <td>chrY</td>\n",
       "      <td>14056227</td>\n",
       "      <td>1</td>\n",
       "      <td>VCY1B</td>\n",
       "      <td>protein_coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9136</th>\n",
       "      <td>chrY</td>\n",
       "      <td>12662368</td>\n",
       "      <td>1</td>\n",
       "      <td>USP9Y</td>\n",
       "      <td>protein_coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>236024</th>\n",
       "      <td>chr1</td>\n",
       "      <td>36479519</td>\n",
       "      <td>-1</td>\n",
       "      <td>CSF3R</td>\n",
       "      <td>protein_coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>236025</th>\n",
       "      <td>chr1</td>\n",
       "      <td>36471474</td>\n",
       "      <td>-1</td>\n",
       "      <td>CSF3R</td>\n",
       "      <td>protein_coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>236033</th>\n",
       "      <td>chr1</td>\n",
       "      <td>36482051</td>\n",
       "      <td>-1</td>\n",
       "      <td>CSF3R</td>\n",
       "      <td>protein_coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>236034</th>\n",
       "      <td>chr1</td>\n",
       "      <td>36323645</td>\n",
       "      <td>-1</td>\n",
       "      <td>EVA1B</td>\n",
       "      <td>protein_coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>236035</th>\n",
       "      <td>chr1</td>\n",
       "      <td>36324154</td>\n",
       "      <td>-1</td>\n",
       "      <td>EVA1B</td>\n",
       "      <td>protein_coding</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>77249 rows × 5 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       Chromosome     Start  Strand   Gene Transcript_type\n",
       "8986         chrY   9337464       1  TSPY4  protein_coding\n",
       "8987         chrY   9337510       1  TSPY4  protein_coding\n",
       "9024         chrY  22490397       1    PRY  protein_coding\n",
       "9100         chrY  14056227       1  VCY1B  protein_coding\n",
       "9136         chrY  12662368       1  USP9Y  protein_coding\n",
       "...           ...       ...     ...    ...             ...\n",
       "236024       chr1  36479519      -1  CSF3R  protein_coding\n",
       "236025       chr1  36471474      -1  CSF3R  protein_coding\n",
       "236033       chr1  36482051      -1  CSF3R  protein_coding\n",
       "236034       chr1  36323645      -1  EVA1B  protein_coding\n",
       "236035       chr1  36324154      -1  EVA1B  protein_coding\n",
       "\n",
       "[77249 rows x 5 columns]"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "f_biomart = '/staging/leuven/stg_00002/lcb/cflerin/analysis/pbmc_atac/analysis2/data_freeze_Jun2021/atac_qc_multiplet_merged/jupyter/biomart_annot.pickle'\n",
    "with open(f_biomart, 'rb') as f:\n",
    "    annot = pickle.load(f)\n",
    "annot"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "21392a7a",
   "metadata": {
    "collapsed": true,
    "jupyter": {
     "outputs_hidden": true
    },
    "tags": []
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "2021-09-14 17:03:26,430\tINFO services.py:1263 -- View the Ray dashboard at \u001b[1m\u001b[32mhttp://127.0.0.1:8265\u001b[39m\u001b[22m\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\u001b[2m\u001b[36m(pid=31203)\u001b[0m 2021-09-14 17:03:29,680 cisTopic     INFO     Reading Broad_2\n",
      "\u001b[2m\u001b[36m(pid=31205)\u001b[0m 2021-09-14 17:03:29,618 cisTopic     INFO     Reading s3atac\n",
      "\u001b[2m\u001b[36m(pid=31201)\u001b[0m 2021-09-14 17:03:29,740 cisTopic     INFO     Reading Broad_mito_2\n",
      "\u001b[2m\u001b[36m(pid=31200)\u001b[0m 2021-09-14 17:03:29,743 cisTopic     INFO     Reading Broad_mito_1\n",
      "\u001b[2m\u001b[36m(pid=31204)\u001b[0m 2021-09-14 17:03:29,720 cisTopic     INFO     Reading Broad_1\n",
      "\u001b[2m\u001b[36m(pid=31202)\u001b[0m 2021-09-14 17:03:29,724 cisTopic     INFO     Reading CNAG_2\n",
      "\u001b[2m\u001b[36m(pid=31199)\u001b[0m 2021-09-14 17:03:29,739 cisTopic     INFO     Reading CNAG_1\n",
      "\u001b[2m\u001b[36m(pid=31201)\u001b[0m 2021-09-14 17:04:36,855 cisTopic     INFO     Computing barcode rank plot for Broad_mito_2\n",
      "\u001b[2m\u001b[36m(pid=31201)\u001b[0m 2021-09-14 17:04:36,856 cisTopic     INFO     Counting fragments\n",
      "\u001b[2m\u001b[36m(pid=31201)\u001b[0m 2021-09-14 17:04:43,396 cisTopic     INFO     Marking barcodes with more than 100\n",
      "\u001b[2m\u001b[36m(pid=31201)\u001b[0m 2021-09-14 17:04:43,413 numexpr.utils INFO     Note: NumExpr detected 36 cores but \"NUMEXPR_MAX_THREADS\" not set, so enforcing safe limit of 8.\n",
      "\u001b[2m\u001b[36m(pid=31201)\u001b[0m 2021-09-14 17:04:43,442 cisTopic     INFO     Returning plot data\n",
      "\u001b[2m\u001b[36m(pid=31201)\u001b[0m 2021-09-14 17:04:43,442 cisTopic     INFO     Returning valid barcodes\n",
      "\u001b[2m\u001b[36m(pid=31200)\u001b[0m 2021-09-14 17:04:46,529 cisTopic     INFO     Computing barcode rank plot for Broad_mito_1\n",
      "\u001b[2m\u001b[36m(pid=31200)\u001b[0m 2021-09-14 17:04:46,530 cisTopic     INFO     Counting fragments\n",
      "\u001b[2m\u001b[36m(pid=31201)\u001b[0m 2021-09-14 17:04:47,486 cisTopic     INFO     Computing duplicate rate plot for Broad_mito_2\n",
      "\u001b[2m\u001b[36m(pid=31200)\u001b[0m 2021-09-14 17:04:54,121 cisTopic     INFO     Marking barcodes with more than 100\n",
      "\u001b[2m\u001b[36m(pid=31200)\u001b[0m 2021-09-14 17:04:54,130 numexpr.utils INFO     Note: NumExpr detected 36 cores but \"NUMEXPR_MAX_THREADS\" not set, so enforcing safe limit of 8.\n",
      "\u001b[2m\u001b[36m(pid=31201)\u001b[0m 2021-09-14 17:04:54,227 cisTopic     INFO     Return plot data\n",
      "\u001b[2m\u001b[36m(pid=31200)\u001b[0m 2021-09-14 17:04:54,161 cisTopic     INFO     Returning plot data\n",
      "\u001b[2m\u001b[36m(pid=31200)\u001b[0m 2021-09-14 17:04:54,162 cisTopic     INFO     Returning valid barcodes\n",
      "\u001b[2m\u001b[36m(pid=31201)\u001b[0m 2021-09-14 17:04:54,352 cisTopic     INFO     Computing insert size distribution for Broad_mito_2\n",
      "\u001b[2m\u001b[36m(pid=31201)\u001b[0m 2021-09-14 17:04:54,353 cisTopic     INFO     Counting fragments\n",
      "\u001b[2m\u001b[36m(pid=31201)\u001b[0m 2021-09-14 17:04:56,235 cisTopic     INFO     Returning plot data\n",
      "\u001b[2m\u001b[36m(pid=31202)\u001b[0m 2021-09-14 17:04:58,911 cisTopic     INFO     Computing barcode rank plot for CNAG_2\n",
      "\u001b[2m\u001b[36m(pid=31202)\u001b[0m 2021-09-14 17:04:58,911 cisTopic     INFO     Counting fragments\n",
      "\u001b[2m\u001b[36m(pid=31200)\u001b[0m 2021-09-14 17:04:59,513 cisTopic     INFO     Computing duplicate rate plot for Broad_mito_1\n",
      "\u001b[2m\u001b[36m(pid=31199)\u001b[0m 2021-09-14 17:05:00,487 cisTopic     INFO     Computing barcode rank plot for CNAG_1\n",
      "\u001b[2m\u001b[36m(pid=31199)\u001b[0m 2021-09-14 17:05:00,488 cisTopic     INFO     Counting fragments\n",
      "\u001b[2m\u001b[36m(pid=31202)\u001b[0m 2021-09-14 17:05:06,443 cisTopic     INFO     Marking barcodes with more than 100\n",
      "\u001b[2m\u001b[36m(pid=31202)\u001b[0m 2021-09-14 17:05:06,448 numexpr.utils INFO     Note: NumExpr detected 36 cores but \"NUMEXPR_MAX_THREADS\" not set, so enforcing safe limit of 8.\n",
      "\u001b[2m\u001b[36m(pid=31202)\u001b[0m 2021-09-14 17:05:06,481 cisTopic     INFO     Returning plot data\n",
      "\u001b[2m\u001b[36m(pid=31202)\u001b[0m 2021-09-14 17:05:06,487 cisTopic     INFO     Returning valid barcodes\n",
      "\u001b[2m\u001b[36m(pid=31199)\u001b[0m 2021-09-14 17:05:08,103 cisTopic     INFO     Marking barcodes with more than 100\n",
      "\u001b[2m\u001b[36m(pid=31199)\u001b[0m 2021-09-14 17:05:08,107 numexpr.utils INFO     Note: NumExpr detected 36 cores but \"NUMEXPR_MAX_THREADS\" not set, so enforcing safe limit of 8.\n",
      "\u001b[2m\u001b[36m(pid=31199)\u001b[0m 2021-09-14 17:05:08,141 cisTopic     INFO     Returning plot data\n",
      "\u001b[2m\u001b[36m(pid=31199)\u001b[0m 2021-09-14 17:05:08,148 cisTopic     INFO     Returning valid barcodes\n",
      "\u001b[2m\u001b[36m(pid=31200)\u001b[0m 2021-09-14 17:05:08,252 cisTopic     INFO     Return plot data\n",
      "\u001b[2m\u001b[36m(pid=31200)\u001b[0m 2021-09-14 17:05:08,446 cisTopic     INFO     Computing insert size distribution for Broad_mito_1\n",
      "\u001b[2m\u001b[36m(pid=31200)\u001b[0m 2021-09-14 17:05:08,446 cisTopic     INFO     Counting fragments\n",
      "\u001b[2m\u001b[36m(pid=31200)\u001b[0m 2021-09-14 17:05:10,829 cisTopic     INFO     Returning plot data\n",
      "\u001b[2m\u001b[36m(pid=31202)\u001b[0m 2021-09-14 17:05:11,796 cisTopic     INFO     Computing duplicate rate plot for CNAG_2\n",
      "\u001b[2m\u001b[36m(pid=31199)\u001b[0m 2021-09-14 17:05:13,370 cisTopic     INFO     Computing duplicate rate plot for CNAG_1\n",
      "\u001b[2m\u001b[36m(pid=31201)\u001b[0m 2021-09-14 17:05:16,174 cisTopic     INFO     Computing TSS profile for Broad_mito_2\n",
      "\u001b[2m\u001b[36m(pid=31202)\u001b[0m 2021-09-14 17:05:20,767 cisTopic     INFO     Return plot data\n",
      "\u001b[2m\u001b[36m(pid=31202)\u001b[0m 2021-09-14 17:05:20,865 cisTopic     INFO     Computing insert size distribution for CNAG_2\n",
      "\u001b[2m\u001b[36m(pid=31202)\u001b[0m 2021-09-14 17:05:20,865 cisTopic     INFO     Counting fragments\n",
      "\u001b[2m\u001b[36m(pid=31202)\u001b[0m 2021-09-14 17:05:22,256 cisTopic     INFO     Returning plot data\n",
      "\u001b[2m\u001b[36m(pid=31199)\u001b[0m 2021-09-14 17:05:22,302 cisTopic     INFO     Return plot data\n",
      "\u001b[2m\u001b[36m(pid=31199)\u001b[0m 2021-09-14 17:05:22,433 cisTopic     INFO     Computing insert size distribution for CNAG_1\n",
      "\u001b[2m\u001b[36m(pid=31199)\u001b[0m 2021-09-14 17:05:22,434 cisTopic     INFO     Counting fragments\n",
      "\u001b[2m\u001b[36m(pid=31199)\u001b[0m 2021-09-14 17:05:23,868 cisTopic     INFO     Returning plot data\n",
      "\u001b[2m\u001b[36m(pid=31201)\u001b[0m 2021-09-14 17:05:29,369 cisTopic     INFO     Formatting annnotation\n",
      "\u001b[2m\u001b[36m(pid=31201)\u001b[0m 2021-09-14 17:05:29,464 cisTopic     INFO     Creating coverage matrix\n",
      "\u001b[2m\u001b[36m(pid=31205)\u001b[0m 2021-09-14 17:05:30,429 cisTopic     INFO     Computing barcode rank plot for s3atac\n",
      "\u001b[2m\u001b[36m(pid=31205)\u001b[0m 2021-09-14 17:05:30,429 cisTopic     INFO     Counting fragments\n",
      "\u001b[2m\u001b[36m(pid=31205)\u001b[0m 2021-09-14 17:05:36,806 cisTopic     INFO     Marking barcodes with more than 100\n",
      "\u001b[2m\u001b[36m(pid=31205)\u001b[0m 2021-09-14 17:05:36,810 numexpr.utils INFO     Note: NumExpr detected 36 cores but \"NUMEXPR_MAX_THREADS\" not set, so enforcing safe limit of 8.\n",
      "\u001b[2m\u001b[36m(pid=31205)\u001b[0m 2021-09-14 17:05:36,815 cisTopic     INFO     Returning plot data\n",
      "\u001b[2m\u001b[36m(pid=31205)\u001b[0m 2021-09-14 17:05:36,816 cisTopic     INFO     Returning valid barcodes\n",
      "\u001b[2m\u001b[36m(pid=31200)\u001b[0m 2021-09-14 17:05:39,240 cisTopic     INFO     Computing TSS profile for Broad_mito_1\n",
      "\u001b[2m\u001b[36m(pid=31202)\u001b[0m 2021-09-14 17:05:43,814 cisTopic     INFO     Computing TSS profile for CNAG_2\n",
      "\u001b[2m\u001b[36m(pid=31205)\u001b[0m 2021-09-14 17:05:44,613 cisTopic     INFO     Computing duplicate rate plot for s3atac\n",
      "\u001b[2m\u001b[36m(pid=31199)\u001b[0m 2021-09-14 17:05:47,463 cisTopic     INFO     Computing TSS profile for CNAG_1\n",
      "\u001b[2m\u001b[36m(pid=31205)\u001b[0m 2021-09-14 17:05:50,867 cisTopic     INFO     Return plot data\n",
      "\u001b[2m\u001b[36m(pid=31205)\u001b[0m 2021-09-14 17:05:50,949 cisTopic     INFO     Computing insert size distribution for s3atac\n",
      "\u001b[2m\u001b[36m(pid=31205)\u001b[0m 2021-09-14 17:05:50,949 cisTopic     INFO     Counting fragments\n",
      "\u001b[2m\u001b[36m(pid=31200)\u001b[0m 2021-09-14 17:05:54,276 cisTopic     INFO     Formatting annnotation\n",
      "\u001b[2m\u001b[36m(pid=31205)\u001b[0m 2021-09-14 17:05:54,384 cisTopic     INFO     Returning plot data\n",
      "\u001b[2m\u001b[36m(pid=31200)\u001b[0m 2021-09-14 17:05:54,367 cisTopic     INFO     Creating coverage matrix\n",
      "\u001b[2m\u001b[36m(pid=31202)\u001b[0m 2021-09-14 17:06:01,247 cisTopic     INFO     Formatting annnotation\n",
      "\u001b[2m\u001b[36m(pid=31202)\u001b[0m 2021-09-14 17:06:01,356 cisTopic     INFO     Creating coverage matrix\n",
      "\u001b[2m\u001b[36m(pid=31199)\u001b[0m 2021-09-14 17:06:04,784 cisTopic     INFO     Formatting annnotation\n",
      "\u001b[2m\u001b[36m(pid=31199)\u001b[0m 2021-09-14 17:06:04,877 cisTopic     INFO     Creating coverage matrix\n",
      "\u001b[2m\u001b[36m(pid=31205)\u001b[0m 2021-09-14 17:06:33,577 cisTopic     INFO     Computing TSS profile for s3atac\n",
      "\u001b[2m\u001b[36m(pid=31203)\u001b[0m 2021-09-14 17:06:38,598 cisTopic     INFO     Computing barcode rank plot for Broad_2\n",
      "\u001b[2m\u001b[36m(pid=31203)\u001b[0m 2021-09-14 17:06:38,598 cisTopic     INFO     Counting fragments\n",
      "\u001b[2m\u001b[36m(pid=31205)\u001b[0m 2021-09-14 17:06:41,578 cisTopic     INFO     Formatting annnotation\n",
      "\u001b[2m\u001b[36m(pid=31205)\u001b[0m 2021-09-14 17:06:41,663 cisTopic     INFO     Creating coverage matrix\n",
      "\u001b[2m\u001b[36m(pid=31204)\u001b[0m 2021-09-14 17:06:50,273 cisTopic     INFO     Computing barcode rank plot for Broad_1\n",
      "\u001b[2m\u001b[36m(pid=31204)\u001b[0m 2021-09-14 17:06:50,274 cisTopic     INFO     Counting fragments\n"
     ]
    }
   ],
   "source": [
    "metadata_bc_dict, profile_data_dict = compute_qc_stats(\n",
    "        fragments_dict=fragments_dict,\n",
    "        tss_annotation=annot,\n",
    "        stats=['barcode_rank_plot', 'duplicate_rate', 'insert_size_distribution', 'profile_tss', 'frip'],\n",
    "        label_list=None,\n",
    "        path_to_regions=consensus_peaks_dict,\n",
    "        n_cpu=7,\n",
    "        valid_bc=None,\n",
    "        n_frag=100,\n",
    "        n_bc=None,\n",
    "        tss_flank_window=2000,\n",
    "        tss_window=50,\n",
    "        tss_minimum_signal_window=100,\n",
    "        tss_rolling_window=10,\n",
    "        min_norm=0.1,\n",
    "        remove_duplicates = True,\n",
    "        )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "206c9c74",
   "metadata": {},
   "outputs": [],
   "source": [
    "f_qc_dir = 'qc__consensus_peaks'\n",
    "if not os.path.exists(os.path.join(wdir, f_qc_dir)):\n",
    "    os.makedirs(os.path.join(wdir, f_qc_dir))\n",
    "    \n",
    "with open(os.path.join(wdir, f_qc_dir, 'metadata.pickle'), 'wb') as f:\n",
    "    pickle.dump(metadata_bc_dict, f)\n",
    "\n",
    "with open(os.path.join(wdir, f_qc_dir, 'profile_data.pickle'), 'wb') as f:\n",
    "    pickle.dump(profile_data_dict, f)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c39a6213",
   "metadata": {},
   "source": [
    "## Create the cisTopic objects for each sample"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "18d3ad05",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Valid barcodes\n",
    "with open(wdir + 'barcodes_passing_filters2.pkl', 'rb') as f:\n",
    "    bc_passing_filters = pickle.load(f)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "c0f53ef5",
   "metadata": {},
   "outputs": [],
   "source": [
    "from pycisTopic.cistopic_class import create_cistopic_object_from_fragments"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "5f6b84a0",
   "metadata": {},
   "outputs": [],
   "source": [
    "#path_to_regions = '/staging/leuven/stg_00002/lcb/cbravo/SCREEN_ENCODE3/V2.hg38-rDHS-Unfiltered.blacklisted.bed'\n",
    "path_to_blacklist = '/staging/leuven/stg_00002/lcb/cflerin/testruns/pycisTopic/pycisTopic/blacklist/hg38-blacklist.v2.bed'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "7397c5de",
   "metadata": {},
   "outputs": [],
   "source": [
    "f_cto_dir = 'cistopic_objs__consensus_peaks'\n",
    "if not os.path.exists(os.path.join(wdir, f_cto_dir)):\n",
    "    os.makedirs(os.path.join(wdir, f_cto_dir))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "9a47dde0",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Skipping Broad_1\n",
      "Skipping Broad_2\n",
      "Skipping Broad_mito_1\n",
      "Skipping Broad_mito_2\n",
      "Skipping CNAG_1\n",
      "Skipping CNAG_2\n",
      "Skipping s3atac\n",
      "Skipping Sanger_1\n",
      "Skipping Sanger_2\n",
      "Skipping Stanford_1\n",
      "Skipping Stanford_2\n",
      "Skipping VIB_1\n",
      "Skipping VIB_2\n"
     ]
    }
   ],
   "source": [
    "#Create all objects\n",
    "\n",
    "for key in bc_passing_filters.keys():\n",
    "    f_out = os.path.join(wdir, f_cto_dir, key + '__cistopic_obj.pkl')\n",
    "    if os.path.isfile(f_out):\n",
    "        print(f\"Skipping {key}\")\n",
    "        continue\n",
    "        \n",
    "    tmp_cto = create_cistopic_object_from_fragments(path_to_fragments=fragments_dict[key],\n",
    "                                                    path_to_regions=consensus_peaks_dict[key],\n",
    "                                                    path_to_blacklist=path_to_blacklist,\n",
    "                                                    #metrics=metadata_bc_dict[key],\n",
    "                                                    valid_bc=bc_passing_filters[key],\n",
    "                                                    n_cpu=1,\n",
    "                                                    partition=20,\n",
    "                                                    project=key)\n",
    "    \n",
    "    with open(f_out, 'wb') as f:\n",
    "        pickle.dump(tmp_cto, f)\n",
    "    \n",
    "    print(f\"DONE {key}\")\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f37a3c64",
   "metadata": {},
   "source": [
    "## Create a merged object\n",
    "\n",
    "Here, use a common set of peaks for each sample, then merge them"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "c7e28925",
   "metadata": {},
   "outputs": [],
   "source": [
    "f_cto_merged_dir = 'cistopic_objs__consensus_peaks_merged'\n",
    "if not os.path.exists(os.path.join(wdir, f_cto_merged_dir)):\n",
    "    os.makedirs(os.path.join(wdir, f_cto_merged_dir))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "3f4eda27",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Skipping Broad_1\n",
      "Skipping Broad_2\n",
      "Skipping Broad_mito_1\n",
      "Skipping Broad_mito_2\n",
      "Skipping CNAG_1\n",
      "Skipping CNAG_2\n",
      "Skipping s3atac\n",
      "Skipping Sanger_1\n",
      "Skipping Sanger_2\n",
      "Skipping Stanford_1\n",
      "Skipping Stanford_2\n",
      "Skipping VIB_1\n",
      "Skipping VIB_2\n"
     ]
    }
   ],
   "source": [
    "#Create all objects\n",
    "\n",
    "for key in bc_passing_filters.keys():\n",
    "    f_out = os.path.join(wdir, f_cto_merged_dir, key + '__cistopic_obj.pkl')\n",
    "    if os.path.isfile(f_out):\n",
    "        print(f\"Skipping {key}\")\n",
    "        continue\n",
    "        \n",
    "    tmp_cto = create_cistopic_object_from_fragments(path_to_fragments=fragments_dict[key],\n",
    "                                                    path_to_regions=consensus_peaks_dict['merged'],\n",
    "                                                    path_to_blacklist=path_to_blacklist,\n",
    "                                                    #metrics=metadata_bc_dict[key],\n",
    "                                                    valid_bc=bc_passing_filters[key],\n",
    "                                                    n_cpu=1,\n",
    "                                                    partition=20,\n",
    "                                                    project=key)\n",
    "    \n",
    "    with open(f_out, 'wb') as f:\n",
    "        pickle.dump(tmp_cto, f)\n",
    "    \n",
    "    print(f\"DONE {key}\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "16d059ed",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Loaded Broad_1\n",
      "Loaded Broad_2\n",
      "Loaded Broad_mito_1\n",
      "Loaded Broad_mito_2\n",
      "Loaded CNAG_1\n",
      "Loaded CNAG_2\n",
      "Loaded s3atac\n",
      "Loaded Sanger_1\n",
      "Loaded Sanger_2\n",
      "Loaded Stanford_1\n",
      "Loaded Stanford_2\n",
      "Loaded VIB_1\n",
      "Loaded VIB_2\n"
     ]
    }
   ],
   "source": [
    "# load objects into dict:\n",
    "cistopic_obj_dict = {}\n",
    "for key in bc_passing_filters.keys():\n",
    "    f_out = os.path.join(wdir, f_cto_merged_dir, key + '__cistopic_obj.pkl')\n",
    "    with open(f_out, 'rb') as f:\n",
    "        cistopic_obj_dict[key] = pickle.load(f)\n",
    "    print(f\"Loaded {key}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "ce8595eb",
   "metadata": {},
   "outputs": [],
   "source": [
    "import copy"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "f357f00d",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2021-09-15 11:35:23,309 cisTopic     INFO     cisTopic object 1 merged\n",
      "2021-09-15 11:35:31,486 cisTopic     INFO     cisTopic object 2 merged\n",
      "2021-09-15 11:35:41,386 cisTopic     INFO     cisTopic object 3 merged\n",
      "2021-09-15 11:35:52,168 cisTopic     INFO     cisTopic object 4 merged\n",
      "2021-09-15 11:36:03,680 cisTopic     INFO     cisTopic object 5 merged\n",
      "2021-09-15 11:36:16,789 cisTopic     INFO     cisTopic object 6 merged\n",
      "2021-09-15 11:36:31,420 cisTopic     INFO     cisTopic object 7 merged\n",
      "2021-09-15 11:36:48,298 cisTopic     INFO     cisTopic object 8 merged\n",
      "2021-09-15 11:37:05,575 cisTopic     INFO     cisTopic object 9 merged\n",
      "2021-09-15 11:37:23,415 cisTopic     INFO     cisTopic object 10 merged\n",
      "2021-09-15 11:37:42,651 cisTopic     INFO     cisTopic object 11 merged\n",
      "2021-09-15 11:38:05,534 cisTopic     INFO     cisTopic object 12 merged\n"
     ]
    }
   ],
   "source": [
    "f_out = os.path.join(wdir, f_cto_merged_dir, 'libds_merged__cistopic_obj.pkl')\n",
    "if os.path.isfile(f_out):\n",
    "    print(f\"Loading {f_out}\")\n",
    "    with open(f_out, 'rb') as f:\n",
    "        cistopic_obj_dict['merged'] = pickle.load(f)\n",
    "else:\n",
    "    # merge:\n",
    "    ct_keys = list(cistopic_obj_dict.keys())\n",
    "    cto_merged = copy.copy(cistopic_obj_dict[ct_keys[0]])\n",
    "    cto_merged.merge([ cistopic_obj_dict[k] for k in ct_keys[1:] ], project='scATAC-seq_benchmark')\n",
    "    with open(f_out, 'wb') as f:\n",
    "        pickle.dump(cto_merged, f)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "23f838e2",
   "metadata": {},
   "source": [
    "## Create a set of cisTopic objects for downstream analysis\n",
    "\n",
    "Include each sample individually (using the sample-specific consensus peaks), plus the merged sample (using the merged consensus peaks)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "3f5658eb",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Loaded Broad_1\n",
      "Loaded Broad_2\n",
      "Loaded Broad_mito_1\n",
      "Loaded Broad_mito_2\n",
      "Loaded CNAG_1\n",
      "Loaded CNAG_2\n",
      "Loaded s3atac\n",
      "Loaded Sanger_1\n",
      "Loaded Sanger_2\n",
      "Loaded Stanford_1\n",
      "Loaded Stanford_2\n",
      "Loaded VIB_1\n",
      "Loaded VIB_2\n",
      "Loading /staging/leuven/stg_00002/lcb/cflerin/analysis/pbmc_atac/analysis2/data_freeze_Jun2021_library_downsampled/atac_qc_multiplet_merged/jupyter/cistopic_objs__consensus_peaks_merged/libds_merged__cistopic_obj.pkl\n"
     ]
    }
   ],
   "source": [
    "del cistopic_obj_dict\n",
    "\n",
    "cistopic_obj_dict = {}\n",
    "# load sample objects into dict:\n",
    "for key in bc_passing_filters.keys():\n",
    "    f_out = os.path.join(wdir, f_cto_dir, key + '__cistopic_obj.pkl')\n",
    "    with open(f_out, 'rb') as f:\n",
    "        cistopic_obj_dict[key] = pickle.load(f)\n",
    "    print(f\"Loaded {key}\")\n",
    "\n",
    "# add the merged object:\n",
    "f_out = os.path.join(wdir, f_cto_merged_dir, 'libds_merged__cistopic_obj.pkl')\n",
    "if os.path.isfile(f_out):\n",
    "    print(f\"Loading {f_out}\")\n",
    "    with open(f_out, 'rb') as f:\n",
    "        cistopic_obj_dict['merged'] = pickle.load(f)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0063351b",
   "metadata": {},
   "source": [
    "## Add cell annotations"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "e6ed71c8",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Dupl_rate</th>\n",
       "      <th>FRIP</th>\n",
       "      <th>Predicted_doublets_fragments</th>\n",
       "      <th>cisTopic_nr_frag</th>\n",
       "      <th>pycisTopic_leiden_10_0.6</th>\n",
       "      <th>cisTopic_log_nr_frag</th>\n",
       "      <th>sample_id</th>\n",
       "      <th>Log_unique_nr_frag</th>\n",
       "      <th>Dupl_nr_frag</th>\n",
       "      <th>Total_nr_frag_in_regions</th>\n",
       "      <th>...</th>\n",
       "      <th>barcode</th>\n",
       "      <th>fmx_droplet_type</th>\n",
       "      <th>cisTopic_log_nr_acc</th>\n",
       "      <th>pycisTopic_leiden_10_0.8</th>\n",
       "      <th>fmx_sample</th>\n",
       "      <th>Log_total_nr_frag</th>\n",
       "      <th>cisTopic_nr_acc</th>\n",
       "      <th>pycisTopic_leiden_10_1.0</th>\n",
       "      <th>pycisTopic_leiden_10_1.2</th>\n",
       "      <th>consensus_cell_type</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>GAGGACATCTGAACTCCTTAA_TGGACCAGTAATACACCTACC_CTCTTGCCAGGCGGGTCGGTT_GCGACTCTCGGATGGGTAACA-Broad_1</th>\n",
       "      <td>0.679883</td>\n",
       "      <td>0.545198</td>\n",
       "      <td>False</td>\n",
       "      <td>10834</td>\n",
       "      <td>1</td>\n",
       "      <td>4.034789</td>\n",
       "      <td>Broad_1</td>\n",
       "      <td>4.140068</td>\n",
       "      <td>29322</td>\n",
       "      <td>24899</td>\n",
       "      <td>...</td>\n",
       "      <td>GAGGACATCTGAACTCCTTAA_TGGACCAGTAATACACCTACC_CT...</td>\n",
       "      <td>SNG</td>\n",
       "      <td>3.992067</td>\n",
       "      <td>0</td>\n",
       "      <td>sampleB</td>\n",
       "      <td>4.634759</td>\n",
       "      <td>9819</td>\n",
       "      <td>0</td>\n",
       "      <td>15</td>\n",
       "      <td>CD4+ T cell</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CGGATCTCGGCCAGGGCGGAA_TTTGGTCGATCACCTCATACA_CGGAGGAGTAATACAATGGCC-Broad_1</th>\n",
       "      <td>0.663091</td>\n",
       "      <td>0.783794</td>\n",
       "      <td>False</td>\n",
       "      <td>14588</td>\n",
       "      <td>10</td>\n",
       "      <td>4.163996</td>\n",
       "      <td>Broad_1</td>\n",
       "      <td>4.131555</td>\n",
       "      <td>26645</td>\n",
       "      <td>32467</td>\n",
       "      <td>...</td>\n",
       "      <td>CGGATCTCGGCCAGGGCGGAA_TTTGGTCGATCACCTCATACA_CG...</td>\n",
       "      <td>SNG</td>\n",
       "      <td>4.120508</td>\n",
       "      <td>8</td>\n",
       "      <td>sampleA</td>\n",
       "      <td>4.604042</td>\n",
       "      <td>13198</td>\n",
       "      <td>7</td>\n",
       "      <td>6</td>\n",
       "      <td>CD14+ monocyte</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CTTGCGACGACACTATAGGCA_TCTTGTAGTTGATGCATACGC_AAGTTATCCTCCTTACTACGA-Broad_1</th>\n",
       "      <td>0.675366</td>\n",
       "      <td>0.803886</td>\n",
       "      <td>False</td>\n",
       "      <td>7822</td>\n",
       "      <td>10</td>\n",
       "      <td>3.893318</td>\n",
       "      <td>Broad_1</td>\n",
       "      <td>3.915664</td>\n",
       "      <td>17132</td>\n",
       "      <td>21308</td>\n",
       "      <td>...</td>\n",
       "      <td>CTTGCGACGACACTATAGGCA_TCTTGTAGTTGATGCATACGC_AA...</td>\n",
       "      <td>SNG</td>\n",
       "      <td>3.853211</td>\n",
       "      <td>8</td>\n",
       "      <td>sampleB</td>\n",
       "      <td>4.404269</td>\n",
       "      <td>7132</td>\n",
       "      <td>7</td>\n",
       "      <td>6</td>\n",
       "      <td>CD14+ monocyte</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TACCGAATGACCGCATTCGTT_CTCATTTACTCAATAGCAACG_TGTCGCTCGATTACCACATGA-Broad_1</th>\n",
       "      <td>0.697144</td>\n",
       "      <td>0.770361</td>\n",
       "      <td>False</td>\n",
       "      <td>7890</td>\n",
       "      <td>1</td>\n",
       "      <td>3.897077</td>\n",
       "      <td>Broad_1</td>\n",
       "      <td>3.882240</td>\n",
       "      <td>17552</td>\n",
       "      <td>20263</td>\n",
       "      <td>...</td>\n",
       "      <td>TACCGAATGACCGCATTCGTT_CTCATTTACTCAATAGCAACG_TG...</td>\n",
       "      <td>SNG</td>\n",
       "      <td>3.845656</td>\n",
       "      <td>0</td>\n",
       "      <td>sampleB</td>\n",
       "      <td>4.401004</td>\n",
       "      <td>7009</td>\n",
       "      <td>0</td>\n",
       "      <td>5</td>\n",
       "      <td>CD4+ T cell</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GAACCGTGGCGTCCTCCGCTC_TGAATCCCCATAATTGGCAAC_TCCGCTCTTGTAAGCAGGCGG-Broad_1</th>\n",
       "      <td>0.663630</td>\n",
       "      <td>0.844259</td>\n",
       "      <td>False</td>\n",
       "      <td>6975</td>\n",
       "      <td>1</td>\n",
       "      <td>3.843544</td>\n",
       "      <td>Broad_1</td>\n",
       "      <td>3.841422</td>\n",
       "      <td>13694</td>\n",
       "      <td>17960</td>\n",
       "      <td>...</td>\n",
       "      <td>GAACCGTGGCGTCCTCCGCTC_TGAATCCCCATAATTGGCAAC_TC...</td>\n",
       "      <td>SNG</td>\n",
       "      <td>3.800992</td>\n",
       "      <td>0</td>\n",
       "      <td>sampleA</td>\n",
       "      <td>4.314604</td>\n",
       "      <td>6324</td>\n",
       "      <td>0</td>\n",
       "      <td>15</td>\n",
       "      <td>CD4+ T cell</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CGCACAGAGGAGAACA-VIB_2</th>\n",
       "      <td>0.364983</td>\n",
       "      <td>0.214522</td>\n",
       "      <td>False</td>\n",
       "      <td>2487</td>\n",
       "      <td>0</td>\n",
       "      <td>3.395676</td>\n",
       "      <td>VIB_2</td>\n",
       "      <td>3.781468</td>\n",
       "      <td>3475</td>\n",
       "      <td>2044</td>\n",
       "      <td>...</td>\n",
       "      <td>CGCACAGAGGAGAACA</td>\n",
       "      <td>SNG</td>\n",
       "      <td>3.361161</td>\n",
       "      <td>1</td>\n",
       "      <td>sampleB</td>\n",
       "      <td>3.978683</td>\n",
       "      <td>2297</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>CD14+ monocyte</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ACCAAACGTTCTTATC-VIB_2</th>\n",
       "      <td>0.111639</td>\n",
       "      <td>0.260886</td>\n",
       "      <td>False</td>\n",
       "      <td>2532</td>\n",
       "      <td>12</td>\n",
       "      <td>3.403464</td>\n",
       "      <td>VIB_2</td>\n",
       "      <td>3.719000</td>\n",
       "      <td>658</td>\n",
       "      <td>1647</td>\n",
       "      <td>...</td>\n",
       "      <td>ACCAAACGTTCTTATC</td>\n",
       "      <td>SNG</td>\n",
       "      <td>3.397419</td>\n",
       "      <td>11</td>\n",
       "      <td>sampleB</td>\n",
       "      <td>3.770410</td>\n",
       "      <td>2497</td>\n",
       "      <td>9</td>\n",
       "      <td>8</td>\n",
       "      <td>CD14+ monocyte</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GACCCAGAGATATGAC-VIB_2</th>\n",
       "      <td>0.223323</td>\n",
       "      <td>0.131992</td>\n",
       "      <td>False</td>\n",
       "      <td>1369</td>\n",
       "      <td>12</td>\n",
       "      <td>3.136403</td>\n",
       "      <td>VIB_2</td>\n",
       "      <td>3.610234</td>\n",
       "      <td>1172</td>\n",
       "      <td>769</td>\n",
       "      <td>...</td>\n",
       "      <td>GACCCAGAGATATGAC</td>\n",
       "      <td>SNG</td>\n",
       "      <td>3.116940</td>\n",
       "      <td>11</td>\n",
       "      <td>sampleA</td>\n",
       "      <td>3.719994</td>\n",
       "      <td>1309</td>\n",
       "      <td>9</td>\n",
       "      <td>8</td>\n",
       "      <td>CD14+ monocyte</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GATGATCCAACTCCCT-VIB_2</th>\n",
       "      <td>0.103598</td>\n",
       "      <td>0.165140</td>\n",
       "      <td>False</td>\n",
       "      <td>1314</td>\n",
       "      <td>7</td>\n",
       "      <td>3.118595</td>\n",
       "      <td>VIB_2</td>\n",
       "      <td>3.569608</td>\n",
       "      <td>429</td>\n",
       "      <td>810</td>\n",
       "      <td>...</td>\n",
       "      <td>GATGATCCAACTCCCT</td>\n",
       "      <td>SNG</td>\n",
       "      <td>3.116276</td>\n",
       "      <td>9</td>\n",
       "      <td>sampleB</td>\n",
       "      <td>3.617105</td>\n",
       "      <td>1307</td>\n",
       "      <td>8</td>\n",
       "      <td>7</td>\n",
       "      <td>Cytotoxic T cell</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TTACTTGCATACAACC-VIB_2</th>\n",
       "      <td>0.189720</td>\n",
       "      <td>0.136021</td>\n",
       "      <td>False</td>\n",
       "      <td>2082</td>\n",
       "      <td>12</td>\n",
       "      <td>3.318481</td>\n",
       "      <td>VIB_2</td>\n",
       "      <td>3.794279</td>\n",
       "      <td>1458</td>\n",
       "      <td>1206</td>\n",
       "      <td>...</td>\n",
       "      <td>TTACTTGCATACAACC</td>\n",
       "      <td>SNG</td>\n",
       "      <td>3.292699</td>\n",
       "      <td>11</td>\n",
       "      <td>sampleA</td>\n",
       "      <td>3.885644</td>\n",
       "      <td>1962</td>\n",
       "      <td>9</td>\n",
       "      <td>8</td>\n",
       "      <td>Cytotoxic T cell</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>40496 rows × 27 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                    Dupl_rate      FRIP  \\\n",
       "GAGGACATCTGAACTCCTTAA_TGGACCAGTAATACACCTACC_CTC...   0.679883  0.545198   \n",
       "CGGATCTCGGCCAGGGCGGAA_TTTGGTCGATCACCTCATACA_CGG...   0.663091  0.783794   \n",
       "CTTGCGACGACACTATAGGCA_TCTTGTAGTTGATGCATACGC_AAG...   0.675366  0.803886   \n",
       "TACCGAATGACCGCATTCGTT_CTCATTTACTCAATAGCAACG_TGT...   0.697144  0.770361   \n",
       "GAACCGTGGCGTCCTCCGCTC_TGAATCCCCATAATTGGCAAC_TCC...   0.663630  0.844259   \n",
       "...                                                       ...       ...   \n",
       "CGCACAGAGGAGAACA-VIB_2                               0.364983  0.214522   \n",
       "ACCAAACGTTCTTATC-VIB_2                               0.111639  0.260886   \n",
       "GACCCAGAGATATGAC-VIB_2                               0.223323  0.131992   \n",
       "GATGATCCAACTCCCT-VIB_2                               0.103598  0.165140   \n",
       "TTACTTGCATACAACC-VIB_2                               0.189720  0.136021   \n",
       "\n",
       "                                                    Predicted_doublets_fragments  \\\n",
       "GAGGACATCTGAACTCCTTAA_TGGACCAGTAATACACCTACC_CTC...                         False   \n",
       "CGGATCTCGGCCAGGGCGGAA_TTTGGTCGATCACCTCATACA_CGG...                         False   \n",
       "CTTGCGACGACACTATAGGCA_TCTTGTAGTTGATGCATACGC_AAG...                         False   \n",
       "TACCGAATGACCGCATTCGTT_CTCATTTACTCAATAGCAACG_TGT...                         False   \n",
       "GAACCGTGGCGTCCTCCGCTC_TGAATCCCCATAATTGGCAAC_TCC...                         False   \n",
       "...                                                                          ...   \n",
       "CGCACAGAGGAGAACA-VIB_2                                                     False   \n",
       "ACCAAACGTTCTTATC-VIB_2                                                     False   \n",
       "GACCCAGAGATATGAC-VIB_2                                                     False   \n",
       "GATGATCCAACTCCCT-VIB_2                                                     False   \n",
       "TTACTTGCATACAACC-VIB_2                                                     False   \n",
       "\n",
       "                                                    cisTopic_nr_frag  \\\n",
       "GAGGACATCTGAACTCCTTAA_TGGACCAGTAATACACCTACC_CTC...             10834   \n",
       "CGGATCTCGGCCAGGGCGGAA_TTTGGTCGATCACCTCATACA_CGG...             14588   \n",
       "CTTGCGACGACACTATAGGCA_TCTTGTAGTTGATGCATACGC_AAG...              7822   \n",
       "TACCGAATGACCGCATTCGTT_CTCATTTACTCAATAGCAACG_TGT...              7890   \n",
       "GAACCGTGGCGTCCTCCGCTC_TGAATCCCCATAATTGGCAAC_TCC...              6975   \n",
       "...                                                              ...   \n",
       "CGCACAGAGGAGAACA-VIB_2                                          2487   \n",
       "ACCAAACGTTCTTATC-VIB_2                                          2532   \n",
       "GACCCAGAGATATGAC-VIB_2                                          1369   \n",
       "GATGATCCAACTCCCT-VIB_2                                          1314   \n",
       "TTACTTGCATACAACC-VIB_2                                          2082   \n",
       "\n",
       "                                                    pycisTopic_leiden_10_0.6  \\\n",
       "GAGGACATCTGAACTCCTTAA_TGGACCAGTAATACACCTACC_CTC...                         1   \n",
       "CGGATCTCGGCCAGGGCGGAA_TTTGGTCGATCACCTCATACA_CGG...                        10   \n",
       "CTTGCGACGACACTATAGGCA_TCTTGTAGTTGATGCATACGC_AAG...                        10   \n",
       "TACCGAATGACCGCATTCGTT_CTCATTTACTCAATAGCAACG_TGT...                         1   \n",
       "GAACCGTGGCGTCCTCCGCTC_TGAATCCCCATAATTGGCAAC_TCC...                         1   \n",
       "...                                                                      ...   \n",
       "CGCACAGAGGAGAACA-VIB_2                                                     0   \n",
       "ACCAAACGTTCTTATC-VIB_2                                                    12   \n",
       "GACCCAGAGATATGAC-VIB_2                                                    12   \n",
       "GATGATCCAACTCCCT-VIB_2                                                     7   \n",
       "TTACTTGCATACAACC-VIB_2                                                    12   \n",
       "\n",
       "                                                    cisTopic_log_nr_frag  \\\n",
       "GAGGACATCTGAACTCCTTAA_TGGACCAGTAATACACCTACC_CTC...              4.034789   \n",
       "CGGATCTCGGCCAGGGCGGAA_TTTGGTCGATCACCTCATACA_CGG...              4.163996   \n",
       "CTTGCGACGACACTATAGGCA_TCTTGTAGTTGATGCATACGC_AAG...              3.893318   \n",
       "TACCGAATGACCGCATTCGTT_CTCATTTACTCAATAGCAACG_TGT...              3.897077   \n",
       "GAACCGTGGCGTCCTCCGCTC_TGAATCCCCATAATTGGCAAC_TCC...              3.843544   \n",
       "...                                                                  ...   \n",
       "CGCACAGAGGAGAACA-VIB_2                                          3.395676   \n",
       "ACCAAACGTTCTTATC-VIB_2                                          3.403464   \n",
       "GACCCAGAGATATGAC-VIB_2                                          3.136403   \n",
       "GATGATCCAACTCCCT-VIB_2                                          3.118595   \n",
       "TTACTTGCATACAACC-VIB_2                                          3.318481   \n",
       "\n",
       "                                                   sample_id  \\\n",
       "GAGGACATCTGAACTCCTTAA_TGGACCAGTAATACACCTACC_CTC...   Broad_1   \n",
       "CGGATCTCGGCCAGGGCGGAA_TTTGGTCGATCACCTCATACA_CGG...   Broad_1   \n",
       "CTTGCGACGACACTATAGGCA_TCTTGTAGTTGATGCATACGC_AAG...   Broad_1   \n",
       "TACCGAATGACCGCATTCGTT_CTCATTTACTCAATAGCAACG_TGT...   Broad_1   \n",
       "GAACCGTGGCGTCCTCCGCTC_TGAATCCCCATAATTGGCAAC_TCC...   Broad_1   \n",
       "...                                                      ...   \n",
       "CGCACAGAGGAGAACA-VIB_2                                 VIB_2   \n",
       "ACCAAACGTTCTTATC-VIB_2                                 VIB_2   \n",
       "GACCCAGAGATATGAC-VIB_2                                 VIB_2   \n",
       "GATGATCCAACTCCCT-VIB_2                                 VIB_2   \n",
       "TTACTTGCATACAACC-VIB_2                                 VIB_2   \n",
       "\n",
       "                                                    Log_unique_nr_frag  \\\n",
       "GAGGACATCTGAACTCCTTAA_TGGACCAGTAATACACCTACC_CTC...            4.140068   \n",
       "CGGATCTCGGCCAGGGCGGAA_TTTGGTCGATCACCTCATACA_CGG...            4.131555   \n",
       "CTTGCGACGACACTATAGGCA_TCTTGTAGTTGATGCATACGC_AAG...            3.915664   \n",
       "TACCGAATGACCGCATTCGTT_CTCATTTACTCAATAGCAACG_TGT...            3.882240   \n",
       "GAACCGTGGCGTCCTCCGCTC_TGAATCCCCATAATTGGCAAC_TCC...            3.841422   \n",
       "...                                                                ...   \n",
       "CGCACAGAGGAGAACA-VIB_2                                        3.781468   \n",
       "ACCAAACGTTCTTATC-VIB_2                                        3.719000   \n",
       "GACCCAGAGATATGAC-VIB_2                                        3.610234   \n",
       "GATGATCCAACTCCCT-VIB_2                                        3.569608   \n",
       "TTACTTGCATACAACC-VIB_2                                        3.794279   \n",
       "\n",
       "                                                    Dupl_nr_frag  \\\n",
       "GAGGACATCTGAACTCCTTAA_TGGACCAGTAATACACCTACC_CTC...         29322   \n",
       "CGGATCTCGGCCAGGGCGGAA_TTTGGTCGATCACCTCATACA_CGG...         26645   \n",
       "CTTGCGACGACACTATAGGCA_TCTTGTAGTTGATGCATACGC_AAG...         17132   \n",
       "TACCGAATGACCGCATTCGTT_CTCATTTACTCAATAGCAACG_TGT...         17552   \n",
       "GAACCGTGGCGTCCTCCGCTC_TGAATCCCCATAATTGGCAAC_TCC...         13694   \n",
       "...                                                          ...   \n",
       "CGCACAGAGGAGAACA-VIB_2                                      3475   \n",
       "ACCAAACGTTCTTATC-VIB_2                                       658   \n",
       "GACCCAGAGATATGAC-VIB_2                                      1172   \n",
       "GATGATCCAACTCCCT-VIB_2                                       429   \n",
       "TTACTTGCATACAACC-VIB_2                                      1458   \n",
       "\n",
       "                                                    Total_nr_frag_in_regions  \\\n",
       "GAGGACATCTGAACTCCTTAA_TGGACCAGTAATACACCTACC_CTC...                     24899   \n",
       "CGGATCTCGGCCAGGGCGGAA_TTTGGTCGATCACCTCATACA_CGG...                     32467   \n",
       "CTTGCGACGACACTATAGGCA_TCTTGTAGTTGATGCATACGC_AAG...                     21308   \n",
       "TACCGAATGACCGCATTCGTT_CTCATTTACTCAATAGCAACG_TGT...                     20263   \n",
       "GAACCGTGGCGTCCTCCGCTC_TGAATCCCCATAATTGGCAAC_TCC...                     17960   \n",
       "...                                                                      ...   \n",
       "CGCACAGAGGAGAACA-VIB_2                                                  2044   \n",
       "ACCAAACGTTCTTATC-VIB_2                                                  1647   \n",
       "GACCCAGAGATATGAC-VIB_2                                                   769   \n",
       "GATGATCCAACTCCCT-VIB_2                                                   810   \n",
       "TTACTTGCATACAACC-VIB_2                                                  1206   \n",
       "\n",
       "                                                    ...  \\\n",
       "GAGGACATCTGAACTCCTTAA_TGGACCAGTAATACACCTACC_CTC...  ...   \n",
       "CGGATCTCGGCCAGGGCGGAA_TTTGGTCGATCACCTCATACA_CGG...  ...   \n",
       "CTTGCGACGACACTATAGGCA_TCTTGTAGTTGATGCATACGC_AAG...  ...   \n",
       "TACCGAATGACCGCATTCGTT_CTCATTTACTCAATAGCAACG_TGT...  ...   \n",
       "GAACCGTGGCGTCCTCCGCTC_TGAATCCCCATAATTGGCAAC_TCC...  ...   \n",
       "...                                                 ...   \n",
       "CGCACAGAGGAGAACA-VIB_2                              ...   \n",
       "ACCAAACGTTCTTATC-VIB_2                              ...   \n",
       "GACCCAGAGATATGAC-VIB_2                              ...   \n",
       "GATGATCCAACTCCCT-VIB_2                              ...   \n",
       "TTACTTGCATACAACC-VIB_2                              ...   \n",
       "\n",
       "                                                                                              barcode  \\\n",
       "GAGGACATCTGAACTCCTTAA_TGGACCAGTAATACACCTACC_CTC...  GAGGACATCTGAACTCCTTAA_TGGACCAGTAATACACCTACC_CT...   \n",
       "CGGATCTCGGCCAGGGCGGAA_TTTGGTCGATCACCTCATACA_CGG...  CGGATCTCGGCCAGGGCGGAA_TTTGGTCGATCACCTCATACA_CG...   \n",
       "CTTGCGACGACACTATAGGCA_TCTTGTAGTTGATGCATACGC_AAG...  CTTGCGACGACACTATAGGCA_TCTTGTAGTTGATGCATACGC_AA...   \n",
       "TACCGAATGACCGCATTCGTT_CTCATTTACTCAATAGCAACG_TGT...  TACCGAATGACCGCATTCGTT_CTCATTTACTCAATAGCAACG_TG...   \n",
       "GAACCGTGGCGTCCTCCGCTC_TGAATCCCCATAATTGGCAAC_TCC...  GAACCGTGGCGTCCTCCGCTC_TGAATCCCCATAATTGGCAAC_TC...   \n",
       "...                                                                                               ...   \n",
       "CGCACAGAGGAGAACA-VIB_2                                                               CGCACAGAGGAGAACA   \n",
       "ACCAAACGTTCTTATC-VIB_2                                                               ACCAAACGTTCTTATC   \n",
       "GACCCAGAGATATGAC-VIB_2                                                               GACCCAGAGATATGAC   \n",
       "GATGATCCAACTCCCT-VIB_2                                                               GATGATCCAACTCCCT   \n",
       "TTACTTGCATACAACC-VIB_2                                                               TTACTTGCATACAACC   \n",
       "\n",
       "                                                    fmx_droplet_type  \\\n",
       "GAGGACATCTGAACTCCTTAA_TGGACCAGTAATACACCTACC_CTC...               SNG   \n",
       "CGGATCTCGGCCAGGGCGGAA_TTTGGTCGATCACCTCATACA_CGG...               SNG   \n",
       "CTTGCGACGACACTATAGGCA_TCTTGTAGTTGATGCATACGC_AAG...               SNG   \n",
       "TACCGAATGACCGCATTCGTT_CTCATTTACTCAATAGCAACG_TGT...               SNG   \n",
       "GAACCGTGGCGTCCTCCGCTC_TGAATCCCCATAATTGGCAAC_TCC...               SNG   \n",
       "...                                                              ...   \n",
       "CGCACAGAGGAGAACA-VIB_2                                           SNG   \n",
       "ACCAAACGTTCTTATC-VIB_2                                           SNG   \n",
       "GACCCAGAGATATGAC-VIB_2                                           SNG   \n",
       "GATGATCCAACTCCCT-VIB_2                                           SNG   \n",
       "TTACTTGCATACAACC-VIB_2                                           SNG   \n",
       "\n",
       "                                                    cisTopic_log_nr_acc  \\\n",
       "GAGGACATCTGAACTCCTTAA_TGGACCAGTAATACACCTACC_CTC...             3.992067   \n",
       "CGGATCTCGGCCAGGGCGGAA_TTTGGTCGATCACCTCATACA_CGG...             4.120508   \n",
       "CTTGCGACGACACTATAGGCA_TCTTGTAGTTGATGCATACGC_AAG...             3.853211   \n",
       "TACCGAATGACCGCATTCGTT_CTCATTTACTCAATAGCAACG_TGT...             3.845656   \n",
       "GAACCGTGGCGTCCTCCGCTC_TGAATCCCCATAATTGGCAAC_TCC...             3.800992   \n",
       "...                                                                 ...   \n",
       "CGCACAGAGGAGAACA-VIB_2                                         3.361161   \n",
       "ACCAAACGTTCTTATC-VIB_2                                         3.397419   \n",
       "GACCCAGAGATATGAC-VIB_2                                         3.116940   \n",
       "GATGATCCAACTCCCT-VIB_2                                         3.116276   \n",
       "TTACTTGCATACAACC-VIB_2                                         3.292699   \n",
       "\n",
       "                                                    pycisTopic_leiden_10_0.8  \\\n",
       "GAGGACATCTGAACTCCTTAA_TGGACCAGTAATACACCTACC_CTC...                         0   \n",
       "CGGATCTCGGCCAGGGCGGAA_TTTGGTCGATCACCTCATACA_CGG...                         8   \n",
       "CTTGCGACGACACTATAGGCA_TCTTGTAGTTGATGCATACGC_AAG...                         8   \n",
       "TACCGAATGACCGCATTCGTT_CTCATTTACTCAATAGCAACG_TGT...                         0   \n",
       "GAACCGTGGCGTCCTCCGCTC_TGAATCCCCATAATTGGCAAC_TCC...                         0   \n",
       "...                                                                      ...   \n",
       "CGCACAGAGGAGAACA-VIB_2                                                     1   \n",
       "ACCAAACGTTCTTATC-VIB_2                                                    11   \n",
       "GACCCAGAGATATGAC-VIB_2                                                    11   \n",
       "GATGATCCAACTCCCT-VIB_2                                                     9   \n",
       "TTACTTGCATACAACC-VIB_2                                                    11   \n",
       "\n",
       "                                                    fmx_sample  \\\n",
       "GAGGACATCTGAACTCCTTAA_TGGACCAGTAATACACCTACC_CTC...     sampleB   \n",
       "CGGATCTCGGCCAGGGCGGAA_TTTGGTCGATCACCTCATACA_CGG...     sampleA   \n",
       "CTTGCGACGACACTATAGGCA_TCTTGTAGTTGATGCATACGC_AAG...     sampleB   \n",
       "TACCGAATGACCGCATTCGTT_CTCATTTACTCAATAGCAACG_TGT...     sampleB   \n",
       "GAACCGTGGCGTCCTCCGCTC_TGAATCCCCATAATTGGCAAC_TCC...     sampleA   \n",
       "...                                                        ...   \n",
       "CGCACAGAGGAGAACA-VIB_2                                 sampleB   \n",
       "ACCAAACGTTCTTATC-VIB_2                                 sampleB   \n",
       "GACCCAGAGATATGAC-VIB_2                                 sampleA   \n",
       "GATGATCCAACTCCCT-VIB_2                                 sampleB   \n",
       "TTACTTGCATACAACC-VIB_2                                 sampleA   \n",
       "\n",
       "                                                    Log_total_nr_frag  \\\n",
       "GAGGACATCTGAACTCCTTAA_TGGACCAGTAATACACCTACC_CTC...           4.634759   \n",
       "CGGATCTCGGCCAGGGCGGAA_TTTGGTCGATCACCTCATACA_CGG...           4.604042   \n",
       "CTTGCGACGACACTATAGGCA_TCTTGTAGTTGATGCATACGC_AAG...           4.404269   \n",
       "TACCGAATGACCGCATTCGTT_CTCATTTACTCAATAGCAACG_TGT...           4.401004   \n",
       "GAACCGTGGCGTCCTCCGCTC_TGAATCCCCATAATTGGCAAC_TCC...           4.314604   \n",
       "...                                                               ...   \n",
       "CGCACAGAGGAGAACA-VIB_2                                       3.978683   \n",
       "ACCAAACGTTCTTATC-VIB_2                                       3.770410   \n",
       "GACCCAGAGATATGAC-VIB_2                                       3.719994   \n",
       "GATGATCCAACTCCCT-VIB_2                                       3.617105   \n",
       "TTACTTGCATACAACC-VIB_2                                       3.885644   \n",
       "\n",
       "                                                    cisTopic_nr_acc  \\\n",
       "GAGGACATCTGAACTCCTTAA_TGGACCAGTAATACACCTACC_CTC...             9819   \n",
       "CGGATCTCGGCCAGGGCGGAA_TTTGGTCGATCACCTCATACA_CGG...            13198   \n",
       "CTTGCGACGACACTATAGGCA_TCTTGTAGTTGATGCATACGC_AAG...             7132   \n",
       "TACCGAATGACCGCATTCGTT_CTCATTTACTCAATAGCAACG_TGT...             7009   \n",
       "GAACCGTGGCGTCCTCCGCTC_TGAATCCCCATAATTGGCAAC_TCC...             6324   \n",
       "...                                                             ...   \n",
       "CGCACAGAGGAGAACA-VIB_2                                         2297   \n",
       "ACCAAACGTTCTTATC-VIB_2                                         2497   \n",
       "GACCCAGAGATATGAC-VIB_2                                         1309   \n",
       "GATGATCCAACTCCCT-VIB_2                                         1307   \n",
       "TTACTTGCATACAACC-VIB_2                                         1962   \n",
       "\n",
       "                                                   pycisTopic_leiden_10_1.0  \\\n",
       "GAGGACATCTGAACTCCTTAA_TGGACCAGTAATACACCTACC_CTC...                        0   \n",
       "CGGATCTCGGCCAGGGCGGAA_TTTGGTCGATCACCTCATACA_CGG...                        7   \n",
       "CTTGCGACGACACTATAGGCA_TCTTGTAGTTGATGCATACGC_AAG...                        7   \n",
       "TACCGAATGACCGCATTCGTT_CTCATTTACTCAATAGCAACG_TGT...                        0   \n",
       "GAACCGTGGCGTCCTCCGCTC_TGAATCCCCATAATTGGCAAC_TCC...                        0   \n",
       "...                                                                     ...   \n",
       "CGCACAGAGGAGAACA-VIB_2                                                    1   \n",
       "ACCAAACGTTCTTATC-VIB_2                                                    9   \n",
       "GACCCAGAGATATGAC-VIB_2                                                    9   \n",
       "GATGATCCAACTCCCT-VIB_2                                                    8   \n",
       "TTACTTGCATACAACC-VIB_2                                                    9   \n",
       "\n",
       "                                                   pycisTopic_leiden_10_1.2  \\\n",
       "GAGGACATCTGAACTCCTTAA_TGGACCAGTAATACACCTACC_CTC...                       15   \n",
       "CGGATCTCGGCCAGGGCGGAA_TTTGGTCGATCACCTCATACA_CGG...                        6   \n",
       "CTTGCGACGACACTATAGGCA_TCTTGTAGTTGATGCATACGC_AAG...                        6   \n",
       "TACCGAATGACCGCATTCGTT_CTCATTTACTCAATAGCAACG_TGT...                        5   \n",
       "GAACCGTGGCGTCCTCCGCTC_TGAATCCCCATAATTGGCAAC_TCC...                       15   \n",
       "...                                                                     ...   \n",
       "CGCACAGAGGAGAACA-VIB_2                                                    0   \n",
       "ACCAAACGTTCTTATC-VIB_2                                                    8   \n",
       "GACCCAGAGATATGAC-VIB_2                                                    8   \n",
       "GATGATCCAACTCCCT-VIB_2                                                    7   \n",
       "TTACTTGCATACAACC-VIB_2                                                    8   \n",
       "\n",
       "                                                    consensus_cell_type  \n",
       "GAGGACATCTGAACTCCTTAA_TGGACCAGTAATACACCTACC_CTC...          CD4+ T cell  \n",
       "CGGATCTCGGCCAGGGCGGAA_TTTGGTCGATCACCTCATACA_CGG...       CD14+ monocyte  \n",
       "CTTGCGACGACACTATAGGCA_TCTTGTAGTTGATGCATACGC_AAG...       CD14+ monocyte  \n",
       "TACCGAATGACCGCATTCGTT_CTCATTTACTCAATAGCAACG_TGT...          CD4+ T cell  \n",
       "GAACCGTGGCGTCCTCCGCTC_TGAATCCCCATAATTGGCAAC_TCC...          CD4+ T cell  \n",
       "...                                                                 ...  \n",
       "CGCACAGAGGAGAACA-VIB_2                                   CD14+ monocyte  \n",
       "ACCAAACGTTCTTATC-VIB_2                                   CD14+ monocyte  \n",
       "GACCCAGAGATATGAC-VIB_2                                   CD14+ monocyte  \n",
       "GATGATCCAACTCCCT-VIB_2                                 Cytotoxic T cell  \n",
       "TTACTTGCATACAACC-VIB_2                                 Cytotoxic T cell  \n",
       "\n",
       "[40496 rows x 27 columns]"
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "cellannot = pd.read_csv(\n",
    "    '/staging/leuven/stg_00002/lcb/cflerin/analysis/pbmc_atac/analysis2/data_freeze_Jun2021_library_downsampled/atac_qc_multiplet_merged/jupyter/fulldata_filtered_cell_data.tsv',\n",
    "    sep='\\t', index_col=0)\n",
    "cellannot"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "57ddcad6",
   "metadata": {},
   "outputs": [],
   "source": [
    "for key in cistopic_obj_dict.keys():\n",
    "    cistopic_obj_dict[key].add_cell_data(\n",
    "        cellannot[['consensus_cell_type']]\n",
    "    )"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "16f7a181",
   "metadata": {},
   "source": [
    "### Run models for visualization/clustering"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "b3e02cdd",
   "metadata": {},
   "outputs": [],
   "source": [
    "from pycisTopic.lda_models import run_cgs_models_mallet"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "dc0aae62",
   "metadata": {},
   "source": [
    "#### mallet models"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "bcda877c",
   "metadata": {},
   "outputs": [],
   "source": [
    "# set the memory limit for mallet (1GB default is not enough)\n",
    "os.environ['MALLET_MEMORY'] = '100G'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "b94d1eae",
   "metadata": {},
   "outputs": [],
   "source": [
    "f_mod_dir = 'models__consensus__mallet'\n",
    "if not os.path.exists(os.path.join(wdir, f_mod_dir)):\n",
    "    os.makedirs(os.path.join(wdir, f_mod_dir))\n",
    "    \n",
    "f_mod_tmpdir = '/scratch/leuven/325/vsc32528/tmp/mallet_libds'\n",
    "if not os.path.exists(f_mod_tmpdir):\n",
    "    os.makedirs(f_mod_tmpdir)\n",
    "\n",
    "n_topics = [2, 5, 10, 15, 20, 25, 30, 35, 40, 45, 50]\n",
    "n_topics_merged = [2, 5, 10, 20, 30, 35, 40, 45, 50, 60, 70, 80, 90, 100]\n",
    "n_iter = 500"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c708acb2",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Loading /staging/leuven/stg_00002/lcb/cflerin/analysis/pbmc_atac/analysis2/data_freeze_Jun2021_library_downsampled/atac_qc_multiplet_merged/jupyter/models__consensus__mallet/Broad_1__models_500_iter.pkl\n",
      "Loading /staging/leuven/stg_00002/lcb/cflerin/analysis/pbmc_atac/analysis2/data_freeze_Jun2021_library_downsampled/atac_qc_multiplet_merged/jupyter/models__consensus__mallet/Broad_2__models_500_iter.pkl\n",
      "Loading /staging/leuven/stg_00002/lcb/cflerin/analysis/pbmc_atac/analysis2/data_freeze_Jun2021_library_downsampled/atac_qc_multiplet_merged/jupyter/models__consensus__mallet/Broad_mito_1__models_500_iter.pkl\n",
      "Loading /staging/leuven/stg_00002/lcb/cflerin/analysis/pbmc_atac/analysis2/data_freeze_Jun2021_library_downsampled/atac_qc_multiplet_merged/jupyter/models__consensus__mallet/Broad_mito_2__models_500_iter.pkl\n",
      "Loading /staging/leuven/stg_00002/lcb/cflerin/analysis/pbmc_atac/analysis2/data_freeze_Jun2021_library_downsampled/atac_qc_multiplet_merged/jupyter/models__consensus__mallet/CNAG_1__models_500_iter.pkl\n",
      "Loading /staging/leuven/stg_00002/lcb/cflerin/analysis/pbmc_atac/analysis2/data_freeze_Jun2021_library_downsampled/atac_qc_multiplet_merged/jupyter/models__consensus__mallet/CNAG_2__models_500_iter.pkl\n",
      "Loading /staging/leuven/stg_00002/lcb/cflerin/analysis/pbmc_atac/analysis2/data_freeze_Jun2021_library_downsampled/atac_qc_multiplet_merged/jupyter/models__consensus__mallet/s3atac__models_500_iter.pkl\n",
      "Loading /staging/leuven/stg_00002/lcb/cflerin/analysis/pbmc_atac/analysis2/data_freeze_Jun2021_library_downsampled/atac_qc_multiplet_merged/jupyter/models__consensus__mallet/Sanger_1__models_500_iter.pkl\n",
      "Loading /staging/leuven/stg_00002/lcb/cflerin/analysis/pbmc_atac/analysis2/data_freeze_Jun2021_library_downsampled/atac_qc_multiplet_merged/jupyter/models__consensus__mallet/Sanger_2__models_500_iter.pkl\n",
      "Loading /staging/leuven/stg_00002/lcb/cflerin/analysis/pbmc_atac/analysis2/data_freeze_Jun2021_library_downsampled/atac_qc_multiplet_merged/jupyter/models__consensus__mallet/Stanford_1__models_500_iter.pkl\n",
      "Loading /staging/leuven/stg_00002/lcb/cflerin/analysis/pbmc_atac/analysis2/data_freeze_Jun2021_library_downsampled/atac_qc_multiplet_merged/jupyter/models__consensus__mallet/Stanford_2__models_500_iter.pkl\n",
      "Loading /staging/leuven/stg_00002/lcb/cflerin/analysis/pbmc_atac/analysis2/data_freeze_Jun2021_library_downsampled/atac_qc_multiplet_merged/jupyter/models__consensus__mallet/VIB_1__models_500_iter.pkl\n",
      "Loading /staging/leuven/stg_00002/lcb/cflerin/analysis/pbmc_atac/analysis2/data_freeze_Jun2021_library_downsampled/atac_qc_multiplet_merged/jupyter/models__consensus__mallet/VIB_2__models_500_iter.pkl\n",
      "Running merged\n",
      "2021-09-15 11:40:35,667 cisTopic     INFO     Formatting input to corpus\n",
      "2021-09-15 11:40:46,160 gensim.corpora.dictionary INFO     adding document #0 to Dictionary(0 unique tokens: [])\n",
      "2021-09-15 11:41:31,096 gensim.corpora.dictionary INFO     adding document #10000 to Dictionary(0 unique tokens: [])\n",
      "2021-09-15 11:42:16,096 gensim.corpora.dictionary INFO     adding document #20000 to Dictionary(0 unique tokens: [])\n",
      "2021-09-15 11:43:01,065 gensim.corpora.dictionary INFO     adding document #30000 to Dictionary(0 unique tokens: [])\n",
      "2021-09-15 11:43:47,581 gensim.corpora.dictionary INFO     adding document #40000 to Dictionary(0 unique tokens: [])\n",
      "2021-09-15 11:43:50,487 gensim.corpora.dictionary INFO     built Dictionary(327964 unique tokens: ['0', '1', '2', '3', '4']...) from 40496 documents (total 242626278 corpus positions)\n",
      "2021-09-15 11:43:50,491 cisTopic     INFO     Running model with 2 topics\n",
      "2021-09-15 11:43:50,502 LDAMalletWrapper INFO     Serializing temporary corpus to /scratch/leuven/325/vsc32528/tmp/mallet_libds/merged_corpus.txt\n",
      "2021-09-15 11:52:20,378 LDAMalletWrapper INFO     Converting temporary corpus to MALLET format with mallet import-file --preserve-case --keep-sequence --remove-stopwords --token-regex \"\\S+\" --input /scratch/leuven/325/vsc32528/tmp/mallet_libds/merged_corpus.txt --output /scratch/leuven/325/vsc32528/tmp/mallet_libds/merged_corpus.mallet\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/bin/bash: warning: setlocale: LC_ALL: cannot change locale (en_US.UTF-8)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2021-09-15 11:54:22,235 LDAMalletWrapper INFO     Training MALLET LDA with mallet train-topics --input /scratch/leuven/325/vsc32528/tmp/mallet_libds/merged_corpus.mallet --num-topics 2  --alpha 50 --beta 0.1 --optimize-interval 0 --num-threads 16 --output-state /scratch/leuven/325/vsc32528/tmp/mallet_libds/merged_c765e5_state.mallet.gz --output-doc-topics /scratch/leuven/325/vsc32528/tmp/mallet_libds/merged_c765e5_doctopics.txt --output-topic-keys /scratch/leuven/325/vsc32528/tmp/mallet_libds/merged_c765e5_topickeys.txt --num-iterations 500 --inferencer-filename /scratch/leuven/325/vsc32528/tmp/mallet_libds/merged_c765e5_inferencer.mallet --doc-topics-threshold 0.0  --random-seed 555\n",
      "2021-09-15 13:01:45,814 LDAMalletWrapper INFO     loading assigned topics from /scratch/leuven/325/vsc32528/tmp/mallet_libds/merged_c765e5_state.mallet.gz\n",
      "2021-09-15 13:12:59,132 cisTopic     INFO     Model with 2 topics done!\n",
      "2021-09-15 13:12:59,133 cisTopic     INFO     Running model with 5 topics\n",
      "2021-09-15 13:12:59,148 LDAMalletWrapper INFO     MALLET corpus already exists, training model\n",
      "2021-09-15 13:12:59,149 LDAMalletWrapper INFO     Training MALLET LDA with mallet train-topics --input /scratch/leuven/325/vsc32528/tmp/mallet_libds/merged_corpus.mallet --num-topics 5  --alpha 50 --beta 0.1 --optimize-interval 0 --num-threads 16 --output-state /scratch/leuven/325/vsc32528/tmp/mallet_libds/merged_99d2fb_state.mallet.gz --output-doc-topics /scratch/leuven/325/vsc32528/tmp/mallet_libds/merged_99d2fb_doctopics.txt --output-topic-keys /scratch/leuven/325/vsc32528/tmp/mallet_libds/merged_99d2fb_topickeys.txt --num-iterations 500 --inferencer-filename /scratch/leuven/325/vsc32528/tmp/mallet_libds/merged_99d2fb_inferencer.mallet --doc-topics-threshold 0.0  --random-seed 555\n",
      "2021-09-15 14:30:16,039 LDAMalletWrapper INFO     loading assigned topics from /scratch/leuven/325/vsc32528/tmp/mallet_libds/merged_99d2fb_state.mallet.gz\n",
      "2021-09-15 14:40:55,873 cisTopic     INFO     Model with 5 topics done!\n",
      "2021-09-15 14:40:55,875 cisTopic     INFO     Running model with 10 topics\n",
      "2021-09-15 14:40:55,889 LDAMalletWrapper INFO     MALLET corpus already exists, training model\n",
      "2021-09-15 14:40:55,889 LDAMalletWrapper INFO     Training MALLET LDA with mallet train-topics --input /scratch/leuven/325/vsc32528/tmp/mallet_libds/merged_corpus.mallet --num-topics 10  --alpha 50 --beta 0.1 --optimize-interval 0 --num-threads 16 --output-state /scratch/leuven/325/vsc32528/tmp/mallet_libds/merged_611260_state.mallet.gz --output-doc-topics /scratch/leuven/325/vsc32528/tmp/mallet_libds/merged_611260_doctopics.txt --output-topic-keys /scratch/leuven/325/vsc32528/tmp/mallet_libds/merged_611260_topickeys.txt --num-iterations 500 --inferencer-filename /scratch/leuven/325/vsc32528/tmp/mallet_libds/merged_611260_inferencer.mallet --doc-topics-threshold 0.0  --random-seed 555\n",
      "2021-09-15 16:01:59,182 LDAMalletWrapper INFO     loading assigned topics from /scratch/leuven/325/vsc32528/tmp/mallet_libds/merged_611260_state.mallet.gz\n",
      "2021-09-15 16:13:14,193 cisTopic     INFO     Model with 10 topics done!\n",
      "2021-09-15 16:13:14,195 cisTopic     INFO     Running model with 20 topics\n",
      "2021-09-15 16:13:14,207 LDAMalletWrapper INFO     MALLET corpus already exists, training model\n",
      "2021-09-15 16:13:14,207 LDAMalletWrapper INFO     Training MALLET LDA with mallet train-topics --input /scratch/leuven/325/vsc32528/tmp/mallet_libds/merged_corpus.mallet --num-topics 20  --alpha 50 --beta 0.1 --optimize-interval 0 --num-threads 16 --output-state /scratch/leuven/325/vsc32528/tmp/mallet_libds/merged_cf9d39_state.mallet.gz --output-doc-topics /scratch/leuven/325/vsc32528/tmp/mallet_libds/merged_cf9d39_doctopics.txt --output-topic-keys /scratch/leuven/325/vsc32528/tmp/mallet_libds/merged_cf9d39_topickeys.txt --num-iterations 500 --inferencer-filename /scratch/leuven/325/vsc32528/tmp/mallet_libds/merged_cf9d39_inferencer.mallet --doc-topics-threshold 0.0  --random-seed 555\n",
      "2021-09-15 17:57:59,951 LDAMalletWrapper INFO     loading assigned topics from /scratch/leuven/325/vsc32528/tmp/mallet_libds/merged_cf9d39_state.mallet.gz\n",
      "2021-09-15 18:09:00,702 cisTopic     INFO     Model with 20 topics done!\n"
     ]
    }
   ],
   "source": [
    "# mallet implementation\n",
    "models_dict = {}\n",
    "for key in cistopic_obj_dict.keys():\n",
    "    f_mod = os.path.join(wdir, f_mod_dir, key + '__models_' + str(n_iter) + '_iter.pkl')\n",
    "    if os.path.isfile(f_mod):\n",
    "        print(f\"Loading {f_mod}\")\n",
    "        with open(f_mod, 'rb') as f:\n",
    "            models_dict[key] = pickle.load(f)\n",
    "    else:\n",
    "        print(f\"Running {key}\")\n",
    "        model = run_cgs_models_mallet(\n",
    "                            'mallet',\n",
    "                            cistopic_obj_dict[key],\n",
    "                            n_topics=n_topics_merged if key=='merged' else n_topics ,\n",
    "                            n_cpu=16,\n",
    "                            n_iter=n_iter,\n",
    "                            random_state=555,\n",
    "                            alpha=50,\n",
    "                            alpha_by_topic=True,\n",
    "                            eta=0.1,\n",
    "                            eta_by_topic=False,\n",
    "                            tmp_path=os.path.join(f_mod_tmpdir, key + '_')\n",
    "                            )\n",
    "        # Save\n",
    "        with open(f_mod, 'wb') as f:\n",
    "            pickle.dump(model, f)\n",
    "        print(f\"Finished {key}\")\n",
    "        models_dict[key] = model"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "397890a7",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "pycisTopic_singularity_20210921",
   "language": "python",
   "name": "pycistopic_singularity"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.6"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
