{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "7f29dc3f",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import glob\n",
    "import os\n",
    "from collections import OrderedDict"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "0e2787a0",
   "metadata": {},
   "outputs": [],
   "source": [
    "filenames = glob.glob('selected_barcodes/*')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "5c4e0149",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'Broad_1': 'selected_barcodes/Broad_1.cell_barcodes.txt',\n",
       " 'Broad_2': 'selected_barcodes/Broad_2.cell_barcodes.txt',\n",
       " 'Broad_mito_1': 'selected_barcodes/Broad_mito_1.cell_barcodes.txt',\n",
       " 'Broad_mito_2': 'selected_barcodes/Broad_mito_2.cell_barcodes.txt',\n",
       " 'CNAG_1': 'selected_barcodes/CNAG_1.cell_barcodes.txt',\n",
       " 'CNAG_2': 'selected_barcodes/CNAG_2.cell_barcodes.txt',\n",
       " 'Sanger_1': 'selected_barcodes/Sanger_1.cell_barcodes.txt',\n",
       " 'Sanger_2': 'selected_barcodes/Sanger_2.cell_barcodes.txt',\n",
       " 'Stanford_1': 'selected_barcodes/Stanford_1.cell_barcodes.txt',\n",
       " 'Stanford_2': 'selected_barcodes/Stanford_2.cell_barcodes.txt',\n",
       " 'VIB_1': 'selected_barcodes/VIB_1.cell_barcodes.txt',\n",
       " 'VIB_2': 'selected_barcodes/VIB_2.cell_barcodes.txt',\n",
       " 'VIB_Hydrop_1': 'selected_barcodes/VIB_Hydrop_1.cell_barcodes.txt',\n",
       " 'VIB_Hydrop_2': 'selected_barcodes/VIB_Hydrop_2.cell_barcodes.txt',\n",
       " 's3atac': 'selected_barcodes/s3atac.cell_barcodes.txt'}"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "filenames = sorted(glob.glob('selected_barcodes/*'))\n",
    "samples = [item.replace(\"selected_barcodes/\", \"\") for item in filenames]\n",
    "samples = [item.replace(\".cell_barcodes.txt\", \"\") for item in samples]\n",
    "bc_dict = {samples[i]: filenames[i] for i in range(len(samples))}\n",
    "bc_dict"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "81d4ecfb",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'Broad_1': 'fragments_postbap/Broad_1.sinto.mm.fragments.tsv',\n",
       " 'Broad_2': 'fragments_postbap/Broad_2.sinto.mm.fragments.tsv',\n",
       " 'Broad_mito_1': 'fragments_postbap/Broad_mito_1.sinto.mm.fragments.tsv',\n",
       " 'Broad_mito_2': 'fragments_postbap/Broad_mito_2.sinto.mm.fragments.tsv',\n",
       " 'CNAG_1': 'fragments_postbap/CNAG_1.sinto.mm.fragments.tsv',\n",
       " 'CNAG_2': 'fragments_postbap/CNAG_2.sinto.mm.fragments.tsv',\n",
       " 'Sanger_1': 'fragments_postbap/Sanger_1.sinto.mm.fragments.tsv',\n",
       " 'Sanger_2': 'fragments_postbap/Sanger_2.sinto.mm.fragments.tsv',\n",
       " 'Stanford_1': 'fragments_postbap/Stanford_1.sinto.mm.fragments.tsv',\n",
       " 'Stanford_2': 'fragments_postbap/Stanford_2.sinto.mm.fragments.tsv',\n",
       " 'VIB_1': 'fragments_postbap/VIB_1.sinto.mm.fragments.tsv',\n",
       " 'VIB_2': 'fragments_postbap/VIB_2.sinto.mm.fragments.tsv',\n",
       " 'VIB_Hydrop_1': 'fragments_postbap/VIB_Hydrop_1.sinto.mm.fragments.tsv',\n",
       " 'VIB_Hydrop_2': 'fragments_postbap/VIB_Hydrop_2.sinto.mm.fragments.tsv',\n",
       " 's3atac': 'fragments_postbap/s3atac.sinto.mm.fragments.tsv'}"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "filenames = sorted(glob.glob('fragments_postbap/*tsv'))\n",
    "samples = [item.replace(\"fragments_postbap/\", \"\") for item in filenames]\n",
    "samples = [item.replace(\".sinto.mm.fragments.tsv\", \"\") for item in samples]\n",
    "frag_dict = {samples[i]: filenames[i] for i in range(len(samples))}\n",
    "frag_dict"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "09a46016",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Broad_1\n",
      "\ttotal reads: 191995191\n",
      "\tcells pre-scrublet/freemuxlet: 4204\n",
      "\tmapped reads per cell: 45669.64581351094\n",
      "\tfraction raw reads in cells: 0.6444203230069445\n",
      "\tfraction unique fragments in cells: 0.5989097027567128\n",
      "Broad_2\n",
      "\ttotal reads: 187364203\n",
      "\tcells pre-scrublet/freemuxlet: 4053\n",
      "\tmapped reads per cell: 46228.52282260054\n",
      "\tfraction raw reads in cells: 0.6112329952376229\n",
      "\tfraction unique fragments in cells: 0.5729772517096624\n",
      "Broad_mito_1\n",
      "\ttotal reads: 91805815\n",
      "\tcells pre-scrublet/freemuxlet: 3466\n",
      "\tmapped reads per cell: 26487.54039238315\n",
      "\tfraction raw reads in cells: 0.9354537618341496\n",
      "\tfraction unique fragments in cells: 0.9209491757864046\n",
      "Broad_mito_2\n",
      "\ttotal reads: 107153707\n",
      "\tcells pre-scrublet/freemuxlet: 3302\n",
      "\tmapped reads per cell: 32451.152937613566\n",
      "\tfraction raw reads in cells: 0.9456769797054245\n",
      "\tfraction unique fragments in cells: 0.9305605465823572\n",
      "CNAG_1\n",
      "\ttotal reads: 205525851\n",
      "\tcells pre-scrublet/freemuxlet: 2696\n",
      "\tmapped reads per cell: 76233.62425816024\n",
      "\tfraction raw reads in cells: 0.4627793075042419\n",
      "\tfraction unique fragments in cells: 0.3766766831812998\n",
      "CNAG_2\n",
      "\ttotal reads: 233220185\n",
      "\tcells pre-scrublet/freemuxlet: 2712\n",
      "\tmapped reads per cell: 85995.64343657818\n",
      "\tfraction raw reads in cells: 0.5026103851173945\n",
      "\tfraction unique fragments in cells: 0.3967154770728165\n",
      "Sanger_1\n",
      "\ttotal reads: 421827496\n",
      "\tcells pre-scrublet/freemuxlet: 3142\n",
      "\tmapped reads per cell: 134254.45448758753\n",
      "\tfraction raw reads in cells: 0.5071430834371214\n",
      "\tfraction unique fragments in cells: 0.4562991065447841\n",
      "Sanger_2\n",
      "\ttotal reads: 439633602\n",
      "\tcells pre-scrublet/freemuxlet: 3814\n",
      "\tmapped reads per cell: 115268.3801782905\n",
      "\tfraction raw reads in cells: 0.48600613790207964\n",
      "\tfraction unique fragments in cells: 0.4392960380592493\n",
      "Stanford_1\n",
      "\ttotal reads: 183360947\n",
      "\tcells pre-scrublet/freemuxlet: 768\n",
      "\tmapped reads per cell: 238751.23307291666\n",
      "\tfraction raw reads in cells: 0.37056446921601033\n",
      "\tfraction unique fragments in cells: 0.33848973699717066\n",
      "Stanford_2\n",
      "\ttotal reads: 139875885\n",
      "\tcells pre-scrublet/freemuxlet: 1403\n",
      "\tmapped reads per cell: 99697.70848182467\n",
      "\tfraction raw reads in cells: 0.3350888110556012\n",
      "\tfraction unique fragments in cells: 0.3282839223822772\n",
      "VIB_1\n",
      "\ttotal reads: 193215267\n",
      "\tcells pre-scrublet/freemuxlet: 2813\n",
      "\tmapped reads per cell: 68686.55065766086\n",
      "\tfraction raw reads in cells: 0.43008775802380045\n",
      "\tfraction unique fragments in cells: 0.33918550222001953\n",
      "VIB_2\n",
      "\ttotal reads: 296982481\n",
      "\tcells pre-scrublet/freemuxlet: 7336\n",
      "\tmapped reads per cell: 40482.88999454744\n",
      "\tfraction raw reads in cells: 0.4095402684712571\n",
      "\tfraction unique fragments in cells: 0.3745887730068407\n",
      "VIB_Hydrop_1\n",
      "\ttotal reads: 124602117\n",
      "\tcells pre-scrublet/freemuxlet: 2410\n",
      "\tmapped reads per cell: 51702.123236514526\n",
      "\tfraction raw reads in cells: 0.7635622916422841\n",
      "\tfraction unique fragments in cells: 0.537952195289631\n",
      "VIB_Hydrop_2\n",
      "\ttotal reads: 100478784\n",
      "\tcells pre-scrublet/freemuxlet: 2744\n",
      "\tmapped reads per cell: 36617.63265306123\n",
      "\tfraction raw reads in cells: 0.7208889191971113\n",
      "\tfraction unique fragments in cells: 0.5328964118709969\n",
      "s3atac\n",
      "\ttotal reads: 1285843224\n",
      "\tcells pre-scrublet/freemuxlet: 3122\n",
      "\tmapped reads per cell: 411865.22229340166\n",
      "\tfraction raw reads in cells: 0.37927564643759404\n",
      "\tfraction unique fragments in cells: 0.36792517150421816\n"
     ]
    }
   ],
   "source": [
    "for key in bc_dict.keys():\n",
    "    print(key)\n",
    "    df_frags = pd.read_csv(frag_dict[key], sep='\\t', header=None, index_col=None)\n",
    "    df_frags.columns = \"chr start end bc count\".split()\n",
    "    df_bc = pd.read_csv(bc_dict[key], header=None, index_col=None)\n",
    "    df_bc.columns = \"bc\".split()\n",
    "\n",
    "\n",
    "    df_frags_incells = df_frags[df_frags['bc'].isin(df_bc['bc'])]\n",
    "    frac_reads = df_frags_incells['count'].sum()/df_frags['count'].sum()\n",
    "    frac_fragments = len(df_frags_incells)/len(df_frags)\n",
    "\n",
    "    print(f'\\ttotal reads: {df_frags[\"count\"].sum()}')\n",
    "    print(f'\\tcells pre-scrublet/freemuxlet: {len(df_bc)}')\n",
    "    print(f'\\tmapped reads per cell: {df_frags[\"count\"].sum()/len(df_bc)}')\n",
    "    print(f'\\tfraction mapped fragments in cells: {frac_reads}')\n",
    "    print(f'\\tfraction unique fragments in cells: {frac_fragments}')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "dc009658",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Broad_1\n",
      "\ttotal reads: 191995191\n",
      "\tcells pre-scrublet/freemuxlet: 4204\n",
      "\tmapped reads per cell: 45669.64581351094\n",
      "\tfraction mapped fragments in cells: 0.6444203230069445\n",
      "\tfraction unique fragments in cells: 0.5989097027567128\n",
      "Broad_2\n",
      "\ttotal reads: 187364203\n",
      "\tcells pre-scrublet/freemuxlet: 4053\n",
      "\tmapped reads per cell: 46228.52282260054\n",
      "\tfraction mapped fragments in cells: 0.6112329952376229\n",
      "\tfraction unique fragments in cells: 0.5729772517096624\n",
      "Broad_mito_1\n",
      "\ttotal reads: 91805815\n",
      "\tcells pre-scrublet/freemuxlet: 3466\n",
      "\tmapped reads per cell: 26487.54039238315\n",
      "\tfraction mapped fragments in cells: 0.9354537618341496\n",
      "\tfraction unique fragments in cells: 0.9209491757864046\n",
      "Broad_mito_2\n",
      "\ttotal reads: 107153707\n",
      "\tcells pre-scrublet/freemuxlet: 3302\n",
      "\tmapped reads per cell: 32451.152937613566\n",
      "\tfraction mapped fragments in cells: 0.9456769797054245\n",
      "\tfraction unique fragments in cells: 0.9305605465823572\n",
      "CNAG_1\n",
      "\ttotal reads: 205525851\n",
      "\tcells pre-scrublet/freemuxlet: 2696\n",
      "\tmapped reads per cell: 76233.62425816024\n",
      "\tfraction mapped fragments in cells: 0.4627793075042419\n",
      "\tfraction unique fragments in cells: 0.3766766831812998\n",
      "CNAG_2\n",
      "\ttotal reads: 233220185\n",
      "\tcells pre-scrublet/freemuxlet: 2712\n",
      "\tmapped reads per cell: 85995.64343657818\n",
      "\tfraction mapped fragments in cells: 0.5026103851173945\n",
      "\tfraction unique fragments in cells: 0.3967154770728165\n",
      "Sanger_1\n",
      "\ttotal reads: 421827496\n",
      "\tcells pre-scrublet/freemuxlet: 3142\n",
      "\tmapped reads per cell: 134254.45448758753\n",
      "\tfraction mapped fragments in cells: 0.5071430834371214\n",
      "\tfraction unique fragments in cells: 0.4562991065447841\n",
      "Sanger_2\n",
      "\ttotal reads: 439633602\n",
      "\tcells pre-scrublet/freemuxlet: 3814\n",
      "\tmapped reads per cell: 115268.3801782905\n",
      "\tfraction mapped fragments in cells: 0.48600613790207964\n",
      "\tfraction unique fragments in cells: 0.4392960380592493\n",
      "Stanford_1\n",
      "\ttotal reads: 183360947\n",
      "\tcells pre-scrublet/freemuxlet: 768\n",
      "\tmapped reads per cell: 238751.23307291666\n",
      "\tfraction mapped fragments in cells: 0.37056446921601033\n",
      "\tfraction unique fragments in cells: 0.33848973699717066\n",
      "Stanford_2\n",
      "\ttotal reads: 139875885\n",
      "\tcells pre-scrublet/freemuxlet: 1403\n",
      "\tmapped reads per cell: 99697.70848182467\n",
      "\tfraction mapped fragments in cells: 0.3350888110556012\n",
      "\tfraction unique fragments in cells: 0.3282839223822772\n",
      "VIB_1\n",
      "\ttotal reads: 193215267\n",
      "\tcells pre-scrublet/freemuxlet: 2813\n",
      "\tmapped reads per cell: 68686.55065766086\n",
      "\tfraction mapped fragments in cells: 0.43008775802380045\n",
      "\tfraction unique fragments in cells: 0.33918550222001953\n",
      "VIB_2\n",
      "\ttotal reads: 296982481\n",
      "\tcells pre-scrublet/freemuxlet: 7336\n",
      "\tmapped reads per cell: 40482.88999454744\n",
      "\tfraction mapped fragments in cells: 0.4095402684712571\n",
      "\tfraction unique fragments in cells: 0.3745887730068407\n",
      "VIB_Hydrop_1\n",
      "\ttotal reads: 124602117\n",
      "\tcells pre-scrublet/freemuxlet: 2410\n",
      "\tmapped reads per cell: 51702.123236514526\n",
      "\tfraction mapped fragments in cells: 0.7635622916422841\n",
      "\tfraction unique fragments in cells: 0.537952195289631\n",
      "VIB_Hydrop_2\n",
      "\ttotal reads: 100478784\n",
      "\tcells pre-scrublet/freemuxlet: 2744\n",
      "\tmapped reads per cell: 36617.63265306123\n",
      "\tfraction mapped fragments in cells: 0.7208889191971113\n",
      "\tfraction unique fragments in cells: 0.5328964118709969\n",
      "s3atac\n",
      "\ttotal reads: 1285843224\n",
      "\tcells pre-scrublet/freemuxlet: 3122\n",
      "\tmapped reads per cell: 411865.22229340166\n",
      "\tfraction mapped fragments in cells: 0.37927564643759404\n",
      "\tfraction unique fragments in cells: 0.36792517150421816\n"
     ]
    }
   ],
   "source": [
    "for key in bc_dict.keys():\n",
    "    print(key)\n",
    "    df_frags = pd.read_csv(frag_dict[key], sep='\\t', header=None, index_col=None)\n",
    "    df_frags.columns = \"chr start end bc count\".split()\n",
    "    df_bc = pd.read_csv(bc_dict[key], header=None, index_col=None)\n",
    "    df_bc.columns = \"bc\".split()\n",
    "\n",
    "\n",
    "    df_frags_incells = df_frags[df_frags['bc'].isin(df_bc['bc'])]\n",
    "    frac_reads = df_frags_incells['count'].sum()/df_frags['count'].sum()\n",
    "    frac_fragments = len(df_frags_incells)/len(df_frags)\n",
    "\n",
    "    print(f'\\ttotal reads: {df_frags[\"count\"].sum()}')\n",
    "    print(f'\\tcells pre-scrublet/freemuxlet: {len(df_bc)}')\n",
    "    print(f'\\tmapped reads per cell: {df_frags[\"count\"].sum()/len(df_bc)}')\n",
    "    print(f'\\tfraction mapped fragments in cells: {frac_reads}')\n",
    "    print(f'\\tfraction unique fragments in cells: {frac_fragments}')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "b3d9d85b",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'Broad_1': 'mapping_stats/Broad_1.mapping_stats.tsv',\n",
       " 'Broad_2': 'mapping_stats/Broad_2.mapping_stats.tsv',\n",
       " 'Broad_mito_1': 'mapping_stats/Broad_mito_1.mapping_stats.tsv',\n",
       " 'Broad_mito_2': 'mapping_stats/Broad_mito_2.mapping_stats.tsv',\n",
       " 'CNAG_1': 'mapping_stats/CNAG_1.mapping_stats.tsv',\n",
       " 'CNAG_2': 'mapping_stats/CNAG_2.mapping_stats.tsv',\n",
       " 'HYA__24010b__20210813_384_PBMC_11_S9': 'mapping_stats/HYA__24010b__20210813_384_PBMC_11_S9.mapping_stats.tsv',\n",
       " 'HYA__2beafa__20210813_384_PBMC_12_S10': 'mapping_stats/HYA__2beafa__20210813_384_PBMC_12_S10.mapping_stats.tsv',\n",
       " 'HYA__3d6da9__20210813_384_PBMC_13_S11': 'mapping_stats/HYA__3d6da9__20210813_384_PBMC_13_S11.mapping_stats.tsv',\n",
       " 'HYA__5028cb__20210813_384_PBMC_14_S12': 'mapping_stats/HYA__5028cb__20210813_384_PBMC_14_S12.mapping_stats.tsv',\n",
       " 'Sanger_1': 'mapping_stats/Sanger_1.mapping_stats.tsv',\n",
       " 'Sanger_2': 'mapping_stats/Sanger_2.mapping_stats.tsv',\n",
       " 'Stanford_1': 'mapping_stats/Stanford_1.mapping_stats.tsv',\n",
       " 'Stanford_2': 'mapping_stats/Stanford_2.mapping_stats.tsv',\n",
       " 'VIB_1': 'mapping_stats/VIB_1.mapping_stats.tsv',\n",
       " 'VIB_2': 'mapping_stats/VIB_2.mapping_stats.tsv'}"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "filenames = sorted(glob.glob('mapping_stats/*'))\n",
    "samples = [item.replace(\"mapping_stats/\", \"\") for item in filenames]\n",
    "samples = [item.replace(\".mapping_stats.tsv\", \"\") for item in samples]\n",
    "stats_dict = {samples[i]: filenames[i] for i in range(len(samples))}\n",
    "stats_dict"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "c1c94ff2",
   "metadata": {},
   "outputs": [],
   "source": [
    "key = 'VIB_1'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "f7b9d03d",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>VIB_1</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>raw total sequences:</th>\n",
       "      <td>4.533154e+08</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>filtered sequences:</th>\n",
       "      <td>0.000000e+00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>sequences:</th>\n",
       "      <td>4.533154e+08</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>is sorted:</th>\n",
       "      <td>1.000000e+00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1st fragments:</th>\n",
       "      <td>2.266577e+08</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>last fragments:</th>\n",
       "      <td>2.266577e+08</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>reads mapped:</th>\n",
       "      <td>4.430560e+08</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>reads mapped and paired:</th>\n",
       "      <td>4.405570e+08</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>reads unmapped:</th>\n",
       "      <td>1.025936e+07</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>reads properly paired:</th>\n",
       "      <td>4.389387e+08</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>reads paired:</th>\n",
       "      <td>4.533154e+08</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>reads duplicated:</th>\n",
       "      <td>8.205966e+07</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>reads MQ0:</th>\n",
       "      <td>3.450323e+07</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>reads QC failed:</th>\n",
       "      <td>0.000000e+00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>non-primary alignments:</th>\n",
       "      <td>0.000000e+00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>supplementary alignments:</th>\n",
       "      <td>1.859000e+03</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>total length:</th>\n",
       "      <td>2.147508e+10</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>total first fragment length:</th>\n",
       "      <td>1.083189e+10</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>total last fragment length:</th>\n",
       "      <td>1.064319e+10</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>bases mapped:</th>\n",
       "      <td>2.098839e+10</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>bases mapped (cigar):</th>\n",
       "      <td>2.088950e+10</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>bases trimmed:</th>\n",
       "      <td>0.000000e+00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>bases duplicated:</th>\n",
       "      <td>3.904326e+09</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>mismatches:</th>\n",
       "      <td>7.722158e+07</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>error rate:</th>\n",
       "      <td>3.696670e-03</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>average length:</th>\n",
       "      <td>4.700000e+01</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>average first fragment length:</th>\n",
       "      <td>4.800000e+01</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>average last fragment length:</th>\n",
       "      <td>4.700000e+01</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>maximum length:</th>\n",
       "      <td>5.000000e+01</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>maximum first fragment length:</th>\n",
       "      <td>5.000000e+01</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>maximum last fragment length:</th>\n",
       "      <td>4.900000e+01</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>average quality:</th>\n",
       "      <td>3.590000e+01</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>insert size average:</th>\n",
       "      <td>1.421000e+02</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>insert size standard deviation:</th>\n",
       "      <td>9.130000e+01</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>inward oriented pairs:</th>\n",
       "      <td>1.986024e+08</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>outward oriented pairs:</th>\n",
       "      <td>2.126929e+07</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>pairs with other orientation:</th>\n",
       "      <td>3.455500e+04</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>pairs on different chromosomes:</th>\n",
       "      <td>3.722420e+05</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>percentage of properly paired reads (%):</th>\n",
       "      <td>9.680000e+01</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Uniquely mapped reads:</th>\n",
       "      <td>4.077298e+08</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Reads mapped with MAPQ&gt;30:</th>\n",
       "      <td>3.976403e+08</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                 VIB_1\n",
       "raw total sequences:                      4.533154e+08\n",
       "filtered sequences:                       0.000000e+00\n",
       "sequences:                                4.533154e+08\n",
       "is sorted:                                1.000000e+00\n",
       "1st fragments:                            2.266577e+08\n",
       "last fragments:                           2.266577e+08\n",
       "reads mapped:                             4.430560e+08\n",
       "reads mapped and paired:                  4.405570e+08\n",
       "reads unmapped:                           1.025936e+07\n",
       "reads properly paired:                    4.389387e+08\n",
       "reads paired:                             4.533154e+08\n",
       "reads duplicated:                         8.205966e+07\n",
       "reads MQ0:                                3.450323e+07\n",
       "reads QC failed:                          0.000000e+00\n",
       "non-primary alignments:                   0.000000e+00\n",
       "supplementary alignments:                 1.859000e+03\n",
       "total length:                             2.147508e+10\n",
       "total first fragment length:              1.083189e+10\n",
       "total last fragment length:               1.064319e+10\n",
       "bases mapped:                             2.098839e+10\n",
       "bases mapped (cigar):                     2.088950e+10\n",
       "bases trimmed:                            0.000000e+00\n",
       "bases duplicated:                         3.904326e+09\n",
       "mismatches:                               7.722158e+07\n",
       "error rate:                               3.696670e-03\n",
       "average length:                           4.700000e+01\n",
       "average first fragment length:            4.800000e+01\n",
       "average last fragment length:             4.700000e+01\n",
       "maximum length:                           5.000000e+01\n",
       "maximum first fragment length:            5.000000e+01\n",
       "maximum last fragment length:             4.900000e+01\n",
       "average quality:                          3.590000e+01\n",
       "insert size average:                      1.421000e+02\n",
       "insert size standard deviation:           9.130000e+01\n",
       "inward oriented pairs:                    1.986024e+08\n",
       "outward oriented pairs:                   2.126929e+07\n",
       "pairs with other orientation:             3.455500e+04\n",
       "pairs on different chromosomes:           3.722420e+05\n",
       "percentage of properly paired reads (%):  9.680000e+01\n",
       "Uniquely mapped reads:                    4.077298e+08\n",
       "Reads mapped with MAPQ>30:                3.976403e+08"
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df = pd.read_csv(stats_dict[key], sep='\\t', header=0, index_col=0)\n",
    "df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "89232e9c",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "b6b045a5",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Broad_1    216532213.0\n",
      "Name: raw total sequences:, dtype: float64\n",
      "Broad_1    195423110.5\n",
      "Name: Reads mapped with MAPQ>30:, dtype: float64\n",
      "frac Q30: Broad_1    1.108017\n",
      "dtype: float64\n",
      "Broad_2    211252689.0\n",
      "Name: raw total sequences:, dtype: float64\n",
      "Broad_2    190746130.5\n",
      "Name: Reads mapped with MAPQ>30:, dtype: float64\n",
      "frac Q30: Broad_2    1.107507\n",
      "dtype: float64\n",
      "Broad_mito_1    145604100.0\n",
      "Name: raw total sequences:, dtype: float64\n",
      "Broad_mito_1    127477001.5\n",
      "Name: Reads mapped with MAPQ>30:, dtype: float64\n",
      "frac Q30: Broad_mito_1    1.142199\n",
      "dtype: float64\n",
      "Broad_mito_2    170196511.0\n",
      "Name: raw total sequences:, dtype: float64\n",
      "Broad_mito_2    149189454.5\n",
      "Name: Reads mapped with MAPQ>30:, dtype: float64\n",
      "frac Q30: Broad_mito_2    1.140808\n",
      "dtype: float64\n",
      "CNAG_1    235210366.0\n",
      "Name: raw total sequences:, dtype: float64\n",
      "CNAG_1    211758281.0\n",
      "Name: Reads mapped with MAPQ>30:, dtype: float64\n",
      "frac Q30: CNAG_1    1.110749\n",
      "dtype: float64\n",
      "CNAG_2    266303672.0\n",
      "Name: raw total sequences:, dtype: float64\n",
      "CNAG_2    239931163.0\n",
      "Name: Reads mapped with MAPQ>30:, dtype: float64\n",
      "frac Q30: CNAG_2    1.109917\n",
      "dtype: float64\n",
      "HYA__24010b__20210813_384_PBMC_11_S9    108047316.0\n",
      "Name: raw total sequences:, dtype: float64\n",
      "HYA__24010b__20210813_384_PBMC_11_S9    91256102.0\n",
      "Name: Reads mapped with MAPQ>30:, dtype: float64\n",
      "frac Q30: HYA__24010b__20210813_384_PBMC_11_S9    1.184001\n",
      "dtype: float64\n",
      "HYA__2beafa__20210813_384_PBMC_12_S10    108531243.0\n",
      "Name: raw total sequences:, dtype: float64\n",
      "HYA__2beafa__20210813_384_PBMC_12_S10    90887797.5\n",
      "Name: Reads mapped with MAPQ>30:, dtype: float64\n",
      "frac Q30: HYA__2beafa__20210813_384_PBMC_12_S10    1.194123\n",
      "dtype: float64\n",
      "HYA__3d6da9__20210813_384_PBMC_13_S11    71583227.0\n",
      "Name: raw total sequences:, dtype: float64\n",
      "HYA__3d6da9__20210813_384_PBMC_13_S11    60471946.5\n",
      "Name: Reads mapped with MAPQ>30:, dtype: float64\n",
      "frac Q30: HYA__3d6da9__20210813_384_PBMC_13_S11    1.183743\n",
      "dtype: float64\n",
      "HYA__5028cb__20210813_384_PBMC_14_S12    97860382.0\n",
      "Name: raw total sequences:, dtype: float64\n",
      "HYA__5028cb__20210813_384_PBMC_14_S12    83398741.0\n",
      "Name: Reads mapped with MAPQ>30:, dtype: float64\n",
      "frac Q30: HYA__5028cb__20210813_384_PBMC_14_S12    1.173404\n",
      "dtype: float64\n",
      "Sanger_1    477505161.0\n",
      "Name: raw total sequences:, dtype: float64\n",
      "Sanger_1    432281465.5\n",
      "Name: Reads mapped with MAPQ>30:, dtype: float64\n",
      "frac Q30: Sanger_1    1.104616\n",
      "dtype: float64\n",
      "Sanger_2    498608426.0\n",
      "Name: raw total sequences:, dtype: float64\n",
      "Sanger_2    451154734.0\n",
      "Name: Reads mapped with MAPQ>30:, dtype: float64\n",
      "frac Q30: Sanger_2    1.105183\n",
      "dtype: float64\n",
      "Stanford_1    221145825.0\n",
      "Name: raw total sequences:, dtype: float64\n",
      "Stanford_1    188284802.0\n",
      "Name: Reads mapped with MAPQ>30:, dtype: float64\n",
      "frac Q30: Stanford_1    1.174528\n",
      "dtype: float64\n",
      "Stanford_2    169040262.0\n",
      "Name: raw total sequences:, dtype: float64\n",
      "Stanford_2    143610574.5\n",
      "Name: Reads mapped with MAPQ>30:, dtype: float64\n",
      "frac Q30: Stanford_2    1.177074\n",
      "dtype: float64\n",
      "VIB_1    226657677.0\n",
      "Name: raw total sequences:, dtype: float64\n",
      "VIB_1    198820149.5\n",
      "Name: Reads mapped with MAPQ>30:, dtype: float64\n",
      "frac Q30: VIB_1    1.140014\n",
      "dtype: float64\n",
      "VIB_2    346959872.0\n",
      "Name: raw total sequences:, dtype: float64\n",
      "VIB_2    306352504.5\n",
      "Name: Reads mapped with MAPQ>30:, dtype: float64\n",
      "frac Q30: VIB_2    1.132551\n",
      "dtype: float64\n"
     ]
    }
   ],
   "source": [
    "for key in bc_dict.keys():\n",
    "    # print(key)\n",
    "    df = pd.read_csv(stats_dict[key], sep='\\t', header=0, index_col=0)\n",
    "    \n",
    "    print(f'{df.loc[\"raw total sequences:\"]/2}')\n",
    "    print(f'{df.loc[\"Reads mapped with MAPQ>30:\"]/2}')\n",
    "    print(f'frac Q30: {df.loc[\"raw total sequences:\"]/df.loc[\"Reads mapped with MAPQ>30:\"]}')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "id": "963c9070",
   "metadata": {},
   "outputs": [
    {
     "ename": "SyntaxError",
     "evalue": "invalid syntax (<ipython-input-30-1524ee5774ae>, line 1)",
     "output_type": "error",
     "traceback": [
      "\u001b[0;36m  File \u001b[0;32m\"<ipython-input-30-1524ee5774ae>\"\u001b[0;36m, line \u001b[0;32m1\u001b[0m\n\u001b[0;31m    df.'raw total sequences:'\u001b[0m\n\u001b[0m                            ^\u001b[0m\n\u001b[0;31mSyntaxError\u001b[0m\u001b[0;31m:\u001b[0m invalid syntax\n"
     ]
    }
   ],
   "source": [
    "df.loc['raw total sequences:']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5a1ca7e1",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3.6",
   "language": "python",
   "name": "p36"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.10"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
